// dmdb/src/net/access.js
"use strict";

var __extends = this && this.__extends || function() {
    var n = function(t, r) {
        return (n = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var e in r) r.hasOwnProperty(e) && (t[e] = r[e]);
        })(t, r);
    };
    return function(t, r) {
        function e() {
            this.constructor = t;
        }
        n(t, r), t.prototype = null === r ? Object.create(r) : (e.prototype = r.prototype, 
        new e());
    };
}(), __awaiter = this && this.__awaiter || function(t, u, o, c) {
    return new (o = o || Promise)(function(e, r) {
        function n(t) {
            try {
                s(c.next(t));
            } catch (t) {
                r(t);
            }
        }
        function i(t) {
            try {
                s(c.throw(t));
            } catch (t) {
                r(t);
            }
        }
        function s(t) {
            var r;
            t.done ? e(t.value) : ((r = t.value) instanceof o ? r : new o(function(t) {
                t(r);
            })).then(n, i);
        }
        s((c = c.apply(t, u || [])).next());
    });
}, __generator = this && this.__generator || function(e, n) {
    var i, s, u, o = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: r(0),
        throw: r(1),
        return: r(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function r(r) {
        return function(t) {
            return function(r) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (i = 1, s && (u = 2 & r[0] ? s.return : r[0] ? s.throw || ((u = s.return) && u.call(s), 
                    0) : s.next) && !(u = u.call(s, r[1])).done) return u;
                    switch (s = 0, u && (r = [ 2 & r[0], u.value ]), r[0]) {
                      case 0:
                      case 1:
                        u = r;
                        break;

                      case 4:
                        return o.label++, {
                            value: r[1],
                            done: !1
                        };

                      case 5:
                        o.label++, s = r[1], r = [ 0 ];
                        continue;

                      case 7:
                        r = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = o.trys).length && u[u.length - 1]) && (6 === r[0] || 2 === r[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === r[0] && (!u || r[1] > u[0] && r[1] < u[3])) {
                            o.label = r[1];
                            break;
                        }
                        if (6 === r[0] && o.label < u[1]) {
                            o.label = u[1], u = r;
                            break;
                        }
                        if (u && o.label < u[2]) {
                            o.label = u[2], o.ops.push(r);
                            break;
                        }
                        u[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    r = n.call(e, o);
                } catch (t) {
                    r = [ 6, t ], s = 0;
                } finally {
                    i = u = 0;
                }
                if (5 & r[0]) throw r[1];
                return {
                    value: r[0] ? r[1] : void 0,
                    done: !0
                };
            }([ r, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var os_1 = require("os"), buffer_1 = require("./buffer"), msg_1 = require("./msg/msg"), startup_1 = require("./msg/startup"), error_1 = require("../driver/error"), const_1 = require("../desc/const"), login_1 = require("./msg/login"), executeRetInfo_1 = require("../desc/executeRetInfo"), parameter_1 = require("../desc/parameter"), stmtAllocate_1 = require("./msg/stmtAllocate"), fetch_1 = require("./msg/fetch"), abstractLob_1 = require("../desc/abstractLob"), offRowBinder_1 = require("../convert/offRowBinder"), commit_1 = require("./msg/commit"), rollback_1 = require("./msg/rollback"), events_1 = require("events"), msgParser_1 = require("./msgParser"), prepare_1 = require("./msg/prepare"), execute_1 = require("./msg/execute"), getLobData_1 = require("./msg/getLobData"), bufferUtil_1 = require("../utils/bufferUtil"), setLobData_1 = require("./msg/setLobData"), getLobLen_1 = require("./msg/getLobLen"), lobTruncate_1 = require("./msg/lobTruncate"), compressUtil_1 = require("../utils/compressUtil"), preExec_1 = require("./msg/preExec"), bufferArray_1 = require("../utils/bufferArray"), putData_1 = require("./msg/putData"), msgSecurity_1 = require("../security/msgSecurity"), symmCipher_1 = require("../security/symmCipher"), execDirect_1 = require("./msg/execDirect"), stmtFree_1 = require("./msg/stmtFree"), tableTs_1 = require("./msg/tableTs"), moreResult_1 = require("./msg/moreResult"), putData2_1 = require("./msg/putData2"), offRowBufferBinder_1 = require("../convert/offRowBufferBinder"), data_1 = require("../convert/data"), Access = function(e) {
    function t(t) {
        var r = e.call(this) || this;
        return r.ot = new Array(), r.ft = new msgParser_1.MsgParser({
            onReceive: r._t.bind(r)
        }), r.vt = null, r.wt = !1, r.mt = !1, r.gt = -1, r.dt = !1, r.lt = "", r.bt = 0, 
        r.pt = t, r.yt = r.pt.socket, r;
    }
    return __extends(t, e), t.prototype.qt = function(r) {
        var e = this;
        this.ot.push(r), this.emit("enqueue", r), r.once("error", function(t) {}).once("packet", function(t) {
            r._timer.refresh(), e.emit("data", t);
        }).once("timeout", function() {
            var t = error_1.DBError.ECJS_SOCKET_TIMEOUT();
            t.fatal = !0, e.Pt(t, r);
        }), r.once("end", function() {
            r.removeAllListeners(), e.xt(r);
        }), 1 === this.ot.length && this.At(r);
    }, t.prototype.Y = function() {}, t.prototype.z = function(t) {
        return this.ft.write(t), !0;
    }, t.prototype._t = function(t) {
        var r, e, n, i, s, u;
        bufferUtil_1.BufferUtil.Dt(t.subarray(msg_1.Msg.Lt, msg_1.Msg.Lt + 2)) !== msg_1.Msg.CMD_HEART_TIME_OUT ? (r = this.ot[0]) && (r._timer.stop(), 
        r.St = new buffer_1.DmBuffer(t), r.Bt(), 0 < (e = r.Rt()) && (this.mt && (n = r.St.m(msg_1.Msg.kt, e), 
        i = this.Ut.decrypt(n, !0), r.St.Tt(msg_1.Msg.kt), r.St.It(i), r.Et(i.length)), 
        this.jt(r) && (e = r.Rt(), s = r.St.m(msg_1.Msg.kt + const_1.Const.ULINT_SIZE, e - const_1.Const.ULINT_SIZE), 
        u = compressUtil_1.CompressUtil.uncompress(s, this.pt.conn_prop_compressID), r.St.Tt(msg_1.Msg.kt), 
        r.St.It(u), r.Et(u.length))), r.Y()) : this.ot.forEach(function(t) {
            return t._timer.refresh();
        });
    }, t.prototype.xt = function(t) {
        this.ot.shift();
        var r = this.ot[0];
        r ? this.At(r) : this.emit("drain");
    }, t.prototype.At = function(t) {
        try {
            this.pt.middlewares.log && this.pt.middlewares.log.debug(this.pt, this.pt.rwStandby ? "accessStandby" : "access", t.Ct()), 
            t.Ft();
            var r, e, n = t.Rt();
            if (0 < n && (this.Mt(t) && (r = compressUtil_1.CompressUtil.compress(t.St.m(msg_1.Msg.kt, n), this.pt.conn_prop_compressID), 
            t.St.Tt(msg_1.Msg.kt), t.St.k(n), t.St.It(r), t.Et(const_1.Const.ULINT_SIZE + r.length), 
            t.St.t(msg_1.Msg.Ot, 1)), this.mt && (n = t.Rt(), e = this.Ut.encrypt(t.St.m(msg_1.Msg.kt, n), !0), 
            t.St.Tt(msg_1.Msg.kt), t.St.It(e), t.Et(e.length))), t.St.Gt > msg_1.Msg.zt) throw error_1.DBError.ECJS_MSG_TOO_LONG();
            t.Ht(), t.St.Jt(0), 0 < t._timeout && t._timer.start(t._timeout), t.Kt();
        } catch (t) {
            throw this.ot.shift(), t;
        }
    }, t.prototype.Nt = function(t) {
        this.emit("data", t);
    }, t.prototype.F = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.Qt() ];

                  case 1:
                    return r = t.sent(), 2 !== this.pt.sslEncrypt ? [ 3, 3 ] : [ 4, this.pt.startTLS(!1) ];

                  case 2:
                    return t.sent(), [ 3, 5 ];

                  case 3:
                    return 1 !== this.pt.sslEncrypt ? [ 3, 5 ] : [ 4, this.pt.startTLS(!0) ];

                  case 4:
                    t.sent(), t.label = 5;

                  case 5:
                    return (this.mt || this.wt) && (e = msgSecurity_1.MsgSecurity.computeSessionKey(this.Vt, r.Wt), 
                    this.pt.msgVersion >= msg_1.Msg.VERSION_SSL_UPDATE && this.pt.msgVersion < msg_1.Msg.VERSION_DETERMINE_ENCRYPT ? -1 === this.gt && (this.gt = msgSecurity_1.MsgSecurity.AES256_CFB) : this.pt.msgVersion >= msg_1.Msg.VERSION_DETERMINE_ENCRYPT ? this.gt = 0 === this.pt.algorighm ? msgSecurity_1.MsgSecurity.DES_CFB : msgSecurity_1.MsgSecurity.AES256_CFB : -1 === this.gt && (this.gt = msgSecurity_1.MsgSecurity.DES_CFB), 
                    this.Xt(this.gt, e, this.pt.conn_prop_cipherPath, -1 === r.Yt ? msgSecurity_1.MsgSecurity.MD5 : r.Yt)), 
                    [ 4, this.Zt() ];

                  case 6:
                    return n = t.sent(), this.pt.conn_prop_socketTimeout !== this.bt && 0 < this.bt && (this.pt.conn_prop_socketTimeout = this.bt), 
                    this.pt.msgVersion < msg_1.Msg.VERSION_PUT_DATA2 && (this.pt.lobOffRowLen = 2048), 
                    [ 2, n ];
                }
            });
        });
    }, t.prototype.Qt = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new startup_1.Startup(r, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Zt = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new login_1.Login(r, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.tt = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, e) {
                    n.qt(new stmtAllocate_1.StmtAllocate(n, i, function(t) {
                        t ? e(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.P = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, e) {
                    n.qt(new stmtFree_1.StmtFree(n, i, function(t) {
                        t ? e(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.st = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new prepare_1.Prepare(r, i, !1, const_1.Const.EXECUTE_TYPE_ANY, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.ht = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new execute_1.Execute(r, i, s, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.rt = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new prepare_1.Prepare(r, i, !0, s, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.et = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new execDirect_1.ExecDirect(r, i, s, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.$t = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, new Promise(function(r, e) {
                        n.qt(new preExec_1.PreExec(n, i, i.bindParams, function(t) {
                            t ? e(t) : r();
                        }));
                    }) ];

                  case 1:
                    return t.sent(), i.preExec = !0, [ 2 ];
                }
            });
        });
    }, t.prototype.tr = function(i, s, u, o) {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, e) {
                    n.qt(new putData_1.PutData(n, i, s, u, o, function(t) {
                        t ? e(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.rr = function(i, s, u, o, c) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new putData2_1.PutData2(r, i, s, u, o, c, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.nt = function(u, o, c) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.pt.msgVersion >= msg_1.Msg.VERSION_PUT_DATA2) return [ 2, this.er(u, o) ];
                    r = new Array(u.paramCount), e = 0, t.label = 1;

                  case 1:
                    return e < u.paramCount ? u.bindParams[e].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.nr(r, o, e) ? c ? [ 3, 3 ] : [ 4, this.$t(u) ] : [ 3, 9 ] : [ 3, 10 ];

                  case 2:
                    t.sent(), c = !0, t.label = 3;

                  case 3:
                    return n = o[e], i = new bufferArray_1.BufferArray(), [ 4, n.read(i) ];

                  case 4:
                    t.sent(), s = 0, t.label = 5;

                  case 5:
                    return !n.isReadOver() || 0 < i.length ? !n.isReadOver() && i.length < this.pt.lobReadWriteLen ? [ 4, n.read(i) ] : [ 3, 7 ] : [ 3, 9 ];

                  case 6:
                    t.sent(), t.label = 7;

                  case 7:
                    return s = Math.min(i.length, this.pt.lobReadWriteLen), [ 4, this.tr(u, e, i, s) ];

                  case 8:
                    return t.sent(), [ 3, 5 ];

                  case 9:
                    return e++, [ 3, 1 ];

                  case 10:
                    return [ 2, this.ht(u, [ r ]) ];
                }
            });
        });
    }, t.prototype.er = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = new Array(i.paramCount), e = 0, t.label = 1;

                  case 1:
                    return e < i.paramCount ? i.bindParams[e].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.nr(r, s, e) ? [ 4, this.ir(i, e, s[e]) ] : [ 3, 3 ] : [ 3, 4 ];

                  case 2:
                    n = t.sent(), r[e] = new data_1.Data(data_1.Data.LEN_PUT_DATA2, n), t.label = 3;

                  case 3:
                    return e++, [ 3, 1 ];

                  case 4:
                    return [ 2, this.ht(i, [ r ]) ];
                }
            });
        });
    }, t.prototype.ut = function(_, h, v) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u, o, c, a, f;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.pt.msgVersion >= msg_1.Msg.VERSION_PUT_DATA2) return [ 2, this.sr(_, h, v) ];
                    e = r = 0, n = !1, i = h.length, s = new executeRetInfo_1.ExecuteRetInfo(), u = new Array(), 
                    t.label = 1;

                  case 1:
                    if (!(r < i)) return [ 3, 6 ];
                    for (e = r; e < i; e++) {
                        for (o = h[e], c = new Array(_.paramCount), n = !1, a = 0; a < _.paramCount; a++) if (_.bindParams[a].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.nr(c, o, a)) {
                            n = !0;
                            break;
                        }
                        if (n) break;
                        u.push(c);
                    }
                    return e === r ? [ 3, 3 ] : [ 4, this.ht(_, u) ];

                  case 2:
                    f = t.sent(), u.splice(0), s.union(f, i, r, e - r), t.label = 3;

                  case 3:
                    return e < i ? [ 4, this.nt(_, h[e], v) ] : [ 3, 5 ];

                  case 4:
                    f = t.sent(), v = !0, s.union(f, i, e, 1), t.label = 5;

                  case 5:
                    return r = e + 1, [ 3, 1 ];

                  case 6:
                    return [ 2, s ];
                }
            });
        });
    }, t.prototype.sr = function(m, g, t) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u, o, c, a, f, _, h, v, w;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = msg_1.Msg.zt - 32768, i = g.length, s = new Array(), a = c = o = u = 0, t.label = 1;

                  case 1:
                    if (!(a < i)) return [ 3, 9 ];
                    for (f = g[a], _ = new Array(m.paramCount), h = new Array(m.paramCount), v = o = 0; v < m.paramCount; v++) m.bindParams[v].ioType !== parameter_1.Parameter.IO_TYPE_OUT && (this.nr(_, f, v) ? (h[v] = !0, 
                    o += 2 + putData2_1.PutData2.ur) : _[v] instanceof Buffer ? o += 2 + _[v].length : o += 2);
                    return 0 < a && r < u + o ? [ 4, this.ht(m, s) ] : [ 3, 3 ];

                  case 2:
                    n = t.sent(), void 0 === e ? e = n : e.union(n, i, c, s.length), c = a, s = [], 
                    u = 0, t.label = 3;

                  case 3:
                    v = 0, t.label = 4;

                  case 4:
                    return v < m.bindParams.length ? h[v] ? [ 4, this.ir(m, v, f[v]) ] : [ 3, 6 ] : [ 3, 7 ];

                  case 5:
                    w = t.sent(), _[v] = new data_1.Data(data_1.Data.LEN_PUT_DATA2, w), t.label = 6;

                  case 6:
                    return v++, [ 3, 4 ];

                  case 7:
                    s.push(_), u += o, t.label = 8;

                  case 8:
                    return a++, [ 3, 1 ];

                  case 9:
                    return 0 < s.length ? [ 4, this.ht(m, s) ] : [ 3, 11 ];

                  case 10:
                    n = t.sent(), void 0 === e ? e = n : e.union(n, i, c, s.length), t.label = 11;

                  case 11:
                    return [ 2, e ];
                }
            });
        });
    }, t.prototype.ct = function(o, c, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = c.length, e = new executeRetInfo_1.ExecuteRetInfo(), n = "", i = 0, t.label = 1;

                  case 1:
                    if (!(i < r)) return [ 3, 6 ];
                    t.label = 2;

                  case 2:
                    return t.trys.push([ 2, 4, , 5 ]), [ 4, this.nt(o, c[i], a || 0 !== i) ];

                  case 3:
                    return s = t.sent(), e.union(s, r, i, 1), [ 3, 5 ];

                  case 4:
                    if (u = t.sent(), e.union(0, r, i, 1), !o.connection.conn_prop_continueBatchOnError) throw error_1.DBError.ECJS_BATCH_ERROR(u);
                    return n += "row[" + i + "]:" + u + os_1.EOL, [ 3, 5 ];

                  case 5:
                    return i++, [ 3, 1 ];

                  case 6:
                    if (n) throw error_1.DBError.ECJS_BATCH_ERROR(n);
                    return [ 2, e ];
                }
            });
        });
    }, t.prototype.K = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new fetch_1.Fetch(r, i, BigInt(s), const_1.Const.INT64_MAX, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.R = function() {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, e) {
                    n.qt(new commit_1.Commit(n, function(t) {
                        t ? e(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.I = function() {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, e) {
                    n.qt(new rollback_1.Rollback(n, function(t) {
                        t ? e(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.L = function(s, u, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = Buffer.alloc(o), n = e = 0, t.label = 1;

                  case 1:
                    return e < o ? ((n = o - e) > this.pt.lobReadWriteLen && (n = this.pt.lobReadWriteLen), 
                    [ 4, this.or(s, u + e, n) ]) : [ 3, 3 ];

                  case 2:
                    return (i = t.sent().data) && 0 !== i.length ? (bufferUtil_1.BufferUtil._(r, e, i), 
                    e += i.length, s.readOver ? [ 3, 3 ] : [ 3, 1 ]) : [ 3, 3 ];

                  case 3:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.C = function(o, c, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = "", n = e = 0, t.label = 1;

                  case 1:
                    return e < a ? ((n = a - e) > this.pt.lobReadWriteLen / 2 && (n = this.pt.lobReadWriteLen / 2), 
                    [ 4, this.or(o, c + e, n) ]) : [ 3, 3 ];

                  case 2:
                    return (u = t.sent(), (i = u.data) && 0 !== i.length) ? (s = bufferUtil_1.BufferUtil.u(i, 0, i.length, o.serverEncoding), 
                    r += s, e += -1 === u.charLen ? s.length : u.charLen, o.readOver ? [ 3, 3 ] : [ 3, 1 ]) : [ 3, 3 ];

                  case 3:
                    return [ 2, [ r, e ] ];
                }
            });
        });
    }, t.prototype.j = function(f, _, h, v, w) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u, o, c, a;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    n = e = r = 0, i = w / this.pt.lobReadWriteLen + 1, u = 1, o = 2, c = s = 0, t.label = 1;

                  case 1:
                    return c < i ? ((s = 0) === c && (s |= u), c === i - 1 && (s |= o), (n = w - e) > this.pt.lobReadWriteLen && (n = this.pt.lobReadWriteLen), 
                    [ 4, this.cr(f, s, _, h, v, n) ]) : [ 3, 4 ];

                  case 2:
                    if ((a = t.sent()) <= 0) return [ 3, 4 ];
                    _ += a, r += a, e += n, v += n, t.label = 3;

                  case 3:
                    return c++, [ 3, 1 ];

                  case 4:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.A = function(v, w, m, g) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s, u, o, c, a, f, _, h;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = bufferUtil_1.BufferUtil.o(m, g), e = 0, n = r.length, u = s = i = 0, o = n / this.pt.lobReadWriteLen + 1, 
                    a = 1, f = 2, _ = c = 0, t.label = 1;

                  case 1:
                    return _ < o ? ((c = 0) == _ && (c |= a), _ == o - 1 && (c |= f), (u = n - s) > this.pt.lobReadWriteLen && (u = this.pt.lobReadWriteLen), 
                    [ 4, this.cr(v, c, w, r, e, u) ]) : [ 3, 4 ];

                  case 2:
                    if ((h = t.sent()) <= 0) return [ 3, 4 ];
                    w += h, i += h, s += u, e += u, t.label = 3;

                  case 3:
                    return _++, [ 3, 1 ];

                  case 4:
                    return [ 2, i ];
                }
            });
        });
    }, t.prototype.or = function(i, s, u) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new getLobData_1.GetLobData(r, i, s, u, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.cr = function(i, s, u, o, c, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new setLobData_1.SetLobData(r, i, s, u, o, c, a, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.q = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new getLobLen_1.GetLobLen(r, i, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.O = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new lobTruncate_1.LobTruncate(r, i, s, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Q = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new tableTs_1.TableTS(r, i, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.it = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new moreResult_1.MoreResult(r, i, s, function(t, r) {
                        t ? n(t) : e(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.J = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(e, n) {
                    r.qt(new fetch_1.Fetch(r, i, const_1.Const.INT64_MAX, BigInt(1), function(t, r) {
                        t ? n(t) : e(r.updateCount);
                    }));
                }) ];
            });
        });
    }, t.prototype.ar = function() {
        return this.Vt || (this.Vt = msgSecurity_1.MsgSecurity.newClientKeyPair()), msgSecurity_1.MsgSecurity.getPublicKey(this.Vt);
    }, t.prototype.Xt = function(t, r, e, n) {
        if (0 < t && t < msgSecurity_1.MsgSecurity.MIN_EXTERNAL_CIPHER_ID && r) this.Ut = new symmCipher_1.SymmCipher(t, r); else if (t >= msgSecurity_1.MsgSecurity.MIN_EXTERNAL_CIPHER_ID) throw new Error("not supported");
    }, t.prototype.Mt = function(t) {
        return t.fr !== msg_1.Msg._r && this.pt.conn_prop_compress !== const_1.Const.MSG_COMPRESS_NO && (this.pt.conn_prop_compress === const_1.Const.MSG_COMPRESS_SIMPLE || this.pt.conn_prop_compress === const_1.Const.MSG_COMPRESS_AUTO && (!this.pt.local && t.Rt() > const_1.Const.MSG_COMPRESS_THRESHOLD));
    }, t.prototype.jt = function(t) {
        return t.fr !== msg_1.Msg._r && this.pt.conn_prop_compress !== const_1.Const.MSG_COMPRESS_NO && (this.pt.conn_prop_compress === const_1.Const.MSG_COMPRESS_SIMPLE || this.pt.conn_prop_compress === const_1.Const.MSG_COMPRESS_AUTO && 1 === t.St.a(msg_1.Msg.Ot));
    }, t.prototype.hr = function(t) {
        return t.fr !== msg_1.Msg._r && 1 == this.pt.sslEncrypt;
    }, t.prototype.nr = function(t, r, e) {
        var n = !1;
        if (null === r[e] || void 0 === r[e]) t[e] = null; else if (r[e] instanceof offRowBinder_1.OffRowBinder) {
            if (r[e].isReadOver()) {
                if (!(r[e] instanceof offRowBufferBinder_1.OffRowBufferBinder)) return t[e] = null, 
                !1;
                r[e].reset();
            }
            n = !0, t[e] = Buffer.alloc(0);
            var i, s = r[e].obj;
            s instanceof abstractLob_1.AbstractLob && ((i = s).canOptimized(this.pt) && (t[e] = i.buildCtlData(), 
            n = !1));
        } else t[e] = r[e];
        return n;
    }, t.prototype.ir = function(s, u, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return s.preExec ? [ 3, 2 ] : [ 4, this.$t(s) ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    r = new bufferArray_1.BufferArray(), e = putData2_1.PutData2.createLobHead(), (n = o).read(r), 
                    t.label = 3;

                  case 3:
                    return !n.isReadOver() || 0 < r.length ? (!n.isReadOver() && r.length < this.pt.lobReadWriteLen && n.read(r), 
                    i = Math.min(r.length, this.pt.lobReadWriteLen), [ 4, this.rr(s, u, r, i, e) ]) : [ 3, 5 ];

                  case 4:
                    return e = t.sent(), [ 3, 3 ];

                  case 5:
                    return [ 2, e ];
                }
            });
        });
    }, t.prototype.M = function(t) {
        t instanceof error_1.DBError && (t.fatal = !0);
        var r = this.ot[0];
        r ? r.Y(t) : this.Pt(t);
    }, t.prototype.Pt = function(r, t) {
        var e = this;
        this.vt || r instanceof error_1.DBError && !r.fatal || (this.vt = r, process.nextTick(function() {
            e.ot.forEach(function(t) {
                t.Y(r);
            }), e.ot.length = 0;
        }), this.emit("end", r));
    }, t;
}(events_1.EventEmitter);

exports.Access = Access;
