// dmdb/src/net/buffer.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, e, s, o) {
    return new (s = s || Promise)(function(i, n) {
        function r(t) {
            try {
                f(o.next(t));
            } catch (t) {
                n(t);
            }
        }
        function u(t) {
            try {
                f(o.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function f(t) {
            var n;
            t.done ? i(t.value) : ((n = t.value) instanceof s ? n : new s(function(t) {
                t(n);
            })).then(r, u);
        }
        f((o = o.apply(t, e || [])).next());
    });
}, __generator = this && this.__generator || function(i, r) {
    var u, f, e, s = {
        label: 0,
        sent: function() {
            if (1 & e[0]) throw e[1];
            return e[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (u) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (u = 1, f && (e = 2 & n[0] ? f.return : n[0] ? f.throw || ((e = f.return) && e.call(f), 
                    0) : f.next) && !(e = e.call(f, n[1])).done) return e;
                    switch (f = 0, e && (n = [ 2 & n[0], e.value ]), n[0]) {
                      case 0:
                      case 1:
                        e = n;
                        break;

                      case 4:
                        return s.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        s.label++, f = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(e = 0 < (e = s.trys).length && e[e.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === n[0] && (!e || n[1] > e[0] && n[1] < e[3])) {
                            s.label = n[1];
                            break;
                        }
                        if (6 === n[0] && s.label < e[1]) {
                            s.label = e[1], e = n;
                            break;
                        }
                        if (e && s.label < e[2]) {
                            s.label = e[2], s.ops.push(n);
                            break;
                        }
                        e[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    n = r.call(i, s);
                } catch (t) {
                    n = [ 6, t ], f = 0;
                } finally {
                    u = e = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var bufferUtil_1 = require("../utils/bufferUtil"), msg_1 = require("./msg/msg"), DmBuffer = function() {
    function t(t) {
        this._readOff = 0, this._writeOff = 0, "number" == typeof t ? (t < 0 && (t = 0), 
        this._buf = Buffer.alloc(t)) : (this._buf = Buffer.from(t), this._writeOff = t.length);
    }
    return Object.defineProperty(t.prototype, "tn", {
        get: function() {
            return this._buf.slice(0, this._writeOff);
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "nn", {
        get: function() {
            return this._readOff;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "Gt", {
        get: function() {
            return this._writeOff;
        },
        enumerable: !0,
        configurable: !0
    }), t.prototype.Tt = function(t) {
        return this._writeOff = t, this._readOff > t && (this._readOff = t), this;
    }, t.prototype.Jt = function(t) {
        return this._readOff = t, this;
    }, t.prototype.in = function() {
        return this.Gt - this._readOff;
    }, t.prototype.rn = function(t, n, i) {
        return n ? i ? this.un(t) : this.Tt(t) : i ? this._readOff += t : this._readOff -= t, 
        this;
    }, t.prototype.fn = function(f, e) {
        return __awaiter(this, void 0, void 0, function() {
            var u = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(t, n) {
                    setTimeout(function() {
                        n(new Error("Socket read timeout"));
                    }, 3e3);
                    var i = null, r = setInterval(function() {
                        null !== i && (clearInterval(r), u.It(i), t(i.length));
                    }, 1);
                    f.once("readable", function() {
                        var t = f.read(e), n = t.readInt32LE(msg_1.Msg.en);
                        0 < n && (t = Buffer.concat([ t, f.read(n) ])), i = t;
                    });
                }) ];
            });
        });
    }, t.prototype.sn = function(u) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, i) {
                    u.write(r._buf, function(t) {
                        t ? i(t) : (console.log("send msg"), n(r));
                    });
                }) ];
            });
        });
    }, t.prototype.on = function() {
        return bufferUtil_1.BufferUtil.hn(this.cn(1));
    }, t.prototype.ln = function() {
        return bufferUtil_1.BufferUtil._n(this.cn(1));
    }, t.prototype.bn = function() {
        return bufferUtil_1.BufferUtil.an(this.cn(2));
    }, t.prototype.T = function() {
        return bufferUtil_1.BufferUtil.Dt(this.cn(2));
    }, t.prototype.D = function() {
        return bufferUtil_1.BufferUtil.Un(this.cn(4));
    }, t.prototype.vn = function() {
        return bufferUtil_1.BufferUtil.mn(this.cn(4));
    }, t.prototype.gn = function() {
        return bufferUtil_1.BufferUtil.V(this.cn(8));
    }, t.prototype.dn = function() {
        return bufferUtil_1.BufferUtil.wn(this.cn(8));
    }, t.prototype.yn = function() {
        return bufferUtil_1.BufferUtil.Bn(this.cn(4));
    }, t.prototype.pn = function() {
        return bufferUtil_1.BufferUtil.kn(this.cn(8));
    }, t.prototype.cn = function(t) {
        var n = bufferUtil_1.BufferUtil.m(this._buf, this._readOff, t);
        return this._readOff += t, n;
    }, t.prototype.On = function() {
        return this.cn(this.vn());
    }, t.prototype.jn = function() {
        return this.cn(this.ln());
    }, t.prototype.xn = function() {
        return this.cn(this.T());
    }, t.prototype.Pn = function() {
        for (var t = 0; 0 != this.ln(); ) t++;
        return this.rn(t, !1, !1), this.cn(t);
    }, t.prototype.Sn = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.cn(t), n);
    }, t.prototype.qn = function(t) {
        return bufferUtil_1.BufferUtil.G(this.On(), t);
    }, t.prototype.Dn = function(t) {
        return bufferUtil_1.BufferUtil.G(this.jn(), t);
    }, t.prototype.S = function(t) {
        return bufferUtil_1.BufferUtil.G(this.xn(), t);
    }, t.prototype.En = function(t) {
        return bufferUtil_1.BufferUtil.G(this.Pn(), t);
    }, t.prototype.In = function(t) {
        return this.It(bufferUtil_1.BufferUtil.Mn(t ? 1 : 0));
    }, t.prototype.Tn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.y(t));
    }, t.prototype.Gn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.Mn(t));
    }, t.prototype.x = function(t) {
        return this.It(bufferUtil_1.BufferUtil.p(t));
    }, t.prototype.zn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.An(t));
    }, t.prototype.k = function(t) {
        return this.It(bufferUtil_1.BufferUtil.h(t));
    }, t.prototype.Cn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.Fn(t));
    }, t.prototype.Hn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.l(t));
    }, t.prototype.Jn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.Kn(t));
    }, t.prototype.Ln = function(t) {
        return this.It(bufferUtil_1.BufferUtil.w(t));
    }, t.prototype.Nn = function(t) {
        return this.It(bufferUtil_1.BufferUtil.v(t));
    }, t.prototype.It = function(t) {
        var n = t.length;
        if (0 == n) return n;
        var i, r, u = this._buf.length;
        return this._writeOff + n > u && (i = Math.max(2 * u, this._writeOff + n), r = Buffer.alloc(i), 
        this._buf = r.fill(this._buf, 0, this._buf.length)), this._buf = this._buf.fill(t, this._writeOff, this._writeOff + n), 
        this._writeOff += n, n;
    }, t.prototype.Qn = function(t) {
        return this.Cn(t.length) + this.It(t);
    }, t.prototype.Rn = function(t) {
        return this.Gn(t.length) + this.It(t);
    }, t.prototype.Vn = function(t) {
        return this.zn(t.length) + this.It(t);
    }, t.prototype.Wn = function(t) {
        return this.It(t) + this.Gn(0);
    }, t.prototype.Xn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.o(t, n);
        return this.Qn(i);
    }, t.prototype.Yn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.o(t, n);
        return this.Rn(i);
    }, t.prototype.N = function(t, n) {
        var i = bufferUtil_1.BufferUtil.o(t, n);
        return this.Vn(i);
    }, t.prototype.Zn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.o(t, n);
        return this.Wn(i);
    }, t.prototype.t = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.y(n));
    }, t.prototype.$n = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.Mn(n));
    }, t.prototype.B = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.p(n));
    }, t.prototype.e = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.An(n));
    }, t.prototype.f = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.h(n));
    }, t.prototype.n = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.Fn(n));
    }, t.prototype.r = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.l(n));
    }, t.prototype.ti = function(t, n) {
        return this._(t, bufferUtil_1.BufferUtil.Kn(n));
    }, t.prototype._ = function(t, n) {
        return this._buf.set(n, t), n.length;
    }, t.prototype.ni = function(t, n) {
        return this.n(t, n.length) + this._(t + 4, n);
    }, t.prototype.ii = function(t, n) {
        return this.$n(t, n.length) + this._(t + 1, n);
    }, t.prototype.ri = function(t, n) {
        return this.e(t, n.length) + this._(t + 2, n);
    }, t.prototype.ui = function(t, n) {
        return this._(t, n) + this.$n(t + n.length, 0);
    }, t.prototype.fi = function(t, n, i) {
        return this.ni(t, bufferUtil_1.BufferUtil.o(n, i));
    }, t.prototype.ei = function(t, n, i) {
        return this.ii(t, bufferUtil_1.BufferUtil.o(n, i));
    }, t.prototype.si = function(t, n, i) {
        return this.ri(t, bufferUtil_1.BufferUtil.o(n, i));
    }, t.prototype.oi = function(t, n, i) {
        return this.ui(t, bufferUtil_1.BufferUtil.o(n, i));
    }, t.prototype.a = function(t) {
        return bufferUtil_1.BufferUtil.hn(this.m(t, 1));
    }, t.prototype.H = function(t) {
        return bufferUtil_1.BufferUtil._n(this.m(t, 1));
    }, t.prototype.c = function(t) {
        return bufferUtil_1.BufferUtil.an(this.m(t, 2));
    }, t.prototype.U = function(t) {
        return bufferUtil_1.BufferUtil.Dt(this.m(t, 2));
    }, t.prototype.s = function(t) {
        return bufferUtil_1.BufferUtil.Un(this.m(t, 4));
    }, t.prototype.g = function(t) {
        return bufferUtil_1.BufferUtil.mn(this.m(t, 4));
    }, t.prototype.i = function(t) {
        return bufferUtil_1.BufferUtil.V(this.m(t, 8));
    }, t.prototype.hi = function(t) {
        return bufferUtil_1.BufferUtil.wn(this.m(t, 8));
    }, t.prototype.d = function(t) {
        return bufferUtil_1.BufferUtil.Bn(this.m(t, 4));
    }, t.prototype.b = function(t) {
        return bufferUtil_1.BufferUtil.kn(this.m(t, 8));
    }, t.prototype.m = function(t, n) {
        return this._buf.subarray(t, t + n);
    }, t.prototype.ci = function(t) {
        var n = this.g(t);
        return this.m(t + 4, n);
    }, t.prototype.li = function(t) {
        var n = this.H(t);
        return this.m(t + 1, n);
    }, t.prototype._i = function(t) {
        var n = this.U(t);
        return this.m(t + 2, n);
    }, t.prototype.bi = function(t) {
        for (var n = 0; 0 !== this.H(t); ) t++, n++;
        return this.m(t - n, n);
    }, t.prototype.ai = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.ci(t), n);
    }, t.prototype.Ui = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.li(t), n);
    }, t.prototype.vi = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this._i(t), n);
    }, t.prototype.mi = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.bi(t), n);
    }, t.prototype.un = function(t) {
        var n, i = this._buf.length;
        this._writeOff + t > i && (n = Buffer.alloc(2 * i), this._buf = n.fill(this._buf, 0, i));
        for (var r = 0; r < t; r++) this._buf.writeUInt8(0, this._writeOff++);
        return this;
    }, t;
}();

exports.DmBuffer = DmBuffer;
