// dmdb/src/middleware/reconnect/reconnectMiddleware.js
"use strict";

var __awaiter = this && this.__awaiter || function(r, c, o, a) {
    return new (o = o || Promise)(function(e, t) {
        function n(r) {
            try {
                u(a.next(r));
            } catch (r) {
                t(r);
            }
        }
        function i(r) {
            try {
                u(a.throw(r));
            } catch (r) {
                t(r);
            }
        }
        function u(r) {
            var t;
            r.done ? e(r.value) : ((t = r.value) instanceof o ? t : new o(function(r) {
                r(t);
            })).then(n, i);
        }
        u((a = a.apply(r, c || [])).next());
    });
}, __generator = this && this.__generator || function(e, n) {
    var i, u, c, o = {
        label: 0,
        sent: function() {
            if (1 & c[0]) throw c[1];
            return c[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: t(0),
        throw: t(1),
        return: t(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function t(t) {
        return function(r) {
            return function(t) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (i = 1, u && (c = 2 & t[0] ? u.return : t[0] ? u.throw || ((c = u.return) && c.call(u), 
                    0) : u.next) && !(c = c.call(u, t[1])).done) return c;
                    switch (u = 0, c && (t = [ 2 & t[0], c.value ]), t[0]) {
                      case 0:
                      case 1:
                        c = t;
                        break;

                      case 4:
                        return o.label++, {
                            value: t[1],
                            done: !1
                        };

                      case 5:
                        o.label++, u = t[1], t = [ 0 ];
                        continue;

                      case 7:
                        t = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(c = 0 < (c = o.trys).length && c[c.length - 1]) && (6 === t[0] || 2 === t[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === t[0] && (!c || t[1] > c[0] && t[1] < c[3])) {
                            o.label = t[1];
                            break;
                        }
                        if (6 === t[0] && o.label < c[1]) {
                            o.label = c[1], c = t;
                            break;
                        }
                        if (c && o.label < c[2]) {
                            o.label = c[2], o.ops.push(t);
                            break;
                        }
                        c[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    t = n.call(e, o);
                } catch (r) {
                    t = [ 6, r ], u = 0;
                } finally {
                    i = c = 0;
                }
                if (5 & t[0]) throw t[1];
                return {
                    value: t[0] ? t[1] : void 0,
                    done: !0
                };
            }([ t, r ]);
        };
    }
}, __values = this && this.__values || function(r) {
    var t = "function" == typeof Symbol && Symbol.iterator, e = t && r[t], n = 0;
    if (e) return e.call(r);
    if (r && "number" == typeof r.length) return {
        next: function() {
            return r && n >= r.length && (r = void 0), {
                value: r && r[n++],
                done: !r
            };
        }
    };
    throw new TypeError(t ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var error_1 = require("../../driver/error"), const_1 = require("../../desc/const"), ep_1 = require("../../desc/ep"), stringUtil_1 = require("../../utils/stringUtil"), rwMiddleware_1 = require("../rw/rwMiddleware"), ReconnectMiddleware = function() {
    function c() {}
    return c.getInstance = function() {
        return this.instance;
    }, c.reconnect = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 5, , 6 ]), 0 < t.conn_prop_rwSeparate ? [ 4, rwMiddleware_1.RWMiddleware.getInstance(t.conn_prop_rwSeparate).rwUtil.reconnect(t) ] : [ 3, 2 ];

                  case 1:
                    return r.sent(), [ 3, 4 ];

                  case 2:
                    return [ 4, t.do_reconnect() ];

                  case 3:
                    r.sent(), r.label = 4;

                  case 4:
                    return [ 3, 6 ];

                  case 5:
                    throw r.sent(), error_1.DBError.ECJS_CONNECTION_SWITCH_FAILED(e);

                  case 6:
                    throw t.closed = !1, error_1.DBError.ECJS_CONNECTION_SWITCHED(e);
                }
            });
        });
    }, c.prototype.autoReconnect = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                if (e instanceof error_1.DBError && (e.errCode === error_1.DBError.ECJS_COMMUNICATION_ERROR().errCode || e.errCode === error_1.DBError.ECJS_CONNECTION_CLOSED().errCode)) return [ 2, c.reconnect(t, e.message) ];
                throw e;
            });
        });
    }, c.prototype.checkAndRecover = function(h) {
        return __awaiter(this, void 0, void 0, function() {
            var t, e, n, i, u, c, o, a, s, _, f;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return h.conn_prop_doSwitch !== const_1.Const.DO_SWITCH_WHEN_EP_RECOVER ? [ 2 ] : (t = 0, 
                    !h.trxFinish || 0 === (t = h.getIndexOnEPGroup()) || Date.now() - h.middlewares.recoverInfo.checkEpRecoverTs < h.conn_prop_switchInterval ? [ 2 ] : h.conn_prop_cluster !== const_1.Const.CLUSTER_TYPE_DSC ? [ 3, 2 ] : [ 4, this.loadDscEpSites(h) ]);

                  case 1:
                    return n = r.sent(), [ 3, 3 ];

                  case 2:
                    n = null, r.label = 3;

                  case 3:
                    if (!(e = n) || 0 === e.length) return [ 2 ];
                    i = !1;
                    try {
                        for (u = __values(e), c = u.next(); !c.done; c = u.next()) if ((o = c.value).epStatus === const_1.Const.EP_STATUS_OK) {
                            for (a = 0; a < t; a++) if (s = h.epGroup.epList[a], o.host === s.host && o.port === s.port) {
                                i = !0;
                                break;
                            }
                            if (i) break;
                        }
                    } catch (r) {
                        _ = {
                            error: r
                        };
                    } finally {
                        try {
                            c && !c.done && (f = u.return) && f.call(u);
                        } finally {
                            if (_) throw _.error;
                        }
                    }
                    return h.middlewares.recoverInfo.checkEpRecoverTs = Date.now(), i ? [ 2, h.do_reconnect() ] : [ 2 ];
                }
            });
        });
    }, c.prototype.loadDscEpSites = function(u) {
        return __awaiter(this, void 0, void 0, function() {
            var t, e, n, i;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 6, , 7 ]), [ 4, u.do_executeInner(c.SQL_GET_DSC_EP_SITE) ];

                  case 1:
                    t = r.sent(), e = [], n = void 0, r.label = 2;

                  case 2:
                    return [ 4, t.resultSet.do_getRow() ];

                  case 3:
                    return (n = r.sent()) ? ((i = new ep_1.EP(n[1], n[2])).epSeqno = n[0], i.epStatus = stringUtil_1.StringUtil.equalsIgnoreCase(n[3], "ok") ? const_1.Const.EP_STATUS_OK : const_1.Const.EP_STATUS_ERROR, 
                    e.push(i), [ 3, 2 ]) : [ 3, 4 ];

                  case 4:
                    return [ 4, t.resultSet.do_close() ];

                  case 5:
                    return r.sent(), [ 2, e ];

                  case 6:
                    return r.sent(), [ 3, 7 ];

                  case 7:
                    return [ 2, null ];
                }
            });
        });
    }, c.prototype.connection_openConnection = function(t, e, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_openConnection(e, n) ];
            });
        });
    }, c.prototype.connection_execute = function(e, n, i, u, c) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 3, , 5 ]), [ 4, this.checkAndRecover(n) ];

                  case 1:
                    return r.sent(), [ 4, e.connection_execute(n, i, u, c) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return t = r.sent(), [ 4, this.autoReconnect(n, t) ];

                  case 4:
                    return r.sent(), [ 2, null ];

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.connection_executeMany = function(e, n, i, u, c) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 3, , 5 ]), [ 4, this.checkAndRecover(n) ];

                  case 1:
                    return r.sent(), [ 4, e.connection_executeMany(n, i, u, c) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return t = r.sent(), [ 4, this.autoReconnect(n, t) ];

                  case 4:
                    return r.sent(), [ 2, null ];

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.connection_createLob = function(t, e, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_createLob(e, n) ];
            });
        });
    }, c.prototype.connection_close = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 3 ]), [ 4, e.connection_close(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 2, this.autoReconnect(n, t) ];

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.connection_free = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return [ 4, t.connection_free(e) ];

                  case 1:
                    return [ 2, r.sent() ];
                }
            });
        });
    }, c.prototype.connection_commit = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 3, , 4 ]), [ 4, e.connection_commit(n) ];

                  case 1:
                    return r.sent(), [ 4, this.checkAndRecover(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return t = r.sent(), [ 2, this.autoReconnect(n, t) ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.connection_getStatementInfo = function(e, n, i) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 3, , 5 ]), [ 4, this.checkAndRecover(n) ];

                  case 1:
                    return r.sent(), [ 4, e.connection_getStatementInfo(n, i) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return t = r.sent(), [ 4, this.autoReconnect(n, t) ];

                  case 4:
                    return r.sent(), [ 2, null ];

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.connection_rollback = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 3 ]), [ 4, e.connection_rollback(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 2, this.autoReconnect(n, t) ];

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.lob_close = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.lob_close(e) ];
            });
        });
    }, c.prototype.lob_getData = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 4 ]), [ 4, e.lob_getData(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 4, this.autoReconnect(n.iLob.conn, t) ];

                  case 3:
                    return r.sent(), [ 2, null ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.lob_getLength = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 4 ]), [ 4, e.lob_getLength(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 4, this.autoReconnect(n.iLob.conn, t) ];

                  case 3:
                    return r.sent(), [ 2, 0 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.pool_getConnection = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.pool_getConnection(e) ];
            });
        });
    }, c.prototype.pool_close = function(t, e, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.pool_close(e, n) ];
            });
        });
    }, c.prototype.resultSet_close = function(t, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.resultSet_close(e) ];
            });
        });
    }, c.prototype.resultSet_getRow = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 3 ]), [ 4, e.resultSet_getRow(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 2, this.autoReconnect(n.conn, t) ];

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.resultSet_getRows = function(e, n, i) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 4 ]), [ 4, e.resultSet_getRows(n, i) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 4, this.autoReconnect(n.conn, t) ];

                  case 3:
                    return r.sent(), [ 2, [] ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, c.prototype.resultSet_getRowCount = function(e, n) {
        return __awaiter(this, void 0, void 0, function() {
            var t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, , 4 ]), [ 4, e.resultSet_getRowCount(n) ];

                  case 1:
                    return [ 2, r.sent() ];

                  case 2:
                    return t = r.sent(), [ 4, this.autoReconnect(n.conn, t) ];

                  case 3:
                    return r.sent(), [ 2, 0 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, c.SQL_GET_DSC_EP_SITE = "SELECT dsc.ep_seqno, (CASE mal.MAL_INST_HOST WHEN '' THEN mal.MAL_HOST ELSE mal.MAL_INST_HOST END) as ep_host, dcr.EP_PORT, dsc.EP_STATUS FROM V$DSC_EP_INFO dsc LEFT join V$DM_MAL_INI mal on dsc.EP_NAME = mal.MAL_INST_NAME LEFT join (SELECT grp.GROUP_TYPE GROUP_TYPE, ep.* FROM SYS.\"V$DCR_GROUP\" grp, SYS.\"V$DCR_EP\" ep where grp.GROUP_NAME = ep.GROUP_NAME) dcr on dsc.EP_NAME = dcr.EP_NAME and GROUP_TYPE = 'DB' order by  dsc.ep_seqno asc;", 
    c.instance = new c(), c;
}();

exports.ReconnectMiddleware = ReconnectMiddleware;
