// dmdb/src/dm.js
"use strict";

var __awaiter = this && this.__awaiter || function(r, u, c, f) {
    return new (c = c || Promise)(function(n, e) {
        function t(r) {
            try {
                i(f.next(r));
            } catch (r) {
                e(r);
            }
        }
        function o(r) {
            try {
                i(f.throw(r));
            } catch (r) {
                e(r);
            }
        }
        function i(r) {
            var e;
            r.done ? n(r.value) : ((e = r.value) instanceof c ? e : new c(function(r) {
                r(e);
            })).then(t, o);
        }
        i((f = f.apply(r, u || [])).next());
    });
}, __generator = this && this.__generator || function(n, t) {
    var o, i, u, c = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: e(0),
        throw: e(1),
        return: e(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function e(e) {
        return function(r) {
            return function(e) {
                if (o) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (o = 1, i && (u = 2 & e[0] ? i.return : e[0] ? i.throw || ((u = i.return) && u.call(i), 
                    0) : i.next) && !(u = u.call(i, e[1])).done) return u;
                    switch (i = 0, u && (e = [ 2 & e[0], u.value ]), e[0]) {
                      case 0:
                      case 1:
                        u = e;
                        break;

                      case 4:
                        return c.label++, {
                            value: e[1],
                            done: !1
                        };

                      case 5:
                        c.label++, i = e[1], e = [ 0 ];
                        continue;

                      case 7:
                        e = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = c.trys).length && u[u.length - 1]) && (6 === e[0] || 2 === e[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === e[0] && (!u || e[1] > u[0] && e[1] < u[3])) {
                            c.label = e[1];
                            break;
                        }
                        if (6 === e[0] && c.label < u[1]) {
                            c.label = u[1], u = e;
                            break;
                        }
                        if (u && c.label < u[2]) {
                            c.label = u[2], c.ops.push(e);
                            break;
                        }
                        u[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    e = t.call(n, c);
                } catch (r) {
                    e = [ 6, r ], i = 0;
                } finally {
                    o = u = 0;
                }
                if (5 & e[0]) throw e[1];
                return {
                    value: e[0] ? e[1] : void 0,
                    done: !0
                };
            }([ e, r ]);
        };
    }
}, __read = this && this.__read || function(r, e) {
    var n = "function" == typeof Symbol && r[Symbol.iterator];
    if (!n) return r;
    var t, o, i = n.call(r), u = [];
    try {
        for (;(void 0 === e || 0 < e--) && !(t = i.next()).done; ) u.push(t.value);
    } catch (r) {
        o = {
            error: r
        };
    } finally {
        try {
            t && !t.done && (n = i.return) && n.call(i);
        } finally {
            if (o) throw o.error;
        }
    }
    return u;
}, __spread = this && this.__spread || function() {
    for (var r = [], e = 0; e < arguments.length; e++) r = r.concat(__read(arguments[e]));
    return r;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var dmdb, connection_1 = require("./driver/connection"), error_1 = require("./driver/error"), properties_1 = require("./desc/properties"), url_1 = require("url"), svcConf_1 = require("./desc/svcConf"), const_1 = require("./desc/const"), lob_1 = require("./driver/lob"), pool_1 = require("./driver/pool"), epGroup_1 = require("./desc/epGroup"), ep_1 = require("./desc/ep"), driver_1 = require("./driver");

!function(t) {
    function r(r) {
        return null != r && (r.connectString || r.connectionString);
    }
    function o(i) {
        return __awaiter(this, void 0, void 0, function() {
            var e, n, t, o;
            return __generator(this, function(r) {
                if (!i.startsWith("dm://")) return [ 2, Promise.reject(new Error("URL must start with 'dm://'")) ];
                for (t in e = url_1.parse(i.trim(), !0), (n = new properties_1.Properties()).set("url", i), 
                e.auth && e.auth.split(":")[0] && (n.set("user", e.auth.split(":")[0]), e.auth.split(":")[1] && n.set("password", e.auth.split(":")[1])), 
                e.query) o = e.query[t], n.set(t, o instanceof Array ? o[o.length - 1] : o);
                return [ 2, u(n, e.host) ];
            });
        });
    }
    function i(o) {
        return __awaiter(this, void 0, void 0, function() {
            var e, n, t;
            return __generator(this, function(r) {
                for (t in o = o || {}, e = new properties_1.Properties(), n = o.connectString || o.connectionString || "", 
                e.setAny("url", n), o) e.setAny(t, o[t]);
                return [ 2, u(e, n) ];
            });
        });
    }
    function u(v, p) {
        return __awaiter(this, void 0, void 0, function() {
            var e, n, t, o, i, u, c, f, s, a, _, d;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return (n = v.getTrimString("svcConfPath", "")) ? [ 4, svcConf_1.SvcConf.load(n, !1) ] : [ 3, 2 ];

                  case 1:
                    return e = r.sent(), [ 3, 5 ];

                  case 2:
                    return svcConf_1.SvcConf.GLOBAL ? [ 3, 4 ] : (t = svcConf_1.SvcConf, [ 4, svcConf_1.SvcConf.load(n, !1) ]);

                  case 3:
                    t.GLOBAL = r.sent(), r.label = 4;

                  case 4:
                    e = svcConf_1.SvcConf.GLOBAL, r.label = 5;

                  case 5:
                    return o = v.getTrimString("addressRemap", "") || e.globalProperties.getTrimString("addressRemap", ""), 
                    p = l(p, o), i = __read([ "", "" ], 2), u = i[0], c = i[1], f = p.lastIndexOf(":"), 
                    c = -1 !== f ? (u = p.substring(0, f), p.substring(f + 1)) : (u = p, const_1.Const.PORT_DEFAULT.toString()), 
                    v.set("host", u), v.set("port", c), s = v.getTrimString("userRemap", "") || e.globalProperties.getTrimString("userRemap", ""), 
                    v.set("user", l(v.getTrimString("user", ""), s)), "" !== (_ = v.getTrimString(p, "")) ? (_.startsWith("(") && _.endsWith(")") && (_ = _.slice(1, _.length - 1)), 
                    d = p + "=" + _, (a = driver_1.Driver.epGroupMap.get(d)) || (a = svcConf_1.SvcConf.parseServerName(p, _)) && driver_1.Driver.epGroupMap.set(d, a)) : e.serverGroupMap.has(p) && (a = e.serverGroupMap.get(p)), 
                    a = a || new epGroup_1.EPGroup(u + ":" + c, [ new ep_1.EP(u, Number.parseInt(c)) ]), 
                    v.setDiffProps(a.props), 0 < v.getNumber("rwSeparate", const_1.Const.RW_SEPARATE_OFF) && (v.set("loginMode", const_1.Const.LOGIN_MODE_PRIMARY_ONLY.toString()), 
                    v.set("loginStatus", const_1.Const.SERVER_STATUS_OPEN.toString()), v.set("doSwitch", "true")), 
                    v.setAny("epGroup", a), v.setDiffProps(e.globalProperties), [ 2, v ];
                }
            });
        });
    }
    function l(r, e) {
        if (!r || !e) return r;
        for (var n = /\(.*?,.*?\)/g, t = null; t = n.exec(e); ) {
            var o = t[0].substring(1, t[0].length - 1).split(",");
            if (o[0].trim() === r) return o[1].trim();
        }
        return r;
    }
    t.EXEC_MANY_FAIL = -1, t.OUT_FORMAT_ARRAY = 4001, t.OUT_FORMAT_OBJECT = 4002, t.OUT_FORMAT_AUTO = 4003, 
    t.BLOB = 2019, t.BUFFER = 2006, t.CLOB = 2017, t.CURSOR = 2021, t.DATE = 2014, t.DEFAULT = 0, 
    t.NUMBER = 2010, t.STRING = 2001, t.BIND_IN = 3001, t.BIND_INOUT = 3002, t.BIND_OUT = 3003, 
    t.POOL_STATUS_OPEN = 6e3, t.POOL_STATUS_DRAINING = 6001, t.POOL_STATUS_CLOSED = 6002, 
    t.STMT_TYPE_UNKNOWN = 0, t.STMT_TYPE_SELECT = 1, t.STMT_TYPE_UPDATE = 2, t.STMT_TYPE_DELETE = 3, 
    t.STMT_TYPE_INSERT = 4, t.STMT_TYPE_CREATE = 5, t.STMT_TYPE_DROP = 6, t.STMT_TYPE_ALTER = 7, 
    t.STMT_TYPE_BEGIN = 8, t.STMT_TYPE_DECLARE = 9, t.STMT_TYPE_CALL = 10, t.STMT_TYPE_EXPLAIN_PLAN = 15, 
    t.STMT_TYPE_MERGE = 16, t.STMT_TYPE_ROLLBACK = 17, t.STMT_TYPE_COMMIT = 21, t.extendedMetaData = !1, 
    t.outFormat = t.OUT_FORMAT_ARRAY, t.outBindFormat = t.OUT_FORMAT_AUTO, t.poolMax = 4, 
    t.poolMin = 0, t.poolTimeout = 60, t.queueMax = 500, t.queueRequests = !0, t.queueTimeout = 6e4, 
    t.fetchAsBuffer = [], t.fetchAsString = [], t.parseJson = !0, t.allowDebug = !1, 
    t.pools = new Map(), t.getConnection = function(r, e) {
        if (void 0 === e) {
            var n;
            if (r instanceof Object) return i(r).then(function(r) {
                return n = r, connection_1.Connection.newInstanceOfConnection(n).openConnection(n.getAny("epGroup"));
            });
            if ("string" != typeof r) throw error_1.DBError.ECJS_INVALID_URL_OR_ALIAS_OR_ATTR();
            if (r.startsWith("dm://")) return o(r).then(function(r) {
                return n = r, connection_1.Connection.newInstanceOfConnection(n).openConnection(n.getAny("epGroup"));
            });
            if (!t.pools.has(r)) throw error_1.DBError.ECJS_INVALID_URL_OR_ALIAS_OR_ATTR();
            return t.pools.get(r).getConnection();
        }
        if ("function" != typeof e) throw error_1.DBError.ECJS_INVALID_PARAM();
        t.getConnection(r).then(function(r) {
            e(null, r);
        }).catch(function(r) {
            e(r, void 0);
        });
    }, t.createPool = function(e, n) {
        if (!r(e)) throw error_1.DBError.ECJS_INVALID_PARAM();
        if (void 0 === n) return new pool_1.Pool(e).openPool().then(function(r) {
            if (e.poolAlias) {
                if (t.pools.has(e.poolAlias)) throw error_1.DBError.ECJS_POOL_ALIAS_CONFLICT();
                t.pools.set(e.poolAlias, r);
            }
            return r;
        });
        if ("function" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
        new pool_1.Pool(e).openPool().then(function(r) {
            if (e.poolAlias) {
                if (t.pools.has(e.poolAlias)) throw error_1.DBError.ECJS_POOL_ALIAS_CONFLICT();
                t.pools.set(e.poolAlias, r);
            }
            n(null, r);
        }).catch(function(r) {
            n(r, void 0);
        });
    }, t.getPool = function(r) {
        if (t.pools.has(r)) return t.pools.get(r);
    }, t.isPoolAttributes = r, t.isBindParameter = function(r) {
        return null != r && (!("string" == typeof r || "number" == typeof r || r instanceof lob_1.Lob || r instanceof Date || r instanceof BigInt) && (void 0 !== r.dir || void 0 !== r.maxArraySize || void 0 !== r.maxSize || void 0 !== r.type || void 0 !== r.val));
    }, t.parseUrl = o, t.parseConnectionAttr = i, t.debug = function() {
        for (var r = [], e = 0; e < arguments.length; e++) r[e] = arguments[e];
        t.allowDebug && console.log.apply(console, __spread(r));
    };
}(dmdb = exports.dmdb || (exports.dmdb = {}));
