// dmdb/src/middleware/log/logWriter.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, n) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, n) {
            t.__proto__ = n;
        } || function(t, n) {
            for (var i in n) n.hasOwnProperty(i) && (t[i] = n[i]);
        })(t, n);
    };
    return function(t, n) {
        function i() {
            this.constructor = t;
        }
        r(t, n), t.prototype = null === n ? Object.create(n) : (i.prototype = n.prototype, 
        new i());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var writer_1 = require("../writer"), LogWriter = function(r) {
    function t(t, n) {
        var i = r.call(this, t, "dm_js", n) || this;
        return i.start(), i;
    }
    return __extends(t, r), t.prototype.start = function() {
        var t = this;
        0 < this.flushFreq && (this.intervalId = setInterval(function() {
            t.doFlush(t.buffer);
        }, 1e3 * this.flushFreq));
    }, t.prototype.end = function() {
        this.intervalId && (clearInterval(this.intervalId), 0 < this.buffer.length && this.doFlush(this.buffer), 
        this.closeCurrentFile());
    }, t.prototype.write = function(t) {
        var n = Buffer.from(t);
        this.buffer.putBuffer(n, 0, n.length);
    }, t.prototype.writeLine = function(t) {
        this.write(t + "\n");
    }, t;
}(writer_1.Writer);

exports.LogWriter = LogWriter;
