// dmdb/src/parser/sqlProcessor.js
"use strict";

var __values = this && this.__values || function(r) {
    var t = "function" == typeof Symbol && Symbol.iterator, e = t && r[t], i = 0;
    if (e) return e.call(r);
    if (r && "number" == typeof r.length) return {
        next: function() {
            return r && i >= r.length && (r = void 0), {
                value: r && r[i++],
                done: !r
            };
        }
    };
    throw new TypeError(t ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var lval_1 = require("./lval"), lexer_1 = require("./lexer"), stringUtil_1 = require("../utils/stringUtil"), dbtype_1 = require("../driver/dbtype"), const_1 = require("../desc/const"), bufferUtil_1 = require("../utils/bufferUtil"), numeric_1 = require("../driver/numeric"), error_1 = require("../driver/error"), dm_1 = require("../dm"), SQLProcessor = function() {
    function p() {}
    return p.lex = function(r) {
        for (var t, e = new lexer_1.Lexer(r, !1), i = 0, a = new Array(); null != (t = e.yylex()); ) t.position = i++, 
        a.push(t);
        return a;
    }, p.escape = function(r, t) {
        var e, i;
        if (0 === t.length && !r.includes("{")) return r;
        var a = new Map();
        if (0 < t.length) try {
            for (var l = __values(t), n = l.next(); !n.done; n = l.next()) {
                var _ = n.value;
                a.set(_.toLowerCase(), null);
            }
        } catch (r) {
            e = {
                error: r
            };
        } finally {
            try {
                n && !n.done && (i = l.return) && i.call(l);
            } finally {
                if (e) throw e.error;
            }
        }
        for (var s = "", u = new Array(), o = this.lex(r), f = 0; f < o.length; f++) {
            var v, c, d, b = o[f];
            b.type == lval_1.Type.NORMAL ? "{" === b.value ? (v = this.next(o, f + 1)) && v.type == lval_1.Type.NORMAL ? stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "escape") || stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "call") ? u.push(!0) : stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "oj") ? (u.push(!0), 
            v.value = "", v.type = lval_1.Type.WHITESPACE_OR_COMMENT) : stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "d") ? (u.push(!0), 
            v.value = "date") : stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "t") ? (u.push(!0), 
            v.value = "time") : stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "ts") ? (u.push(!0), 
            v.value = "datetime") : stringUtil_1.StringUtil.equalsIgnoreCase(v.value, "fn") ? (u.push(!0), 
            v.value = "", v.type = lval_1.Type.WHITESPACE_OR_COMMENT, (c = p.next(o, v.position + 1)) && c.type == lval_1.Type.NORMAL && stringUtil_1.StringUtil.equalsIgnoreCase(c.value, "database") && (c.value = "cur_database")) : "?" === v.value && (c = p.next(o, v.position + 1)) && c.type == lval_1.Type.NORMAL && stringUtil_1.StringUtil.equalsIgnoreCase(c.value, "=") && (d = p.next(o, c.position + 1)) && d.type == lval_1.Type.NORMAL && stringUtil_1.StringUtil.equalsIgnoreCase(d.value, "call") ? (u.push(!0), 
            d.value = "", d.type = lval_1.Type.WHITESPACE_OR_COMMENT) : (u.push(!1), s += b.value) : (u.push(!1), 
            s += b.value) : "}" === b.value ? 0 < u.length && u.pop() || (s += b.value) : a && a.has(b.value.toLowerCase()) ? s += '"' + stringUtil_1.StringUtil.processDoubleQuoteOfName(b.value.toUpperCase()) + '"' : s += b.value : b.type == lval_1.Type.STRING ? s += "'" + stringUtil_1.StringUtil.processSingleQuoteOfName(b.value) + "'" : s += b.value;
        }
        return s;
    }, p.execOpt = function(r, t, e, i) {
        var a = "", l = this.lex(r);
        if (!l || 0 === l.length) return r;
        var n = l[0].value;
        if (!(stringUtil_1.StringUtil.equalsIgnoreCase(n, "insert") || stringUtil_1.StringUtil.equalsIgnoreCase(n, "select") || stringUtil_1.StringUtil.equalsIgnoreCase(n, "update") || stringUtil_1.StringUtil.equalsIgnoreCase(n, "delete"))) return r;
        for (var _ = 0; _ < l.length; _++) {
            var s = l[_], u = void 0;
            switch (s.type) {
              case lval_1.Type.NULL:
                a += "?", t.push(new OptParameter(null, dbtype_1.DBType.NULL, dbtype_1.DBType.NULL_PREC));
                break;

              case lval_1.Type.INT:
                a += "?";
                try {
                    u = BigInt(s.value);
                } catch (r) {
                    throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR(r + "");
                }
                u <= const_1.Const.INT32_MAX && u >= const_1.Const.INT32_MIN ? t.push(new OptParameter(bufferUtil_1.BufferUtil.h(Number(u)), dbtype_1.DBType.INT, dbtype_1.DBType.INT_PREC)) : t.push(new OptParameter(bufferUtil_1.BufferUtil.l(u), dbtype_1.DBType.BIGINT, dbtype_1.DBType.BIGINT_PREC));
                break;

              case lval_1.Type.DOUBLE:
                a += "?";
                try {
                    t.push(new OptParameter(numeric_1.Numeric.fromString(s.value, 0, 0).encode(!0), dbtype_1.DBType.DECIMAL, 0));
                } catch (r) {
                    if (u = parseFloat(s.value), Number.isNaN(u)) throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
                    t.push(new OptParameter(bufferUtil_1.BufferUtil.v(u), dbtype_1.DBType.DOUBLE, dbtype_1.DBType.DOUBLE_PREC));
                }
                break;

              case lval_1.Type.DECIMAL:
                a += "?", t.push(new OptParameter(numeric_1.Numeric.fromString(s.value, 0, 0).encode(!1), dbtype_1.DBType.DECIMAL, 0));
                break;

              case lval_1.Type.STRING:
                if (s.value.length > const_1.Const.INT16_MAX) {
                    a += "'" + stringUtil_1.StringUtil.processSingleQuoteOfName(s.value) + "'";
                    break;
                }
                a += "?", i && (s.value = stringUtil_1.StringUtil.translate(s.value)), t.push(new OptParameter(bufferUtil_1.BufferUtil.o(s.value, e), dbtype_1.DBType.VARCHAR, dbtype_1.DBType.VARCHAR_PREC));
                break;

              case lval_1.Type.HEX_INT:
              default:
                a += s.value;
            }
            0;
        }
        return a;
    }, p.arrayConstant = function(r, t) {
        var e = "", i = this.lex(r);
        if (!i || 0 === i.length) return [ r, t ];
        var a = i[0].value;
        if (!(stringUtil_1.StringUtil.equalsIgnoreCase(a, "insert") || stringUtil_1.StringUtil.equalsIgnoreCase(a, "select") || stringUtil_1.StringUtil.equalsIgnoreCase(a, "update") || stringUtil_1.StringUtil.equalsIgnoreCase(a, "delete"))) return [ r, t ];
        var l, n = new Map();
        if (Array.isArray(t)) {
            l = new Array();
            for (var _ = 0; _ < t.length; _++) {
                var s = dm_1.dmdb.isBindParameter(t[_]) ? t[_].val : t[_];
                Array.isArray(s) ? n.set(_, stringUtil_1.StringUtil.array2String(s)) : l.push(t[_]);
            }
        } else for (var u in l = {}, t) {
            s = dm_1.dmdb.isBindParameter(t[u]) ? t[u].val : t[u];
            Array.isArray(s) ? n.set(u, stringUtil_1.StringUtil.array2String(s)) : l[u] = t[u];
        }
        if (0 === n.size) return [ r, t ];
        for (var o = -1, _ = 0; _ < i.length; _++) {
            var f, v = i[_];
            switch (v.type) {
              case lval_1.Type.NORMAL:
                "?" === v.value ? (o++, n.has(o) ? e += n.get(o) : e += v.value) : ":" === v.value && (f = this.next(i, _ + 1)) ? (o++, 
                u = Array.isArray(t) ? o : f.value, n.has(u) ? (e += n.get(u), _ = f.position) : e += v.value) : e += v.value;
                break;

              default:
                e += v.value;
            }
        }
        return [ e, l ];
    }, p.next = function(r, t) {
        for (var e = null, i = t; i < r.length && (e = r[i]).type === lval_1.Type.WHITESPACE_OR_COMMENT; i++) ;
        return e;
    }, p;
}();

exports.SQLProcessor = SQLProcessor;

var OptParameter = function() {
    function r(r, t, e) {
        this.ioType = 0, this.scale = 0, this.bytes = r, this.type = t, this.prec = e;
    }
    return r.prototype.toString = function() {
        return this.bytes ? this.bytes.toString() : "null";
    }, r;
}();

exports.OptParameter = OptParameter;
