// dmdb/src/middleware/concurrent/concurrentMiddle.js
"use strict";

var __awaiter = this && this.__awaiter || function(r, o, c, a) {
    return new (c = c || Promise)(function(n, t) {
        function e(r) {
            try {
                u(a.next(r));
            } catch (r) {
                t(r);
            }
        }
        function i(r) {
            try {
                u(a.throw(r));
            } catch (r) {
                t(r);
            }
        }
        function u(r) {
            var t;
            r.done ? n(r.value) : ((t = r.value) instanceof c ? t : new c(function(r) {
                r(t);
            })).then(e, i);
        }
        u((a = a.apply(r, o || [])).next());
    });
}, __generator = this && this.__generator || function(n, e) {
    var i, u, o, c = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: t(0),
        throw: t(1),
        return: t(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function t(t) {
        return function(r) {
            return function(t) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (i = 1, u && (o = 2 & t[0] ? u.return : t[0] ? u.throw || ((o = u.return) && o.call(u), 
                    0) : u.next) && !(o = o.call(u, t[1])).done) return o;
                    switch (u = 0, o && (t = [ 2 & t[0], o.value ]), t[0]) {
                      case 0:
                      case 1:
                        o = t;
                        break;

                      case 4:
                        return c.label++, {
                            value: t[1],
                            done: !1
                        };

                      case 5:
                        c.label++, u = t[1], t = [ 0 ];
                        continue;

                      case 7:
                        t = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = c.trys).length && o[o.length - 1]) && (6 === t[0] || 2 === t[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === t[0] && (!o || t[1] > o[0] && t[1] < o[3])) {
                            c.label = t[1];
                            break;
                        }
                        if (6 === t[0] && c.label < o[1]) {
                            c.label = o[1], o = t;
                            break;
                        }
                        if (o && c.label < o[2]) {
                            c.label = o[2], c.ops.push(t);
                            break;
                        }
                        o[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    t = e.call(n, c);
                } catch (r) {
                    t = [ 6, r ], u = 0;
                } finally {
                    i = o = 0;
                }
                if (5 & t[0]) throw t[1];
                return {
                    value: t[0] ? t[1] : void 0,
                    done: !0
                };
            }([ t, r ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var error_1 = require("../../driver/error"), ConcurrentMiddleware = function() {
    function r() {}
    return r.prototype.connection_openConnection = function(t, n, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_openConnection(n, e) ];
            });
        });
    }, r.prototype.connection_execute = function(t, n, e, i, u) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_execute(n, e, i, u) ];
            });
        });
    }, r.prototype.connection_executeMany = function(t, n, e, i, u) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_executeMany(n, e, i, u) ];
            });
        });
    }, r.prototype.connection_createLob = function(t, n, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_createLob(n, e) ];
            });
        });
    }, r.prototype.connection_close = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_close(n) ];
            });
        });
    }, r.prototype.connection_commit = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_commit(n) ];
            });
        });
    }, r.prototype.connection_getStatementInfo = function(t, n, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_getStatementInfo(n, e) ];
            });
        });
    }, r.prototype.connection_rollback = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_rollback(n) ];
            });
        });
    }, r.prototype.connection_free = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.connection_free(n) ];
            });
        });
    }, r.prototype.lob_close = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("Lob");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.lob_close(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.lob_getData = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("Lob");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.lob_getData(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.lob_getLength = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("Lob");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.lob_getLength(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.pool_getConnection = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.pool_getConnection(n) ];
            });
        });
    }, r.prototype.pool_close = function(t, n, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                return [ 2, t.pool_close(n, e) ];
            });
        });
    }, r.prototype.resultSet_close = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("ResultSet");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.resultSet_close(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.resultSet_getRow = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("ResultSet");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.resultSet_getRow(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.resultSet_getRows = function(t, n, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("ResultSet");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.resultSet_getRows(n, e) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.resultSet_getRowCount = function(t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (n._locked) throw error_1.DBError.ECJS_IS_BUSY("ResultSet");
                    n._locked = !0, r.label = 1;

                  case 1:
                    return r.trys.push([ 1, , 3, 4 ]), [ 4, t.resultSet_getRowCount(n) ];

                  case 2:
                    return [ 2, r.sent() ];

                  case 3:
                    return n._locked = !1, [ 7 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r;
}();

exports.ConcurrentMiddleware = ConcurrentMiddleware;
