// dmdb/src/net/msg/execDirect.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, r) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var e in r) r.hasOwnProperty(e) && (t[e] = r[e]);
        })(t, r);
    };
    return function(t, r) {
        function e() {
            this.constructor = t;
        }
        i(t, r), t.prototype = null === r ? Object.create(r) : (e.prototype = r.prototype, 
        new e());
    };
}(), __values = this && this.__values || function(t) {
    var r = "function" == typeof Symbol && Symbol.iterator, e = r && t[r], i = 0;
    if (e) return e.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && i >= t.length && (t = void 0), {
                value: t && t[i++],
                done: !t
            };
        }
    };
    throw new TypeError(r ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var execute_1 = require("./execute"), msg_1 = require("./msg"), error_1 = require("../../driver/error"), const_1 = require("../../desc/const"), dbtype_1 = require("../../driver/dbtype"), ExecDirect = function(o) {
    function t(t, r, e, i) {
        var n = o.call(this, t, msg_1.Msg.lr, r, i) || this;
        return n.vr = e, n;
    }
    return __extends(t, o), t.prototype.te = function() {
        var r, t, e, i, n = this.vr.length;
        this.yr(n, 1), this.St.Zn(this.dr.nativeSql, this.dr.connection.getServerEncoding());
        try {
            for (var o = __values(this.vr), s = o.next(); !s.done; s = o.next()) {
                var u = s.value;
                this.St.Tn(u.ioType), this.St.k(u.type), this.St.k(u.prec), this.St.k(u.scale);
            }
        } catch (t) {
            r = {
                error: t
            };
        } finally {
            try {
                s && !s.done && (t = o.return) && t.call(o);
            } finally {
                if (r) throw r.error;
            }
        }
        try {
            for (var c = __values(this.vr), f = c.next(); !f.done; f = c.next()) {
                if (null === (u = f.value).bytes) this.St.zn(const_1.Const.DATA_NULL_LEN); else {
                    if (u.bytes.length > const_1.Const.MAX_PARAM_DATA_LEN) {
                        if (this.br.pt.msgVersion >= msg_1.Msg.VERSION_PARAMETER_DATA_LEN_4 && u.bytes.length < 4294967295 && dbtype_1.DBType.isComplexType(u.type, u.scale)) {
                            this.St.x(const_1.Const.DATA_FULL_LEN), this.St.Qn(u.bytes);
                            continue;
                        }
                        throw error_1.DBError.ECJS_DATA_TOO_LONG();
                    }
                    this.St.Vn(u.bytes);
                }
            }
        } catch (t) {
            e = {
                error: t
            };
        } finally {
            try {
                f && !f.done && (i = c.return) && i.call(c);
            } finally {
                if (e) throw e.error;
            }
        }
    }, t.prototype.ee = function() {
        return o.prototype.ee.call(this);
    }, t;
}(execute_1.Execute);

exports.ExecDirect = ExecDirect;
