// dmdb/src/net/msg/execute.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, s) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, s) {
            t.__proto__ = s;
        } || function(t, s) {
            for (var e in s) s.hasOwnProperty(e) && (t[e] = s[e]);
        })(t, s);
    };
    return function(t, s) {
        function e() {
            this.constructor = t;
        }
        i(t, s), t.prototype = null === s ? Object.create(s) : (e.prototype = s.prototype, 
        new e());
    };
}(), __values = this && this.__values || function(t) {
    var s = "function" == typeof Symbol && Symbol.iterator, e = s && t[s], i = 0;
    if (e) return e.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && i >= t.length && (t = void 0), {
                value: t && t[i++],
                done: !t
            };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), executeRetInfo_1 = require("../../desc/executeRetInfo"), statement_1 = require("../../driver/statement"), parameter_1 = require("../../desc/parameter"), const_1 = require("../../desc/const"), dbtype_1 = require("../../driver/dbtype"), typeDescriptor_1 = require("../../desc/typeDescriptor"), error_1 = require("../../driver/error"), bufferUtil_1 = require("../../utils/bufferUtil"), column_1 = require("../../desc/column"), dm_1 = require("../../dm"), svcConf_1 = require("../../desc/svcConf"), rowid_1 = require("../../driver/rowid"), data_1 = require("../../convert/data"), Execute = function(n) {
    function u(t, s, e, i) {
        var r = this;
        return "number" == typeof s && e instanceof statement_1.Statement ? r = n.call(this, t, s, e, i) || this : s instanceof statement_1.PreparedStatement && e instanceof Array && ((r = n.call(this, t, t.pt.execute2 ? msg_1.Msg.ss : msg_1.Msg.es, s, i) || this).ts = s.bindParams, 
        r.rs = e), r;
    }
    return __extends(u, n), u.prototype.te = function() {
        var s, t, e = void 0 === this.ts ? 0 : this.ts.length;
        if (this.yr(e, void 0 === this.rs ? 0 : this.rs.length), 0 < e && (this.ns(this.ts), 
        void 0 !== this.rs && 0 < this.rs.length)) try {
            for (var i = __values(this.rs), r = i.next(); !r.done; r = i.next()) {
                var n = r.value;
                this.os(this.ts, n);
            }
        } catch (t) {
            s = {
                error: t
            };
        } finally {
            try {
                r && !r.done && (t = i.return) && t.call(i);
            } finally {
                if (s) throw s.error;
            }
        }
    }, u.prototype.ee = function() {
        var t = new executeRetInfo_1.ExecuteRetInfo(), s = this.dr, e = s.connection, i = msg_1.Msg.cs;
        t.retSqlType = this.St.c(i), i += const_1.Const.USINT_SIZE;
        var r = this.St.c(i);
        i += const_1.Const.USINT_SIZE, t.updateCount = this.St.i(i), i += const_1.Const.DDWORD_SIZE;
        var n = this.St.U(i);
        i += const_1.Const.USINT_SIZE, t.rsUpdatable = 0 != this.St.a(i), i += const_1.Const.BYTE_SIZE;
        var o = this.St.s(i);
        i += const_1.Const.ULINT_SIZE, t.printLen = this.St.s(i), i += const_1.Const.ULINT_SIZE;
        var c = -1;
        t.retSqlType === const_1.Const.SQL_TYPE_DML_SELECT || t.retSqlType === const_1.Const.SQL_TYPE_DML_CALL ? (t.rsBdta = this.St.a(i) === const_1.Const.RS_BDTA, 
        i += const_1.Const.BYTE_SIZE, c = this.St.c(i), i += const_1.Const.USINT_SIZE, i += 5) : (e.msgVersion < msg_1.Msg.VERSION_ROWID_B12 && (t.retSqlType === const_1.Const.SQL_TYPE_DML_INSERT || t.retSqlType === const_1.Const.SQL_TYPE_DML_UPDATE || t.retSqlType === const_1.Const.SQL_TYPE_DML_DELETE) && (t.rowid = new rowid_1.RowId(this.St.m(i, const_1.Const.DDWORD_SIZE))), 
        i += const_1.Const.DDWORD_SIZE), t.execId = this.St.s(i), i += const_1.Const.ULINT_SIZE, 
        t.rsCacheOffset = this.St.s(i), i += const_1.Const.ULINT_SIZE;
        var _ = this.St.a(i);
        i += const_1.Const.BYTE_SIZE;
        var h, a, f = 1 == (1 & _);
        switch (e.trxStatus = this.St.s(i), e.setTrxFinish(e.trxStatus), i += const_1.Const.ULINT_SIZE, 
        0 < t.printLen && (h = this.St.cn(t.printLen), t.printMsg = bufferUtil_1.BufferUtil.u(h, 0, h.length, e.getServerEncoding())), 
        e.msgVersion >= msg_1.Msg.VERSION_ROWID_B12 && (t.retSqlType === const_1.Const.SQL_TYPE_DML_INSERT || t.retSqlType === const_1.Const.SQL_TYPE_DML_UPDATE || t.retSqlType === const_1.Const.SQL_TYPE_DML_CURSOR_UPDATE) && (t.rowid = new rowid_1.RowId(this.St.cn(const_1.Const.ROWID_SIZE))), 
        0 < n && (t.outParamDatas = this._s(n, this.rs ? this.rs.length : 1)), t.retSqlType) {
          case const_1.Const.SQL_TYPE_SET_TIME_ZONE:
            e.conn_prop_localTimezone = this.St.bn();
            break;

          case const_1.Const.SQL_TYPE_DML_SELECT:
            t.hasResultSet = !0, 0 < r && (s.columns = this.hs(r, t.rsBdta)), this.as(t, s.columns.length, o, c);
            break;

          case const_1.Const.SQL_TYPE_DML_CALL:
            (0 < r || 0 < o) && (t.hasResultSet = !0), 0 < r && (s.columns = this.hs(r, t.rsBdta)), 
            this.as(t, s.columns.length, o, c);
            break;

          case const_1.Const.SQL_TYPE_SET_SESS_TRAN:
            e.isoLevel = this.St.bn(), e.readOnly = 1 == this.St.on();
            break;

          case const_1.Const.SQL_TYPE_SET_CURSCH:
            e.schema = this.St.qn(e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_EXPLAIN:
            t.explain = this.St.qn(e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DML_INSERT:
          case const_1.Const.SQL_TYPE_DML_UPDATE:
          case const_1.Const.SQL_TYPE_DML_DELETE:
            if (f) {
                for (var u = this.St.D(), d = new Array(u), p = 0; p < u; p++) d[p] = Number(this.St.gn());
                t.updateCounts = d;
            } else void 0 !== this.rs && 1 === this.rs.length && (t.updateCounts = [ Number(t.updateCount) ]);
            this.decodeGeneratedKeys(t), this.fs == error_1.DBError.EC_BP_WITH_ERROR().errCode && (t.batchErrors = this.us(t.updateCounts));
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_DATEFMT:
            a = this.St.ln(), e.formatDate = this.St.Sn(a, e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_DTFMT:
            a = this.St.ln(), e.formatTimestamp = this.St.Sn(a, e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_DTTZFMT:
            a = this.St.ln(), e.formatTimestampTZ = this.St.Sn(a, e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_TIMEFMT:
            a = this.St.ln(), e.formatTime = this.St.Sn(a, e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_TIMETZFMT:
            a = this.St.ln(), e.formatTimeTZ = this.St.Sn(a, e.getServerEncoding());
            break;

          case const_1.Const.SQL_TYPE_DDL_ALTSESS_DATE_LANGUAGE:
            e.oracleDateLanguage = this.St.ln();
            break;

          case const_1.Const.SQL_TYPE_NLS_NUMERIC_CHARACTERS:
            e.formatNLS = this.St.Sn(2, e.getServerEncoding());
        }
        return t;
    }, u.prototype.yr = function(t, s) {
        var e, i = this.br.pt, r = this.dr, n = msg_1.Msg.cs;
        if (r.execOpts && (e = r.execOpts.autoCommit), void 0 === e && (e = dm_1.dmdb.autoCommit), 
        void 0 === e && (e = i.conn_prop_autoCommit), n += this.St.t(n, e ? 1 : 0), t > parameter_1.Parameter.PARAM_COUNT_LIMIT) throw error_1.DBError.ECJS_PARAM_COUNT_LIMIT();
        n += this.St.e(n, t), n += this.St.t(n, 1), n += this.St.r(n, BigInt(s)), n += this.St.r(n, BigInt(r.cursorUpdateRow)), 
        n += this.St.r(n, r.maxRows <= 0 || r.connection.conn_prop_enRsCache ? BigInt(const_1.Const.INT64_MAX) : BigInt(r.maxRows)), 
        n += this.St.t(n, 0);
        var o = i.conn_prop_continueBatchOnError ? 1 : 0;
        try {
            r.execOpts.batchErrors && (o = 1);
        } catch (t) {}
        r.multiRowsError && (o |= 128), n += this.St.$n(n, o), n += this.St.t(n, 0), n += this.St.t(n, 0), 
        n += this.St.f(n, 0 === r.queryTimeout ? -1 : r.queryTimeout), n += this.St.f(n, r.connection.conn_prop_batchAllowMaxErrors), 
        this.St.t(n, r.innerExec ? 1 : 0);
    }, u.prototype.ns = function(t) {
        var r = this;
        t.forEach(function(t) {
            t.mask == const_1.Const.MASK_ORACLE_DATE ? t.scale = t.scale | dbtype_1.DBType.ORACLE_DATE_SCALE_MASK : t.mask == const_1.Const.MASK_LOCAL_DATETIME ? t.scale = t.scale | dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK : t.mask == const_1.Const.MASK_ORACLE_FLOAT && (t.prec = t.originalPrec, 
            t.scale = dbtype_1.DBType.ORACLE_FLOAT_SCALE_MASK), t.type === dbtype_1.DBType.CURSOR && t.ioType === parameter_1.Parameter.IO_TYPE_OUT ? r.St.Tn(parameter_1.Parameter.IO_TYPE_INOUT) : r.St.Tn(t.ioType), 
            r.St.k(t.type);
            var s = t.prec, e = t.scale, i = t.typeDescriptor;
            switch (t.type) {
              case dbtype_1.DBType.ARRAY:
              case dbtype_1.DBType.SARRAY:
                s = typeDescriptor_1.TypeDescriptor.getPackArraySize(i);
                break;

              case dbtype_1.DBType.PLTYPE_RECORD:
                s = typeDescriptor_1.TypeDescriptor.getPackRecordSize(i);
                break;

              case dbtype_1.DBType.CLASS:
                s = typeDescriptor_1.TypeDescriptor.getPackClassSize(i);
                break;

              case dbtype_1.DBType.BLOB:
                dbtype_1.DBType.isComplexType(t.type, t.scale) && 4 == (s = i.getObjId()) && (s = i.getOuterId());
            }
            switch (r.St.k(s), r.St.k(e), t.type) {
              case dbtype_1.DBType.ARRAY:
              case dbtype_1.DBType.SARRAY:
                typeDescriptor_1.TypeDescriptor.packArray(i, r.St);
                break;

              case dbtype_1.DBType.PLTYPE_RECORD:
                typeDescriptor_1.TypeDescriptor.packRecord(i, r.St);
                break;

              case dbtype_1.DBType.CLASS:
                typeDescriptor_1.TypeDescriptor.packClass(i, r.St);
            }
        });
    }, u.prototype.os = function(t, s) {
        for (var e = 0; e < t.length; e++) {
            var i = t[e], r = s[e];
            if (i.type !== dbtype_1.DBType.CURSOR) {
                if (i.ioType !== parameter_1.Parameter.IO_TYPE_OUT) if (null === r) this.St.zn(const_1.Const.DATA_NULL_LEN); else if (r instanceof Buffer) {
                    if (r.length > const_1.Const.MAX_PARAM_DATA_LEN) {
                        if (this.br.pt.msgVersion >= msg_1.Msg.VERSION_PARAMETER_DATA_LEN_4 && r.length < 4294967295 && dbtype_1.DBType.isComplexType(i.type, i.scale)) {
                            this.St.x(const_1.Const.DATA_FULL_LEN), this.St.Qn(r);
                            continue;
                        }
                        throw error_1.DBError.ECJS_DATA_TOO_LONG();
                    }
                    this.St.Vn(r);
                } else {
                    if (!(r instanceof data_1.Data)) throw new Error("Bind param data failed by invalid param data type");
                    this.St.zn(r.len), this.St.It(r.value);
                }
            } else this.St.x(const_1.Const.ULINT_SIZE), this.St.k(i.cursorStmt.handle);
        }
    }, u.prototype._s = function(t, s) {
        for (var e = new Array(t), i = 0, r = this.ts, n = 0; n < r.length; n++) r[n].ioType !== parameter_1.Parameter.IO_TYPE_INOUT && r[n].ioType !== parameter_1.Parameter.IO_TYPE_OUT || (e[i] = n, 
        i++);
        var o, c = new Array();
        if (this.br.pt.msgVersion >= msg_1.Msg.VERSION_PARAM_BY_COLUMN) for (n = 0; n < t; n++) for (var _ = this.getRowNum(this.St.cn(const_1.Const.ULINT_SIZE)), h = 0; h < _; h++) h < c.length ? o = c[h] : (o = new Array(r.length), 
        c.push(o)), o = this.fillOutParam(o, e[n]); else for (h = 0; h < s; h++) {
            o = new Array(r.length);
            for (n = 0; n < t; n++) o = this.fillOutParam(o, e[n]);
            c.push(o);
        }
        return c;
    }, u.prototype.decodeGeneratedKeys = function(t) {
        if (2 == (2 & this.St.a(u.MSG_RSP_EXECITE_RET_ID_FLAG))) if (this.br.pt.msgVersion < msg_1.Msg.VERSION_GENERATE_KEY) for (var s = this.St.D(), e = 0; e < s; e++) this.St.cn(const_1.Const.DDWORD_SIZE); else for (var i = this.St.vn(), e = 0; e < i; e++) {
            var r = this.St.T();
            r === const_1.Const.REC4_SQL_NULL || r === const_1.Const.REC4_SQL_NULL2 || (r === const_1.Const.REC4_SQL_FULL ? this.St.On() : this.St.cn(r));
        } else t.updateCount, BigInt(1);
    }, u.prototype.getRowNum = function(t) {
        return bufferUtil_1.BufferUtil.s(t, 0) << 1 >>> 1;
    }, u.prototype.fillOutParam = function(t, s) {
        var e = !1, i = this.St.T();
        return i === const_1.Const.DATA_NULL_LEN || i === const_1.Const.DATA_NULL_LEN2 ? e = !(i = 0) : i === const_1.Const.DATA_FULL_LEN && (i = this.St.D()), 
        t[s] = e ? null : this.St.cn(i), t;
    }, u.prototype.hs = function(t, s) {
        for (var e, i, r, n, o = this.br.pt.getServerEncoding(), c = new Array(t), _ = 0; _ < t; _++) {
            var h = new column_1.Column();
            h.type = this.St.D(), h.prec = this.St.D(), h.scale = this.St.D(), h.nullable = 0 != this.St.D();
            var a = this.St.bn();
            h.lob = 0 != (a & u.ds), h.identity = 0 != (a & u.ps), h.readonly = 0 != (a & u.ys), 
            this.St.rn(4, !1, !0), this.St.rn(2, !1, !0), e = this.St.bn(), i = this.St.bn(), 
            r = this.St.bn(), n = this.St.bn(), h.name = this.St.Sn(e, o), h.typeName = this.St.Sn(i, o), 
            h.tableName = this.St.Sn(r, o), h.schemaName = this.St.Sn(n, o), this.dr.readBaseColName && (h.baseName = this.St.S(o)), 
            h.lob && (h.lobTabId = this.St.D(), h.lobColId = this.St.bn()), h.type === dbtype_1.DBType.DATETIME || h.type == dbtype_1.DBType.DATETIME2 ? 0 != (h.scale & dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK) ? (h.scale = h.scale & ~dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK, 
            h.mask = const_1.Const.MASK_LOCAL_DATETIME) : 0 != (h.scale & dbtype_1.DBType.ORACLE_DATE_SCALE_MASK) && (h.scale = h.scale & ~dbtype_1.DBType.ORACLE_DATE_SCALE_MASK, 
            h.mask = const_1.Const.MASK_ORACLE_DATE) : h.type === dbtype_1.DBType.DECIMAL && h.scale === dbtype_1.DBType.ORACLE_FLOAT_SCALE_MASK ? (h.originalPrec = h.prec, 
            h.prec = Math.round(.30103 * h.prec) + 1, h.scale = -1, h.mask = const_1.Const.MASK_ORACLE_FLOAT) : h.type === dbtype_1.DBType.VARCHAR && h.prec === dbtype_1.DBType.BFILE_PREC && h.scale === dbtype_1.DBType.BFILE_SCALE && (h.mask = const_1.Const.MASK_BFILE), 
            c[_] = h;
        }
        for (var f, _ = 0; _ < t; _++) {
            dbtype_1.DBType.isComplexType(c[_].type, c[_].scale) && ((f = new typeDescriptor_1.TypeDescriptor(this.br.pt)).unpack(this.St), 
            c[_].typeDescriptor = f);
        }
        return c;
    }, u.prototype.as = function(t, s, e, i) {
        if (0 < e) {
            var r = this.St.nn;
            if (!t.rsBdta) {
                for (var n = new Array(e), o = 0; o < e; o++) {
                    n[o] = new Array(s + 1), this.St.rn(2, !1, !0), n[o][0] = this.St.cn(this.br.pt.msgVersion < msg_1.Msg.VERSION_ROWID_B12 ? const_1.Const.LINT64_SIZE : const_1.Const.ROWID_SIZE), 
                    this.St.rn(2 * s, !1, !0);
                    for (var c = 1; c < s + 1; ++c) {
                        var _ = this.St.T();
                        _ == const_1.Const.REC4_SQL_NULL || _ === const_1.Const.REC4_SQL_NULL2 ? n[o][c] = null : _ != const_1.Const.REC4_SQL_FULL ? n[o][c] = this.St.cn(_) : n[o][c] = this.St.On();
                    }
                }
                t.rsDatas = n;
            }
            t.rsSizeof = this.St.nn - r;
        }
        if (0 < t.rsCacheOffset) {
            for (var h = this.St.bn(), a = new Array(h), f = new Array(h), o = 0; o < h; o++) a[o] = BigInt(this.St.D()), 
            f[o] = this.St.gn();
            t.tbIds = a, t.tbTss = f;
        }
    }, u.prototype.us = function(t) {
        this.St.rn(4, !1, !0);
        var s = this.St.D(), e = this.br.pt.getServerEncoding();
        svcConf_1.SvcConf.LOCALE !== const_1.Const.LANGUAGE_EN && e === const_1.Const.ENCODING_EUCKR && (e = const_1.Const.ENCODING_GB18030), 
        svcConf_1.SvcConf.LOCALE === const_1.Const.LANGUAGE_CNT_HK && (e = e === const_1.Const.ENCODING_UTF8 ? e : const_1.Const.ENCODING_BIG5);
        for (var i = new Array(), r = 0; r < s; r++) {
            var n = this.St.D();
            t[n] = dm_1.dmdb.EXEC_MANY_FAIL;
            var o = this.St.D(), c = this.St.S(e);
            i.push("row[" + n + "]:" + o + ", " + c);
        }
        return i;
    }, u.ps = 1, u.ds = 2, u.ys = 4, u.bs = 8, u.vs = 256, u.MSG_REQ_EXECUTE_RET_ID_FLAG = 48, 
    u.MSG_RSP_EXECITE_RET_ID_FLAG = 59, u;
}(msg_1.Msg);

exports.Execute = Execute;
