// dmdb/src/net/msg/login.js
"use strict";

var __extends = this && this.__extends || function() {
    var h = function(t, i) {
        return (h = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, i) {
            t.__proto__ = i;
        } || function(t, i) {
            for (var s in i) i.hasOwnProperty(s) && (t[s] = i[s]);
        })(t, i);
    };
    return function(t, i) {
        function s() {
            this.constructor = t;
        }
        h(t, i), t.prototype = null === i ? Object.create(i) : (s.prototype = i.prototype, 
        new s());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), const_1 = require("../../desc/const"), connection_1 = require("../../driver/connection"), bufferUtil_1 = require("../../utils/bufferUtil"), error_1 = require("../../driver/error"), rsaCipher_1 = require("../../security/rsaCipher"), driverUtil_1 = require("../../utils/driverUtil"), driver_1 = require("../../driver"), Login = function(h) {
    function _(t, i) {
        var s = h.call(this, t, msg_1.Msg.Mi, null, i) || this;
        return s.Hi = !1, s.pt = t.pt, s;
    }
    return __extends(_, h), _.processServerVersion = function(t, i) {
        if (driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_LOWEST_DATABASE) < 0) throw error_1.DBError.ECJS_ERR_SERVER_VERSION();
        i.malini2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_MALINI2), 
        i.execute2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_EXECUTE2), 
        i.lobEmptyCompOrcl = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_LOB_EMPTY_COMP_ORCL), 
        i.lobOptimized = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_SERVER_LOB_OPTIMIZED);
    }, _.prototype.te = function() {
        var t = this.pt.getServerEncoding();
        this.St.f(_.qi, this.pt.newClientType ? const_1.Const.SESS3_CLT_NODE : const_1.Const.SESS3_CLT_JDBC), 
        this.St.f(_.Fi, connection_1.Connection.node2dbIsoLevel(this.pt.isoLevel)), this.St.f(_.Si, this.pt.conn_prop_locale), 
        this.St.B(_.Bi, this.pt.conn_prop_localTimezone), this.St.t(_.wi, this.pt.readOnly ? const_1.Const.TRX_ACCESS_MODE_READ_ONLY : const_1.Const.TRX_ACCESS_MODE_READ_WRITE), 
        this.St.f(_.Di, this.pt.conn_prop_sessionTimeout), this.St.t(_.Yi, this.pt.conn_prop_mppLocal ? 1 : 0), 
        this.St.t(_.Ai, this.pt.conn_prop_rwSeparate === const_1.Const.RW_SEPARATE_DB_APPLY_WAIT ? const_1.Const.RW_SEPARATE_DB_APPLY_WAIT : 0 < this.pt.conn_prop_rwSeparate ? 1 : 0), 
        this.St.t(_.Ti, this.pt.newLobFlag ? 1 : 0), this.St.t(_.Zi, this.pt.conn_prop_osAuthType), 
        this.St.t(_.Ci, this.pt.longLobFlag ? 1 : 0), this.St.t(_.Ni, 1), this.br.lt;
        var i, s = bufferUtil_1.BufferUtil.o(this.pt.conn_prop_user, t), h = bufferUtil_1.BufferUtil.o(this.pt.conn_prop_pwd, t);
        if (s.length > const_1.Const.MAX_NAME_LEN) throw error_1.DBError.ECJS_USERNAME_TOO_LONG();
        if (h.length > const_1.Const.MAX_NAME_LEN) throw error_1.DBError.ECJS_PASSWORD_TOO_LONG();
        this.br.wt && this.pt.conn_prop_loginCertificate ? (s = (i = new rsaCipher_1.RSACipher(this.pt.conn_prop_loginCertificate)).encrypt(s), 
        h = i.encrypt(h)) : this.br.wt && (s = this.br.Ut.encrypt(s, !1), h = this.br.Ut.encrypt(h, !1)), 
        this.St.Qn(s), this.St.Qn(h), this.St.Xn(this.pt.conn_prop_appName, t), this.St.Xn(this.pt.conn_prop_osName, t), 
        this.St.Xn(const_1.Const.HOST_NAME, t), this.St.Tn(this.pt.rwStandby ? 1 : 0);
    }, _.prototype.ee = function() {
        var t = this.pt.getServerEncoding();
        this.yi(this.pt), this.St.qn(t);
        var i, s, h, r, n, o = this.St.D();
        if (0 === o && 0 < this.pt.msgVersion ? this.pt.schema = this.pt.conn_prop_user.toUpperCase() : this.pt.schema = this.St.Sn(o, t), 
        this.St.qn(t), this.St.qn(t), this.St.D(), this.St.D(), this.St.D(), this.pt.guid = this.St.qn(t), 
        this.St.qn(t), 1 === this.St.a(_.ki) && (this.pt.standbyHost = this.St.qn(t), this.pt.standbyPort = this.St.D(), 
        this.pt.standbyCount = this.St.T()), 0 < this.St.in() && (this.pt.sessionID = this.St.gn()), 
        0 < this.St.in() && 1 == this.St.on() && (this.pt.formatDate = "DD-MON-YY", this.pt.formatTime = "HH12.MI.SS.FF6 AM", 
        this.pt.formatTimestamp = "DD-MON-YY HH12.MI.SS.FF6 AM", this.pt.formatTimestampTZ = "DD-MON-YY HH12.MI.SS.FF6 AM +TZH:TZM", 
        this.pt.formatTimeTZ = "HH12.MI.SS.FF6 AM +TZH:TZM"), 0 < this.St.in() && ((i = this.St.Dn(t)) && (this.pt.formatDate = i), 
        (i = this.St.Dn(t)) && (this.pt.formatTime = i), (i = this.St.Dn(t)) && (this.pt.formatTimestamp = i), 
        (i = this.St.Dn(t)) && (this.pt.formatTimestampTZ = i), (i = this.St.Dn(t)) && (this.pt.formatTimeTZ = i)), 
        0 < this.St.in() && this.St.Dn(t), 0 < this.St.in() && this.St.rn(1, !1, !0), this.pt.msgVersion >= msg_1.Msg.VERSION_ROWID_B12 && (s = 0 != this.St.vn(), 
        h = this.St.vn(), r = this.St.vn(), this.pt.rowidNBitsEpno = this.utl_rowid_get_n_bits_for_epno_org(h, s, r), 
        this.pt.rowidMaxHpno = this.utl_rowid_get_max_hpno_org(h, s, r), this.pt.rowidMaxEpno = this.utl_rowid_get_max_epno_org(this.pt.rowidNBitsEpno, s, r), 
        this.pt.rowidNBitsReal = this.utl_rowid_get_n_bits_for_real_org(h, s, r)), 0 < this.St.in() && this.St.rn(2, !1, !0), 
        0 < this.St.in()) {
            var e = this.St.bn();
            switch (Math.min(e, _.MAX_INI_COUNT)) {
              case 1:
                this.St.rn(4, !1, !0);
            }
        }
        return 0 < this.St.in() && this.pt.msgVersion >= msg_1.Msg.VERSION_NEW_SERVER_VERSION && (n = this.St.qn(t), 
        _.processServerVersion(n, this.pt)), this.pt;
    }, _.prototype.yi = function(t) {
        t.maxRowSize = this.St.s(_.zi), t.ddlAutoCommit = 1 == this.St.a(_.Ei), t.isoLevel = connection_1.Connection.db2nodeIsoLevel(this.St.s(_.Gi)), 
        t.conn_prop_caseSensitive = 1 == this.St.a(_.Ji), t.backSlashFlag = 1 == this.St.a(_.Ki), 
        t.svrStat = this.St.c(_.Pi), t.svrMode = this.St.c(_.Qi), t.constParaOpt = 1 == this.St.a(_.Ri), 
        t.dbTimezone = this.St.c(_.Vi), t.newLobFlag = 1 == this.St.a(_.Wi), t.longLobFlag = 1 == this.St.a(_.Xi), 
        this.St.a(_.$i), this.St.a(_.ls), this.br.bt = 1e3 * this.St.H(_.gs), 0 === t.conn_prop_bufPrefetch && (t.conn_prop_bufPrefetch = this.St.s(_.Ms)), 
        this.St.c(_.Is), t.dscControl = 1 === this.St.a(_.Us);
    }, _.prototype.utl_rowid_get_n_bits_for_epno_org = function(t, i, s) {
        var h = 1, r = 0;
        if (!i && 0 == s) return r;
        for (;h < t; ) h <<= 1, r++;
        return r;
    }, _.prototype.utl_rowid_get_max_hpno_org = function(t, i, s) {
        var h, r = i || 0 != s ? (h = this.utl_rowid_get_n_bits_for_epno_org(t, i, s), 0 != s ? (BigInt(1) << BigInt(20 - h)) - BigInt(1) : (BigInt(1) << BigInt(16 - h)) - BigInt(1)) : BigInt(65534);
        return r;
    }, _.prototype.utl_rowid_get_max_epno_org = function(t, i, s) {
        var h = i || 0 != s ? (1 << t) - 1 : 0;
        return h;
    }, _.prototype.utl_rowid_get_n_bits_for_real_org = function(t, i, s) {
        return 0 != s ? 44 : 48;
    }, _.qi = msg_1.Msg.cs, _.Fi = _.qi + const_1.Const.ULINT_SIZE, _.Si = _.Fi + const_1.Const.ULINT_SIZE, 
    _.wi = _.Si + const_1.Const.ULINT_SIZE, _.Bi = _.wi + const_1.Const.BYTE_SIZE, _.Di = _.Bi + const_1.Const.USINT_SIZE, 
    _.Yi = _.Di + const_1.Const.ULINT_SIZE, _.Ai = _.Yi + const_1.Const.BYTE_SIZE, _.Ti = _.Ai + const_1.Const.BYTE_SIZE, 
    _.Zi = _.Ti + const_1.Const.BYTE_SIZE, _.Ci = _.Zi + const_1.Const.BYTE_SIZE, _.Ni = _.Ci + const_1.Const.BYTE_SIZE, 
    _.Hs = (_.zi = _.cs) + const_1.Const.ULINT_SIZE, _.Ei = _.Hs + const_1.Const.ULINT_SIZE, 
    _.Gi = _.Ei + const_1.Const.BYTE_SIZE, _.Ji = _.Gi + const_1.Const.ULINT_SIZE, _.Ki = _.Ji + const_1.Const.BYTE_SIZE, 
    _.Qi = _.Ki + const_1.Const.BYTE_SIZE, _.Pi = _.Qi + const_1.Const.USINT_SIZE, _.Ri = _.Pi + const_1.Const.USINT_SIZE, 
    _.Vi = _.Ri + const_1.Const.BYTE_SIZE, _.ki = _.Vi + const_1.Const.USINT_SIZE, _.Wi = _.ki + const_1.Const.BYTE_SIZE, 
    _.Ms = _.Wi + const_1.Const.BYTE_SIZE, _.Is = _.Ms + const_1.Const.ULINT_SIZE, _.Us = _.Is + const_1.Const.USINT_SIZE, 
    _.Xi = _.Us + const_1.Const.BYTE_SIZE, _.$i = _.Xi + const_1.Const.BYTE_SIZE, _.gs = _.$i + const_1.Const.BYTE_SIZE, 
    _.ls = 63, _.MAX_INI_COUNT = 1, _;
}(msg_1.Msg);

exports.Login = Login;
