// dmdb/src/net/msg/prepare.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, e) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, e) {
            t.__proto__ = e;
        } || function(t, e) {
            for (var s in e) e.hasOwnProperty(s) && (t[s] = e[s]);
        })(t, e);
    };
    return function(t, e) {
        function s() {
            this.constructor = t;
        }
        i(t, e), t.prototype = null === e ? Object.create(e) : (s.prototype = e.prototype, 
        new s());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var execute_1 = require("./execute"), msg_1 = require("./msg"), const_1 = require("../../desc/const"), executeRetInfo_1 = require("../../desc/executeRetInfo"), parameter_1 = require("../../desc/parameter"), dbtype_1 = require("../../driver/dbtype"), typeDescriptor_1 = require("../../desc/typeDescriptor"), dm_1 = require("../../dm"), Prepare = function(c) {
    function t(t, e, s, i, r) {
        var h = c.call(this, t, msg_1.Msg.ks, e, r) || this;
        return h.ye = s, h.be = i, h;
    }
    return __extends(t, c), t.prototype.te = function() {
        var t, e = this.br.pt, s = this.dr, i = msg_1.Msg.cs;
        s.execOpts && (t = s.execOpts.autoCommit), void 0 === t && (t = dm_1.dmdb.autoCommit), 
        void 0 === t && (t = e.conn_prop_autoCommit), i += this.St.t(i, t ? 1 : 0), i += this.St.t(i, this.ye ? 1 : 0), 
        i += this.St.t(i, 0), i += this.St.t(i, 1), i += this.St.t(i, e.compatibleOracle() ? 0 : 1), 
        i += this.St.B(i, this.be), i += this.St.r(i, s.maxRows <= 0 || e.conn_prop_enRsCache ? const_1.Const.INT64_MAX : s.maxRows), 
        i += this.St.t(i, e.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        i += this.St.B(i, 0), i += this.St.t(i, 0), i += this.St.t(i, 0), i += this.St.t(i, 0), 
        i += this.St.f(i, s.queryTimeout), this.St.t(i, s.innerExec ? 1 : 0), this.St.Zn(s.nativeSql, e.getServerEncoding());
    }, t.prototype.ee = function() {
        if (this.ye) return c.prototype.ee.call(this);
        var t = new executeRetInfo_1.ExecuteRetInfo(), e = this.br.pt, s = msg_1.Msg.cs;
        t.retSqlType = this.St.c(s), s += const_1.Const.USINT_SIZE;
        var i = this.St.U(s);
        s += const_1.Const.USINT_SIZE;
        var r = this.St.c(s);
        if (s += const_1.Const.USINT_SIZE, this.St.i(s), s += const_1.Const.DDWORD_SIZE, 
        e.trxStatus = this.St.s(s), const_1.Const.ULINT_SIZE, 0 < i) {
            this.dr.serverParams = this.me(i), this.dr.bindParams = new Array(i);
            for (var h = 0; h < i; h++) {
                var n = this.dr.serverParams[h], _ = new parameter_1.Parameter();
                _.type = n.type, _.prec = n.prec, _.scale = n.scale, _.nullable = n.nullable, _.hasDefault = n.hasDefault, 
                _.typeFlag = n.typeFlag, _.lob = n.lob, _.ioType = n.ioType, _.name = n.name, _.typeName = n.typeName, 
                _.tableName = n.tableName, _.schemaName = n.schemaName, _.lobTabId = n.lobTabId, 
                _.lobColId = n.lobColId, _.mask = n.mask, _.typeDescriptor = n.typeDescriptor, this.dr.bindParams[h] = _;
            }
            this.dr.paramCount = i;
        } else this.dr.serverParams = new Array(0), this.dr.bindParams = new Array(0), this.dr.paramCount = 0;
        return this.dr.columns = 0 < r ? this.hs(r, t.rsBdta) : new Array(0), t;
    }, t.prototype.me = function(t) {
        for (var e, s, i, r, h = new Array(t), n = 0; n < t; n++) {
            var _ = new parameter_1.Parameter();
            _.type = this.St.D(), _.prec = this.St.D(), _.scale = this.St.D(), _.nullable = 0 != this.St.D();
            var c = this.St.bn();
            _.hasDefault = 0 != (c & execute_1.Execute.vs), _.typeFlag = 0 != (c & execute_1.Execute.bs) ? parameter_1.Parameter.TYPE_FLAG_RECOMMEND : parameter_1.Parameter.TYPE_FLAG_EXACT, 
            _.lob = 0 != (c & execute_1.Execute.ds), this.St.D(), _.ioType = this.St.bn(), e = this.St.bn(), 
            s = this.St.bn(), i = this.St.bn(), r = this.St.bn(), _.name = this.St.Sn(e, this.br.pt.getServerEncoding()), 
            _.typeName = this.St.Sn(s, this.br.pt.getServerEncoding()), _.tableName = this.St.Sn(i, this.br.pt.getServerEncoding()), 
            _.schemaName = this.St.Sn(r, this.br.pt.getServerEncoding()), _.lob && (_.lobTabId = this.St.D(), 
            _.lobColId = this.St.bn()), _.type === dbtype_1.DBType.DATETIME || _.type === dbtype_1.DBType.DATETIME2 ? 0 != (_.scale & dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK) ? (_.scale = _.scale & ~dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK, 
            _.mask = const_1.Const.MASK_LOCAL_DATETIME) : 0 != (_.scale & dbtype_1.DBType.ORACLE_DATE_SCALE_MASK) && (_.scale = _.scale & ~dbtype_1.DBType.ORACLE_DATE_SCALE_MASK, 
            _.mask = const_1.Const.MASK_ORACLE_DATE) : _.type === dbtype_1.DBType.DECIMAL && _.scale === dbtype_1.DBType.ORACLE_FLOAT_SCALE_MASK ? (_.originalPrec = _.prec, 
            _.prec = Math.round(.30103 * _.prec), _.scale = -1, _.mask = const_1.Const.MASK_ORACLE_FLOAT) : _.type === dbtype_1.DBType.VARCHAR && _.prec === dbtype_1.DBType.BFILE_PREC && _.scale === dbtype_1.DBType.BFILE_SCALE && (_.mask = const_1.Const.MASK_BFILE), 
            h[n] = _;
        }
        for (var o, n = 0; n < t; n++) {
            dbtype_1.DBType.isComplexType(h[n].type, h[n].scale) && ((o = new typeDescriptor_1.TypeDescriptor(this.br.pt)).unpack(this.St), 
            h[n].typeDescriptor = o);
        }
        return h;
    }, t;
}(execute_1.Execute);

exports.Prepare = Prepare;
