// dmdb/src/net/msg/startup.js
"use strict";

var __extends = this && this.__extends || function() {
    var h = function(t, s) {
        return (h = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, s) {
            t.__proto__ = s;
        } || function(t, s) {
            for (var i in s) s.hasOwnProperty(i) && (t[i] = s[i]);
        })(t, s);
    };
    return function(t, s) {
        function i() {
            this.constructor = t;
        }
        h(t, s), t.prototype = null === s ? Object.create(s) : (i.prototype = s.prototype, 
        new i());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), const_1 = require("../../desc/const"), driver_1 = require("../../driver"), login_1 = require("./login"), Startup = function(h) {
    function n(t, s) {
        var i = h.call(this, t, msg_1.Msg._r, null, s) || this;
        return i.pt = t.pt, i;
    }
    return __extends(n, h), n.prototype.te = function() {
        var t, s = this, i = this.pt.getServerEncoding();
        this.nh = this.pt.conn_prop_uKeyName ? 1 : 0, this.St.f(n._h, 0), this.St.f(n.ch, this.pt.conn_prop_compress), 
        this.St.t(n.oh, this.pt.conn_prop_loginEncrypt ? n.GEN_KEYPAIR_SIMPLE : 0), this.St.t(n.rh, this.pt.conn_prop_loginEncrypt ? n.COMM_ENC_SIMPLE : 0), 
        this.St.t(n.eh, this.pt.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        this.St.t(n.uh, this.pt.conn_prop_compressID), this.St.t(n.ah, "" !== this.pt.conn_prop_loginCertificate ? 1 : 0), 
        this.St.t(n.fh, 0), this.St.t(n.gh, 1), this.St.e(n.mh, msg_1.Msg.VERSION), this.St.Xn(driver_1.Driver.VERSION, i), 
        this.St.Tn(0), this.nh, this.pt.conn_prop_loginEncrypt && (t = this.br.ar(), this.St.Qn(t)), 
        "" !== this.pt.conn_prop_catalog && this.St.Xn(this.pt.conn_prop_catalog, i), this.St.zn(n.vh.length), 
        n.vh.forEach(function(t) {
            return s.St.zn(t);
        });
    }, n.prototype.Os = function() {
        this.St.Jt(msg_1.Msg.kt);
    }, n.prototype.ee = function() {
        switch (this.pt.sslEncrypt = this.St.s(n.lh), this.St.s(n.dh), this.St.s(n.bh)) {
          case 1:
            this.pt.serverEncoding = const_1.Const.ENCODING_UTF8;
            break;

          case 2:
            this.pt.serverEncoding = const_1.Const.ENCODING_EUCKR;
            break;

          case 3:
            this.pt.serverEncoding = const_1.Const.ENCODING_GB18030;
        }
        this.pt.conn_prop_compress = this.St.s(n.ph);
        var t, s, i = this.St.a(n.qh), h = this.St.a(n.xh);
        return this.pt.conn_prop_isBdtaRS = 0 < this.St.a(n.jh), this.pt.conn_prop_compressID = this.St.a(n.Oh), 
        this.br.dt = 1 === this.St.a(n.kh), this.pt.newClientType = 1 == this.St.a(n.wh), 
        this.pt.msgVersion = Math.min(msg_1.Msg.VERSION, this.St.U(n.MSG_RES_STARTUP_PROTOCOL_VERSION)), 
        this.Ls(), this.pt.msgVersion < msg_1.Msg.VERSION_NEW_SERVER_VERSION && (t = this.pt.getServerEncoding(), 
        s = this.St.qn(t), login_1.Login.processServerVersion(s, this.pt)), this.nh, this.Yt = -1, 
        this.br.gt = -1, this.br.wt = !1, this.br.mt = !1, 0 < h && (this.br.gt = this.St.D()), 
        i === n.GEN_KEYPAIR_SIMPLE && (-1 === this.br.gt ? this.br.wt = !0 : this.br.mt = !0, 
        this.Wt = this.St.On()), h === n.COMM_ENC_SIMPLE && (this.Yt = this.St.D()), this.pt.msgVersion >= msg_1.Msg.VERSION_DETERMINE_ENCRYPT && (this.St.rn(8, !1, !0), 
        this.pt.algorighm = this.St.T()), this;
    }, n._h = msg_1.Msg.cs, n.ch = n._h + const_1.Const.ULINT_SIZE, n.oh = n.ch + const_1.Const.ULINT_SIZE, 
    n.rh = n.oh + const_1.Const.BYTE_SIZE, n.eh = n.rh + const_1.Const.BYTE_SIZE, n.uh = n.eh + const_1.Const.BYTE_SIZE, 
    n.ah = n.uh + const_1.Const.BYTE_SIZE, n.fh = n.ah + const_1.Const.BYTE_SIZE, n.gh = n.fh + const_1.Const.BYTE_SIZE, 
    n.mh = n.gh + const_1.Const.BYTE_SIZE, n.lh = msg_1.Msg.cs, n.dh = n.lh + const_1.Const.ULINT_SIZE, 
    n.bh = n.dh + const_1.Const.ULINT_SIZE, n.ph = n.bh + const_1.Const.ULINT_SIZE, 
    n.Mh = n.ph + const_1.Const.ULINT_SIZE, n.qh = n.Mh + const_1.Const.ULINT_SIZE, 
    n.xh = n.qh + const_1.Const.BYTE_SIZE, n.jh = n.xh + const_1.Const.BYTE_SIZE, n.Oh = n.jh + const_1.Const.BYTE_SIZE, 
    n.Sh = n.Oh + const_1.Const.BYTE_SIZE, n.kh = n.Sh + const_1.Const.BYTE_SIZE, n.wh = n.kh + const_1.Const.USINT_SIZE, 
    n.MSG_RES_STARTUP_PROTOCOL_VERSION = n.wh + const_1.Const.BYTE_SIZE, n.GEN_KEYPAIR_SIMPLE = 1, 
    n.GEN_KEYPAIR_ZB = 2, n.COMM_ENC_SIMPLE = 2, n.COMM_ENC_ZB = 3, n.vh = [ 1, 0 ], 
    n;
}(msg_1.Msg);

exports.Startup = Startup;
