// dmdb/src/net/msg/tableTs.js
"use strict";

var __extends = this && this.__extends || function() {
    var e = function(t, r) {
        return (e = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var n in r) r.hasOwnProperty(n) && (t[n] = r[n]);
        })(t, r);
    };
    return function(t, r) {
        function n() {
            this.constructor = t;
        }
        e(t, r), t.prototype = null === r ? Object.create(r) : (n.prototype = r.prototype, 
        new n());
    };
}(), __values = this && this.__values || function(t) {
    var r = "function" == typeof Symbol && Symbol.iterator, n = r && t[r], e = 0;
    if (n) return n.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && e >= t.length && (t = void 0), {
                value: t && t[e++],
                done: !t
            };
        }
    };
    throw new TypeError(r ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), TableTS = function(i) {
    function t(t, r, n) {
        var e = i.call(this, t, msg_1.Msg.Cr, null, n) || this;
        return e.pr = r, e;
    }
    return __extends(t, i), t.prototype.te = function() {
        var r, t;
        this.St.e(msg_1.Msg.cs, this.pr.length);
        try {
            for (var n = __values(this.pr), e = n.next(); !e.done; e = n.next()) {
                var i = e.value;
                this.St.Cn(Number(i));
            }
        } catch (t) {
            r = {
                error: t
            };
        } finally {
            try {
                e && !e.done && (t = n.return) && t.call(n);
            } finally {
                if (r) throw r.error;
            }
        }
    }, t.prototype.ee = function() {
        var t = this.St.U(msg_1.Msg.cs);
        if (t <= 0) return [];
        for (var r = new Array(t), n = 0; n < t; n++) r[n] = this.St.gn();
        return r;
    }, t;
}(msg_1.Msg);

exports.TableTS = TableTS;
