// dmdb/src/driver/connection.js
"use strict";

var __extends = this && this.__extends || function() {
    var n = function(t, i) {
        return (n = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, i) {
            t.__proto__ = i;
        } || function(t, i) {
            for (var r in i) i.hasOwnProperty(r) && (t[r] = i[r]);
        })(t, i);
    };
    return function(t, i) {
        function r() {
            this.constructor = t;
        }
        n(t, i), t.prototype = null === i ? Object.create(i) : (r.prototype = i.prototype, 
        new r());
    };
}(), __awaiter = this && this.__awaiter || function(t, o, c, h) {
    return new (c = c || Promise)(function(r, i) {
        function n(t) {
            try {
                s(h.next(t));
            } catch (t) {
                i(t);
            }
        }
        function e(t) {
            try {
                s(h.throw(t));
            } catch (t) {
                i(t);
            }
        }
        function s(t) {
            var i;
            t.done ? r(t.value) : ((i = t.value) instanceof c ? i : new c(function(t) {
                t(i);
            })).then(n, e);
        }
        s((h = h.apply(t, o || [])).next());
    });
}, __generator = this && this.__generator || function(r, n) {
    var e, s, o, c = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: i(0),
        throw: i(1),
        return: i(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function i(i) {
        return function(t) {
            return function(i) {
                if (e) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (e = 1, s && (o = 2 & i[0] ? s.return : i[0] ? s.throw || ((o = s.return) && o.call(s), 
                    0) : s.next) && !(o = o.call(s, i[1])).done) return o;
                    switch (s = 0, o && (i = [ 2 & i[0], o.value ]), i[0]) {
                      case 0:
                      case 1:
                        o = i;
                        break;

                      case 4:
                        return c.label++, {
                            value: i[1],
                            done: !1
                        };

                      case 5:
                        c.label++, s = i[1], i = [ 0 ];
                        continue;

                      case 7:
                        i = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = c.trys).length && o[o.length - 1]) && (6 === i[0] || 2 === i[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === i[0] && (!o || i[1] > o[0] && i[1] < o[3])) {
                            c.label = i[1];
                            break;
                        }
                        if (6 === i[0] && c.label < o[1]) {
                            c.label = o[1], o = i;
                            break;
                        }
                        if (o && c.label < o[2]) {
                            c.label = o[2], c.ops.push(i);
                            break;
                        }
                        o[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    i = n.call(r, c);
                } catch (t) {
                    i = [ 6, t ], s = 0;
                } finally {
                    e = o = 0;
                }
                if (5 & i[0]) throw i[1];
                return {
                    value: i[0] ? i[1] : void 0,
                    done: !0
                };
            }([ i, t ]);
        };
    }
}, __values = this && this.__values || function(t) {
    var i = "function" == typeof Symbol && Symbol.iterator, r = i && t[i], n = 0;
    if (r) return r.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && n >= t.length && (t = void 0), {
                value: t && t[n++],
                done: !t
            };
        }
    };
    throw new TypeError(i ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var path_1 = require("path"), access_1 = require("../net/access"), error_1 = require("./error"), const_1 = require("../desc/const"), driverUtil_1 = require("../utils/driverUtil"), stringUtil_1 = require("../utils/stringUtil"), dm_1 = require("../dm"), lob_1 = require("./lob"), events_1 = require("events"), net_1 = require("net"), statement_1 = require("./statement"), tls_1 = require("tls"), fs_1 = require("fs"), os_1 = require("os"), svcConf_1 = require("../desc/svcConf"), middleware_1 = require("../middleware/middleware"), cacheQueue_1 = require("../utils/cacheQueue"), timer_1 = require("../utils/timer"), msg_1 = require("../net/msg/msg"), queryStream_1 = require("./queryStream"), Connection = function(i) {
    function r(t) {
        var r = i.call(this) || this;
        return r.serverEncoding = const_1.Const.ENCODING_GB18030, r.dbTimezone = 0, r.sslEncrypt = 0, 
        r.malini2 = !0, r.execute2 = !0, r.isoLevel = -1, r.lobEmptyCompOrcl = !1, r.lobOptimized = !1, 
        r.newLobFlag = !0, r.longLobFlag = !0, r.lobOffRowLen = 16384, r.lobReadWriteLen = 32e3, 
        r.constParaOpt = !1, r.standbyCount = 0, r.sessionID = BigInt(-1), r.trxStatus = const_1.Const.TRX_STATUS_NOT_START, 
        r.trxFinish = !0, r.dscControl = !1, r.newClientType = !1, r.backSlashFlag = !1, 
        r.algorighm = 0, r.msgVersion = msg_1.Msg.VERSION, r.conn_prop_locale = 0, r.conn_prop_compress = const_1.Const.MSG_COMPRESS_NO, 
        r.conn_prop_compressID = const_1.Const.MSG_CPR_FUN_ID_ZIP, r.conn_prop_loginEncrypt = !0, 
        r.conn_prop_isBdtaRS = !1, r.conn_prop_loginCertificate = "", r.conn_prop_sessEncode = "", 
        r.conn_prop_uKeyName = "", r.conn_prop_uKeyPin = "", r.conn_prop_columnNameCase = const_1.Const.COLUMN_NAME_NATURAL_CASE, 
        r.conn_prop_localTimezone = -new Date().getTimezoneOffset(), r.conn_prop_socketTimeout = 0, 
        r.conn_prop_connectTimeout = 5e3, r.conn_prop_sessionTimeout = 0, r.conn_prop_mppLocal = !1, 
        r.conn_prop_rwSeparate = const_1.Const.RW_SEPARATE_OFF, r.conn_prop_rwAutoDistribute = !0, 
        r.conn_prop_rwPercent = 25, r.conn_prop_rwHA = !1, r.conn_prop_rwStandbyRecoverTime = 1e3, 
        r.conn_prop_rwIgnoreSql = !1, r.conn_prop_doSwitch = const_1.Const.DO_SWITCH_OFF, 
        r.conn_prop_cluster = const_1.Const.CLUSTER_TYPE_NORMAL, r.conn_prop_cipherPath = "", 
        r.conn_prop_keyWords = new Array(), r.conn_prop_osAuthType = const_1.Const.OSAUTH_DEFAULT, 
        r.conn_prop_url = "", r.conn_prop_host = "localhost", r.conn_prop_port = const_1.Const.PORT_DEFAULT, 
        r.conn_prop_user = "SYSDBA", r.conn_prop_pwd = "SYSDBA", r.conn_prop_schema = "", 
        r.conn_prop_catalog = "", r.conn_prop_appName = "", r.conn_prop_osName = os_1.type(), 
        r.conn_prop_caseSensitive = !0, r.conn_prop_bufPrefetch = 0, r.conn_prop_escapeProcess = !1, 
        r.conn_prop_maxRows = 0, r.conn_prop_autoCommit = !0, r.conn_prop_enRsCache = !1, 
        r.conn_prop_epSelector = 0, r.conn_prop_epSelectorDynamic = !1, r.conn_prop_rsCacheSize = 20, 
        r.conn_prop_rsRefreshFreq = 10, r.conn_prop_loginMode = const_1.Const.LOGIN_MODE_NORMAL_FIRST, 
        r.conn_prop_loginStatus = 0, r.conn_prop_loginDscCtrl = !1, r.conn_prop_switchTimes = 1, 
        r.conn_prop_switchInterval = 200, r.conn_prop_continueBatchOnError = !1, r.conn_prop_batchAllowMaxErrors = 0, 
        r.conn_prop_batchType = 1, r.conn_prop_batchNotOnCall = !1, r.conn_prop_compatibleMode = 0, 
        r.conn_prop_lobMode = 1, r.conn_prop_stmtPoolSize = 15, r.conn_prop_alwaysAllowCommit = !0, 
        r.conn_prop_sslPath = "", r.conn_prop_logLevel = const_1.Const.LOG_OFF, r.conn_prop_logDir = process.cwd(), 
        r.conn_prop_logFlushFreq = 30, r.conn_prop_injectArray = !1, r.closed = !0, r.readOnly = !1, 
        r.local = !1, r.rwStandby = !1, r.props = t, r.setAttributes(t), 0 < r.conn_prop_stmtPoolSize && (r.stmtPool = new cacheQueue_1.CacheQueue(r.conn_prop_stmtPoolSize, !1), 
        r.stmtPool.beforeRemove = function(i) {
            return __awaiter(r, void 0, void 0, function() {
                return __generator(this, function(t) {
                    switch (t.label) {
                      case 0:
                        return t.trys.push([ 0, 2, , 3 ]), [ 4, this.access.P(i.handle) ];

                      case 1:
                        return t.sent(), [ 3, 3 ];

                      case 2:
                        return t.sent(), [ 3, 3 ];

                      case 3:
                        return [ 2 ];
                    }
                });
            });
        }), r;
    }
    return __extends(r, i), r.newInstanceOfConnection = function(t) {
        var i = new r(t);
        return i.newMiddlewares(), i;
    }, r.prototype.newMiddlewares = function() {
        this.middlewares = new middleware_1.Middlewares(this, this.props, r.nextID++, null);
    }, r.prototype.setAttributes = function(t) {
        var i, r;
        0 !== t.size() && (this.conn_prop_url = t.getTrimString("url", this.conn_prop_url), 
        this.conn_prop_host = t.getTrimString("host", this.conn_prop_host), this.conn_prop_port = t.getNumber("port", this.conn_prop_port), 
        this.conn_prop_user = t.getTrimString("user", this.conn_prop_user), this.conn_prop_pwd = t.getTrimString("password", this.conn_prop_pwd), 
        this.conn_prop_schema = t.getTrimString("schema", this.conn_prop_schema), this.conn_prop_catalog = t.getTrimString("catalog", this.conn_prop_catalog), 
        this.rwStandby = t.getBoolean("rwStandby", this.rwStandby), this.conn_prop_compress = t.getBoolean("isCompress", !1) ? const_1.Const.MSG_COMPRESS_AUTO : this.conn_prop_compress, 
        this.conn_prop_compress = t.getNumber("compress", this.conn_prop_compress, 0, 2), 
        this.conn_prop_compressID = t.getNumber("compressID", this.conn_prop_compressID, 0, 1), 
        this.parseLanguage(t), (i = t.getTrimString("sessEncode", this.conn_prop_sessEncode)) && !driverUtil_1.DriverUtil.isSupportedCharset(i) ? this.conn_prop_sessEncode = "" : this.conn_prop_sessEncode = i, 
        this.conn_prop_enRsCache = t.getBoolean("enRsCache", this.conn_prop_enRsCache), 
        this.conn_prop_epSelector = t.getNumber("epSelector", this.conn_prop_epSelector, 0), 
        this.conn_prop_epSelectorDynamic = t.getBoolean("epSelectorDynamic", this.conn_prop_epSelectorDynamic), 
        this.conn_prop_localTimezone = t.getNumber("localTimezone", this.conn_prop_localTimezone, -779, 840), 
        this.conn_prop_rsCacheSize = t.getNumber("rsCacheSize", this.conn_prop_rsCacheSize, 0), 
        this.conn_prop_rsRefreshFreq = t.getNumber("rsRefreshFreq", this.conn_prop_rsRefreshFreq, 0), 
        this.conn_prop_loginMode = t.getNumber("loginMode", this.conn_prop_loginMode, 0, 4), 
        this.conn_prop_loginStatus = t.getNumber("loginStatus", this.conn_prop_loginStatus, 0), 
        [ 0, 3, 4, 5 ].includes(this.conn_prop_loginStatus) || (this.conn_prop_loginStatus = 0), 
        this.conn_prop_loginDscCtrl = t.getBoolean("loginDscCtrl", this.conn_prop_loginDscCtrl), 
        this.conn_prop_switchTimes = t.getNumber("switchTimes", this.conn_prop_switchTimes, 0), 
        this.conn_prop_switchInterval = t.getNumber("switchInterval", this.conn_prop_switchInterval, 0), 
        this.conn_prop_loginEncrypt = t.getBoolean("loginEncrypt", this.conn_prop_loginEncrypt), 
        this.conn_prop_loginCertificate = t.getTrimString("loginCertificate", this.conn_prop_loginCertificate), 
        this.conn_prop_rwSeparate = t.getNumber("rwSeparate", this.conn_prop_rwSeparate, 0, 4), 
        this.conn_prop_rwAutoDistribute = t.getBoolean("rwAutoDistribute", this.conn_prop_rwAutoDistribute), 
        this.conn_prop_rwPercent = t.getNumber("rwPercent", this.conn_prop_rwPercent, 0, 100), 
        this.conn_prop_rwHA = t.getBoolean("rwHA", this.conn_prop_rwHA), this.conn_prop_rwStandbyRecoverTime = t.getNumber("rwStandbyRecoverTime", this.conn_prop_rwStandbyRecoverTime, 0), 
        this.conn_prop_rwIgnoreSql = t.getBoolean("rwIgnoreSql", this.conn_prop_rwIgnoreSql), 
        this.conn_prop_doSwitch = t.getNumber("doSwitch", this.conn_prop_doSwitch, 0, 2), 
        this.conn_prop_cluster = this.parseCluster(t), this.conn_prop_cipherPath = t.getTrimString("cipherPath", this.conn_prop_cipherPath), 
        t.getBoolean("comOra", !1) && (this.conn_prop_compatibleMode = const_1.Const.COMPATIBLE_MODE_ORACLE), 
        this.conn_prop_compatibleMode = svcConf_1.SvcConf.parseCompatibleMode(t), this.conn_prop_keyWords = t.getStringArray("keyWords", this.conn_prop_keyWords), 
        this.conn_prop_appName = t.getTrimString("appName", this.conn_prop_appName), this.conn_prop_osName = t.getTrimString("osName", this.conn_prop_osName), 
        this.conn_prop_mppLocal = t.getBoolean("mppLocal", this.conn_prop_mppLocal), this.conn_prop_socketTimeout = t.getNumber("socketTimeout", this.conn_prop_socketTimeout, 0), 
        this.conn_prop_connectTimeout = t.getNumber("connectTimeout", this.conn_prop_connectTimeout, 0), 
        this.conn_prop_sessionTimeout = t.getNumber("sessionTimeout", this.conn_prop_sessionTimeout, 0), 
        this.parseOsAuthType(t), this.conn_prop_continueBatchOnError = t.getBoolean("continueBatchOnError", this.conn_prop_continueBatchOnError), 
        this.conn_prop_batchAllowMaxErrors = t.getNumber("batchAllowMaxErrors", this.conn_prop_batchAllowMaxErrors, 0), 
        this.conn_prop_escapeProcess = t.getBoolean("escapeProcess", this.conn_prop_escapeProcess), 
        this.conn_prop_autoCommit = t.getBoolean("autoCommit", this.conn_prop_autoCommit), 
        this.conn_prop_maxRows = t.getNumber("maxRows", this.conn_prop_maxRows, 0), this.conn_prop_bufPrefetch = t.getNumber("bufPrefetch", this.conn_prop_bufPrefetch, const_1.Const.MIN_BUF_PREFETCH, const_1.Const.MAX_BUF_PREFETCH), 
        this.conn_prop_lobMode = t.getNumber("LobMode", this.conn_prop_lobMode, 1, 2), this.conn_prop_stmtPoolSize = t.getNumber("StmtPoolSize", this.conn_prop_stmtPoolSize, 0), 
        this.conn_prop_alwaysAllowCommit = t.getBoolean("alwaysAllowCommit", this.conn_prop_alwaysAllowCommit), 
        this.conn_prop_batchType = t.getNumber("batchType", this.conn_prop_batchType, 1, 2), 
        this.conn_prop_batchNotOnCall = t.getBoolean("batchNotOnCall", this.conn_prop_batchNotOnCall), 
        this.conn_prop_isBdtaRS = t.getBoolean("isBdtaRS", this.conn_prop_isBdtaRS), this.conn_prop_sslPath = t.getTrimString("sslPath", this.conn_prop_sslPath), 
        this.conn_prop_injectArray = t.getBoolean("injectArray", this.conn_prop_injectArray), 
        this.conn_prop_uKeyName = t.getTrimString("uKeyName", this.conn_prop_uKeyName), 
        this.conn_prop_uKeyPin = t.getTrimString("uKeyPin", this.conn_prop_uKeyPin), t.getBoolean("columnNameUpperCase", !1) && (this.conn_prop_columnNameCase = const_1.Const.COLUMN_NAME_UPPER_CASE), 
        "upper" === (r = t.getTrimString("columnNameCase", "")).toLowerCase() ? this.conn_prop_columnNameCase = const_1.Const.COLUMN_NAME_UPPER_CASE : "lower" === r.toLowerCase() && (this.conn_prop_columnNameCase = const_1.Const.COLUMN_NAME_LOWER_CASE), 
        this.epGroup = t.getAny("epGroup"), this.conn_prop_logLevel = svcConf_1.SvcConf.parseLogLevel(t), 
        this.conn_prop_logDir = driverUtil_1.DriverUtil.formatDir(t.getTrimString("logDir", this.conn_prop_logDir)), 
        this.conn_prop_logFlushFreq = t.getNumber("logFlushFreq", this.conn_prop_logFlushFreq, 0));
    }, r.prototype.parseOsAuthType = function(t) {
        var i = t.getTrimString("osAuthType", "");
        if (i && !stringUtil_1.StringUtil.isDigit(i) ? stringUtil_1.StringUtil.equalsIgnoreCase(i, "ON") ? this.conn_prop_osAuthType = const_1.Const.OSAUTH_DEFAULT : stringUtil_1.StringUtil.equalsIgnoreCase(i, "SYSDBA") ? this.conn_prop_osAuthType = const_1.Const.OSAUTH_DBA : stringUtil_1.StringUtil.equalsIgnoreCase(i, "SYSAUDITOR") ? this.conn_prop_osAuthType = const_1.Const.OSAUTH_AUDITOR : stringUtil_1.StringUtil.equalsIgnoreCase(i, "SYSSSO") ? this.conn_prop_osAuthType = const_1.Const.OSAUTH_SSO : stringUtil_1.StringUtil.equalsIgnoreCase(i, "AUTO") ? this.conn_prop_osAuthType = const_1.Const.OSAUTH_AUTOMATCH : stringUtil_1.StringUtil.equalsIgnoreCase(i, "OFF") && (this.conn_prop_osAuthType = const_1.Const.OSAUTH_OFF) : this.conn_prop_osAuthType = t.getNumber("osAuthType", this.conn_prop_osAuthType, 0, 4), 
        this.conn_prop_user || this.conn_prop_osAuthType !== const_1.Const.OSAUTH_OFF) {
            if (this.conn_prop_osAuthType !== const_1.Const.OSAUTH_OFF && this.conn_prop_user) throw error_1.DBError.ECJS_OSAUTH_ERROR();
            this.conn_prop_osAuthType !== const_1.Const.OSAUTH_OFF && (this.conn_prop_user = process.env["user.name"] || "", 
            this.conn_prop_pwd = "");
        } else this.conn_prop_user = "SYSDBA";
    }, r.prototype.parseLanguage = function(t) {
        var i = t.getTrimString("language", "cn");
        stringUtil_1.StringUtil.equalsIgnoreCase(i, "cn") ? this.conn_prop_locale = const_1.Const.LANGUAGE_CN : stringUtil_1.StringUtil.equalsIgnoreCase(i, "en") ? this.conn_prop_locale = const_1.Const.LANGUAGE_EN : (stringUtil_1.StringUtil.equalsIgnoreCase(i, "cnt_hk") || stringUtil_1.StringUtil.equalsIgnoreCase(i, "cnt_tw") || stringUtil_1.StringUtil.equalsIgnoreCase(i, "hk") || stringUtil_1.StringUtil.equalsIgnoreCase(i, "tw")) && (this.conn_prop_locale = const_1.Const.LANGUAGE_CNT_HK), 
        svcConf_1.SvcConf.LOCALE = this.conn_prop_locale, this.oracleDateLanguage = this.conn_prop_locale;
    }, r.prototype.parseCluster = function(t) {
        var i = t.getTrimString("cluster", "");
        return stringUtil_1.StringUtil.equalsIgnoreCase(i, "dsc") ? const_1.Const.CLUSTER_TYPE_DSC : stringUtil_1.StringUtil.equalsIgnoreCase(i, "rw") ? const_1.Const.CLUSTER_TYPE_RW : stringUtil_1.StringUtil.equalsIgnoreCase(i, "dw") ? const_1.Const.CLUSTER_TYPE_DW : stringUtil_1.StringUtil.equalsIgnoreCase(i, "mpp") ? const_1.Const.CLUSTER_TYPE_MPP : const_1.Const.CLUSTER_TYPE_NORMAL;
    }, r.prototype.setTrxFinish = function(t) {
        switch (t & const_1.Const.TRX_STATUS_STATE_MASK) {
          case const_1.Const.TRX_STATUS_NOT_START:
          case const_1.Const.TRX_STATUS_COMMITTED:
          case const_1.Const.TRX_STATUS_ROLLBACKED:
            this.trxFinish = !0;
            break;

          default:
            this.trxFinish = !1;
        }
    }, r.node2dbIsoLevel = function(t) {
        switch (t) {
          case const_1.Const.CLI_READ_UNCOMMITTED:
            return const_1.Const.ISO_LEVEL_READ_UNCOMMITTED;

          case const_1.Const.CLI_READ_COMMITTED:
            return const_1.Const.ISO_LEVEL_READ_COMMITTED;

          case const_1.Const.CLI_REPEATABLE_READ:
            return const_1.Const.ISO_LEVEL_REPEATABLE_READ;

          case const_1.Const.CLI_SERIALIZABLE:
            return const_1.Const.ISO_LEVEL_SERIALIZABLE;

          default:
            return -1;
        }
    }, r.db2nodeIsoLevel = function(t) {
        switch (t) {
          case const_1.Const.ISO_LEVEL_READ_UNCOMMITTED:
            return const_1.Const.CLI_READ_UNCOMMITTED;

          case const_1.Const.ISO_LEVEL_READ_COMMITTED:
            return const_1.Const.CLI_READ_COMMITTED;

          case const_1.Const.ISO_LEVEL_REPEATABLE_READ:
            return const_1.Const.CLI_REPEATABLE_READ;

          case const_1.Const.ISO_LEVEL_SERIALIZABLE:
            return const_1.Const.CLI_SERIALIZABLE;

          default:
            return -1;
        }
    }, r.prototype.compatibleOracle = function() {
        return this.conn_prop_compatibleMode === const_1.Const.COMPATIBLE_MODE_ORACLE;
    }, r.prototype.checkClosed = function() {
        if (this.closed) throw error_1.DBError.ECJS_CONNECTION_CLOSED();
    }, r.prototype.getIndexOnEPGroup = function() {
        if (!this.epGroup || !this.epGroup.epList) return -1;
        for (var t = 0; t < this.epGroup.epList.length; t++) {
            var i = this.epGroup.epList[t];
            if (i.host === this.conn_prop_host && i.port === this.conn_prop_port) return t;
        }
        return -1;
    }, r.prototype.openConnection = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.middlewares.valid ? this.middlewares.reset().connection_openConnection(this, i) : this.do_openConnection(i) ];
            });
        });
    }, r.prototype.execute = function(t, i, r, n) {
        if (void 0 === i && void 0 === r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_execute(this, t) : this.do_execute(t);
        if ("function" == typeof i && void 0 === r && void 0 === n) this.execute(t).then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, void 0);
        }); else {
            if ("function" != typeof i && void 0 === r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_execute(this, t, i) : this.do_execute(t, i);
            if ("function" != typeof i && "function" == typeof r && void 0 === n) this.execute(t, i).then(function(t) {
                r(null, t);
            }).catch(function(t) {
                r(t, void 0);
            }); else {
                if ("function" != typeof i && "function" != typeof r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_execute(this, t, i, r) : this.do_execute(t, i, r);
                if ("function" == typeof i || "function" == typeof r || "function" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
                this.execute(t, i, r).then(function(t) {
                    n(null, t);
                }).catch(function(t) {
                    n(t, void 0);
                });
            }
        }
    }, r.prototype.executeMany = function(t, i, r, n) {
        if ("number" == typeof i && void 0 === r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_executeMany(this, t, i) : this.do_executeMany(t, i);
        if ("number" == typeof i && "function" == typeof r && void 0 === n) this.executeMany(t, i).then(function(t) {
            r(null, t);
        }).catch(function(t) {
            r(t, void 0);
        }); else {
            if ("number" == typeof i && "function" != typeof r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_executeMany(this, t, i, r) : this.do_executeMany(t, i, r);
            if ("number" == typeof i && "function" != typeof r && "function" == typeof n) this.executeMany(t, i, r).then(function(t) {
                n(null, t);
            }).catch(function(t) {
                n(t, void 0);
            }); else {
                if (Array.isArray(i) && void 0 === r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_executeMany(this, t, i) : this.do_executeMany(t, i);
                if (Array.isArray(i) && "function" == typeof r && void 0 === n) this.executeMany(t, i).then(function(t) {
                    r(null, t);
                }).catch(function(t) {
                    r(t, void 0);
                }); else {
                    if (Array.isArray(i) && "function" != typeof r && void 0 === n) return this.middlewares.valid ? this.middlewares.reset().connection_executeMany(this, t, i, r) : this.do_executeMany(t, i, r);
                    if (!Array.isArray(i) || "function" == typeof r || "function" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
                    this.executeMany(t, i, r).then(function(t) {
                        n(null, t);
                    }).catch(function(t) {
                        n(t, void 0);
                    });
                }
            }
        }
    }, r.prototype.queryStream = function(n, e, s) {
        var t = this;
        (s = s || {}).resultSet = !0;
        var o = new queryStream_1.QueryStream();
        return process.nextTick(function() {
            return __awaiter(t, void 0, void 0, function() {
                var i, r;
                return __generator(this, function(t) {
                    switch (t.label) {
                      case 0:
                        return t.trys.push([ 0, 2, , 3 ]), [ 4, this.execute(n, e || [], s) ];

                      case 1:
                        if (!(i = t.sent())) throw error_1.DBError.ECJS_RESULTSET_ABSENT();
                        return o._open(i.resultSet), [ 3, 3 ];

                      case 2:
                        return r = t.sent(), o.destroy(r), [ 2 ];

                      case 3:
                        return [ 2 ];
                    }
                });
            });
        }), o;
    }, r.prototype.createLob = function(t, i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().connection_createLob(this, t) : this.do_createLob(t);
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.createLob(t).then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, void 0);
        });
    }, r.prototype.close = function(i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().connection_close(this) : this.do_close();
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.close().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
    }, r.prototype.commit = function(i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().connection_commit(this) : this.do_commit();
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.commit().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
    }, r.prototype.getStatementInfo = function(t, i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().connection_getStatementInfo(this, t) : this.do_getStatementInfo(t);
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getStatementInfo(t).then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, void 0);
        });
    }, r.prototype.ping = function(i) {
        if (void 0 === i) return this.execute("select 1").then(function() {});
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.ping().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
    }, r.prototype.release = function(i) {
        if (void 0 === i) return this.close();
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.release().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
    }, r.prototype.rollback = function(i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().connection_rollback(this) : this.do_rollback();
        if ("function" == typeof i) return this.rollback().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
        throw error_1.DBError.ECJS_INVALID_PARAM();
    }, r.prototype.changePassword = function(t, i, r, n) {
        if (void 0 === n) {
            if ("string" != typeof t || "string" != typeof i || "string" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
            return this.execute('alter user "' + t + '" identified by "' + r + '"').then(function() {});
        }
        if ("function" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.changePassword(t, i, r).then(function() {
            n(null);
        }).catch(function(t) {
            n(t);
        });
    }, r.prototype._innerFree = function() {
        return this.middlewares.valid ? this.middlewares.reset().connection_free(this) : this.do_free();
    }, r.prototype.do_free = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i, r = this;
            return __generator(this, function(t) {
                return this.stmtPool && this.stmtPool.clear(), (i = this.middlewares.log) && i.logWriter.end(), 
                [ 2, new Promise(function(t, i) {
                    process.nextTick(function() {
                        r.socket ? (r.doUnbinds(r.socket), r.socket.destroyed ? t() : (r.socket.destroy(), 
                        r.socket.once("close", function() {
                            r.closed = !0, t();
                        }), r.socket.once("error", function(t) {
                            i(t);
                        }))) : (r.closed = !0, t());
                    });
                }) ];
            });
        });
    }, r.prototype.do_close = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.closed) return [ 3, 9 ];
                    t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, , 4 ]), [ 4, this.compatibleOracle() ? this.commit() : this.rollback() ];

                  case 2:
                    return t.sent(), [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 4 ];

                  case 4:
                    return this instanceof PoolConnection ? [ 4, this.return() ] : [ 3, 6 ];

                  case 5:
                    return t.sent(), [ 3, 8 ];

                  case 6:
                    return [ 4, this.do_free() ];

                  case 7:
                    t.sent(), t.label = 8;

                  case 8:
                    t.label = 9;

                  case 9:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.getServerEncoding = function() {
        return this.conn_prop_sessEncode || this.serverEncoding;
    }, r.prototype.lobFetchAll = function() {
        return 2 === this.conn_prop_lobMode;
    }, r.prototype.do_openConnection = function(e) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, n = this;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (e && 0 < e.epList.length) return [ 2, e.connect(this) ];
                    t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 8, 9, 12 ]), this.access = new access_1.Access(this), [ 4, new Promise(function(t, i) {
                        var r = setTimeout(function() {
                            n.socket && (n.socket.removeAllListeners(), n.socket.destroy()), i(error_1.DBError.ECJS_CONNECT_TIMEOUT());
                        }, n.conn_prop_connectTimeout);
                        n.socket = net_1.createConnection({
                            host: n.conn_prop_host,
                            port: n.conn_prop_port
                        }), n.socket.once("connect", function() {
                            clearTimeout(r), n.socket.removeAllListeners("error"), n.doBinds(n.socket), n.emit("connect"), 
                            t();
                        }), n.socket.on("error", function(t) {
                            clearTimeout(r), i(t);
                        });
                    }) ];

                  case 2:
                    return t.sent(), [ 4, this.access.F() ];

                  case 3:
                    if (t.sent(), "" === this.conn_prop_schema) return [ 3, 7 ];
                    t.label = 4;

                  case 4:
                    return t.trys.push([ 4, 6, , 7 ]), this.closed = !1, [ 4, this.do_executeInner('set schema "' + stringUtil_1.StringUtil.processDoubleQuoteOfName(this.conn_prop_schema) + '"') ];

                  case 5:
                    return t.sent(), [ 3, 7 ];

                  case 6:
                    throw i = t.sent(), this.closed = !0, i;

                  case 7:
                    return this.closed = !1, [ 2, this ];

                  case 8:
                    throw (r = t.sent()) instanceof error_1.DBError ? r : error_1.DBError.ECJS_COMMUNICATION_ERROR(r + "");

                  case 9:
                    return this.closed ? [ 4, this.do_free() ] : [ 3, 11 ];

                  case 10:
                    t.sent(), t.label = 11;

                  case 11:
                    return [ 7 ];

                  case 12:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.do_execute = function(o, c, h) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, n, e, s;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if ("string" != typeof o) throw error_1.DBError.ECJS_INVALID_PARAM();
                    if (this.checkClosed(), (h = {
                        autoCommit: (h = h || {}).autoCommit,
                        extendedMetaData: void 0 !== h.extendedMetaData ? h.extendedMetaData : dm_1.dmdb.extendedMetaData,
                        fetchInfo: h.fetchInfo,
                        maxRows: h.maxRows,
                        outFormat: void 0 !== h.outFormat ? h.outFormat : dm_1.dmdb.outFormat,
                        resultSet: void 0 !== h.resultSet && h.resultSet,
                        innerExec: void 0 !== h.innerExec && h.innerExec
                    }).outFormat !== dm_1.dmdb.OUT_FORMAT_ARRAY && h.outFormat !== dm_1.dmdb.OUT_FORMAT_OBJECT) throw error_1.DBError.ECJS_INVALID_OUTFORMAT();
                    return Array.isArray(c) && 0 < c.length ? [ 4, (i = new statement_1.PreparedStatement(this, h, !1)).allocateHandle() ] : [ 3, 6 ];

                  case 1:
                    return t.sent(), [ 4, i.prepareSql(o, c) ];

                  case 2:
                    if ((c = t.sent()).length > i.serverParams.length) throw error_1.DBError.ECJS_TOO_MANY_BINDS();
                    for (r = new Array(i.serverParams.length), n = 0; n < i.serverParams.length; n++) n < c.length ? r[n] = dm_1.dmdb.isBindParameter(c[n]) ? c[n] : {
                        dir: dm_1.dmdb.BIND_IN,
                        val: c[n]
                    } : r[n] = {
                        dir: dm_1.dmdb.BIND_OUT,
                        val: void 0
                    };
                    return [ 4, i.executePstmt(r) ];

                  case 3:
                    return t.sent(), i.execInfo.result.resultSet ? [ 3, 5 ] : [ 4, i.close() ];

                  case 4:
                    t.sent(), t.label = 5;

                  case 5:
                    return [ 2, i.execInfo.result ];

                  case 6:
                    return "object" != typeof c || Array.isArray(c) ? [ 3, 12 ] : ((i = new statement_1.PreparedStatement(this, h, !1)).isBindArray = !1, 
                    [ 4, i.allocateHandle() ]);

                  case 7:
                    return t.sent(), [ 4, i.prepareSql(o, c) ];

                  case 8:
                    for (c = t.sent(), r = new Array(i.serverParams.length), n = 0; n < i.serverParams.length; n++) {
                        for (s in e = i.serverParams[n], c) if (c.hasOwnProperty(s) && stringUtil_1.StringUtil.equalsIgnoreCase(s, e.name)) {
                            e._bindName = s, r[n] = dm_1.dmdb.isBindParameter(c[s]) ? c[s] : {
                                dir: dm_1.dmdb.BIND_IN,
                                val: c[s]
                            };
                            break;
                        }
                        void 0 === r[n] && (r[n] = {
                            dir: dm_1.dmdb.BIND_OUT,
                            val: void 0
                        });
                    }
                    return [ 4, i.executePstmt(r) ];

                  case 9:
                    return t.sent(), i.execInfo.result.resultSet ? [ 3, 11 ] : [ 4, i.close() ];

                  case 10:
                    t.sent(), t.label = 11;

                  case 11:
                    return [ 2, i.execInfo.result ];

                  case 12:
                    return [ 2, this.do_executeNoParam(o, h) ];
                }
            });
        });
    }, r.prototype.do_executeMany = function(v, w, m) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, n, e, s, o, c, h, u, a, f, _, d, l;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if ("string" != typeof v) throw error_1.DBError.ECJS_INVALID_PARAM();
                    return (this.checkClosed(), m = {
                        autoCommit: (m = m || {}).autoCommit,
                        batchErrors: void 0 !== m.batchErrors ? m.batchErrors : this.conn_prop_continueBatchOnError,
                        dmlRowCounts: void 0 !== m.dmlRowCounts && m.dmlRowCounts,
                        bindDefs: m.bindDefs
                    }, "number" != typeof w) ? [ 3, 7 ] : (error_1.DBError.assertParamValue(Number.isSafeInteger(w) && 0 < w, "iterations", w), 
                    [ 4, (i = new statement_1.PreparedStatement(this, m, !0)).allocateHandle() ]);

                  case 1:
                    return t.sent(), [ 4, i.prepareSql("BEGIN FOR i IN 1.." + w + " LOOP " + v + (v.trimRight().endsWith(";") ? "" : ";") + " END LOOP; END;") ];

                  case 2:
                    return t.sent(), 0 < i.paramCount ? [ 4, i.close() ] : [ 3, 4 ];

                  case 3:
                    throw t.sent(), error_1.DBError.ECJS_INVALID_SQL();

                  case 4:
                    return [ 4, i.executeManyPstmt([]) ];

                  case 5:
                    return t.sent(), [ 4, i.close() ];

                  case 6:
                    return t.sent(), delete i.execInfo.results.dmlRowCounts, delete i.execInfo.results.rowsAffected, 
                    [ 2, i.execInfo.results ];

                  case 7:
                    return Array.isArray(w) ? [ 4, (i = new statement_1.PreparedStatement(this, m, !0)).allocateHandle() ] : [ 3, 12 ];

                  case 8:
                    return t.sent(), [ 4, i.prepareSql(v) ];

                  case 9:
                    t.sent(), r = void 0;
                    try {
                        for (n = __values(w), e = n.next(); !e.done; e = n.next()) if (c = e.value, Array.isArray(c)) {
                            if (void 0 !== r && !r) throw error_1.DBError.ECJS_MIXED_BIND();
                            r = !0;
                        } else if ("object" == typeof c) {
                            if (void 0 !== r && r) throw error_1.DBError.ECJS_MIXED_BIND();
                            r = !1;
                        }
                    } catch (t) {
                        d = {
                            error: t
                        };
                    } finally {
                        try {
                            e && !e.done && (l = n.return) && l.call(n);
                        } finally {
                            if (d) throw d.error;
                        }
                    }
                    if (void 0 !== m.bindDefs) if (Array.isArray(m.bindDefs)) {
                        if (void 0 !== r && !r) throw error_1.DBError.ECJS_MIXED_BIND();
                        r = !0;
                    } else if ("object" == typeof m.bindDefs) {
                        if (void 0 !== r && r) throw error_1.DBError.ECJS_MIXED_BIND();
                        r = !1;
                    }
                    for (i.isBindArray = r, s = new Array(w.length), o = 0; o < w.length; o++) if (c = w[o], 
                    Array.isArray(c)) {
                        if (c.length > i.serverParams.length) throw error_1.DBError.ECJS_TOO_MANY_BINDS();
                        for (u = new Array(i.serverParams.length), h = 0; h < i.serverParams.length; h++) h < c.length ? dm_1.dmdb.isBindParameter(c[h]) ? u[h] = c[h] : u[h] = {
                            dir: dm_1.dmdb.BIND_IN,
                            val: c[h]
                        } : u[h] = {
                            dir: dm_1.dmdb.BIND_OUT,
                            val: void 0
                        }, m.bindDefs && Array.isArray(m.bindDefs) && h < m.bindDefs.length && (void 0 !== m.bindDefs[h].dir && (u[h].dir = m.bindDefs[h].dir), 
                        void 0 !== m.bindDefs[h].type && (u[h].type = m.bindDefs[h].type));
                        s[o] = u;
                    } else if ("object" == typeof c) {
                        for (u = new Array(i.serverParams.length), a = 0; a < i.serverParams.length; a++) {
                            for (_ in f = i.serverParams[a], c) if (c.hasOwnProperty(_) && stringUtil_1.StringUtil.equalsIgnoreCase(_, f.name)) {
                                f._bindName = _, dm_1.dmdb.isBindParameter(c[_]) ? u[a] = c[_] : u[a] = {
                                    dir: dm_1.dmdb.BIND_IN,
                                    val: c[_],
                                    type: void 0
                                };
                                break;
                            }
                            if (void 0 === u[a] && (u[a] = {
                                dir: dm_1.dmdb.BIND_OUT,
                                val: void 0,
                                type: void 0
                            }), m.bindDefs && !Array.isArray(m.bindDefs)) for (_ in m.bindDefs) if (m.bindDefs.hasOwnProperty(_) && stringUtil_1.StringUtil.equalsIgnoreCase(_, f.name)) {
                                void 0 !== m.bindDefs[_].dir && (u[a].dir = m.bindDefs[_].dir), void 0 !== m.bindDefs[_].type && (u[a].type = m.bindDefs[_].type);
                                break;
                            }
                        }
                        s[o] = u;
                    }
                    return [ 4, i.executeManyPstmt(s) ];

                  case 10:
                    return t.sent(), [ 4, i.close() ];

                  case 11:
                    return t.sent(), [ 2, i.execInfo.results ];

                  case 12:
                    throw new Error("Function is called with invalid params!");
                }
            });
        });
    }, r.prototype.do_createLob = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return this.checkClosed(), [ 2, lob_1.Lob.newLob(i, this) ];
            });
        });
    }, r.prototype.do_commit = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.checkClosed(), void 0 === (i = dm_1.dmdb.autoCommit) && (i = this.conn_prop_autoCommit), 
                    !i || this.conn_prop_alwaysAllowCommit) return [ 3, 1 ];
                    throw error_1.DBError.ECJS_COMMIT_IN_AUTOCOMMIT_MODE();

                  case 1:
                    return [ 4, this.access.R() ];

                  case 2:
                    t.sent(), this.trxFinish = !0, t.label = 3;

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.do_getStatementInfo = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if ("string" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
                    return this.checkClosed(), [ 4, (i = new statement_1.PreparedStatement(this, {
                        extendedMetaData: dm_1.dmdb.extendedMetaData,
                        outFormat: dm_1.dmdb.outFormat,
                        resultSet: !1
                    }, !1)).allocateHandle() ];

                  case 1:
                    return t.sent(), [ 4, i.prepareSql(n) ];

                  case 2:
                    return t.sent(), r = {}, 0 < i.serverParams.length && (r.bindNames = [], i.serverParams.forEach(function(t) {
                        r.bindNames.push(t.name);
                    })), 0 < i.columns.length && (r.metaData = i.getMetaData(dm_1.dmdb.extendedMetaData)), 
                    r.statementType = i.execInfo.retSqlType, [ 2, r ];
                }
            });
        });
    }, r.prototype.do_rollback = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.checkClosed(), void 0 === (i = dm_1.dmdb.autoCommit) && (i = this.conn_prop_autoCommit), 
                    !i || this.conn_prop_alwaysAllowCommit) return [ 3, 1 ];
                    throw error_1.DBError.ECJS_ROLLBACK_IN_AUTOCOMMIT_MODE();

                  case 1:
                    return [ 4, this.access.I() ];

                  case 2:
                    t.sent(), this.trxFinish = !0, t.label = 3;

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.do_reconnect = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.do_free() ];

                  case 1:
                    return t.sent(), this.epGroup ? [ 4, this.epGroup.connect(this) ] : [ 3, 3 ];

                  case 2:
                    return t.sent(), [ 3, 5 ];

                  case 3:
                    return [ 4, this.do_openConnection() ];

                  case 4:
                    t.sent(), t.label = 5;

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.do_executeInner = function(r, n) {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                return i = {
                    extendedMetaData: !1,
                    maxRows: 0,
                    outFormat: dm_1.dmdb.OUT_FORMAT_ARRAY,
                    resultSet: !0,
                    innerExec: !0
                }, [ 2, this.do_execute(r, n, i) ];
            });
        });
    }, r.prototype.do_executeNoParam = function(n, e) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, (i = new statement_1.Statement(this, e, !1)).allocateHandle() ];

                  case 1:
                    return t.sent(), [ 4, i.execute(n) ];

                  case 2:
                    return (r = t.sent()).resultSet ? [ 3, 4 ] : [ 4, i.close() ];

                  case 3:
                    t.sent(), t.label = 4;

                  case 4:
                    return [ 2, r ];
                }
            });
        });
    }, r.prototype._handleNetWorkError = function(t) {
        this.access.M(t);
    }, r.prototype._handleConnectTimeout = function() {
        this.socket && (this.socket.setTimeout(0), this.socket.destroy()), this._handleNetWorkError(error_1.DBError.ECJS_CONNECT_TIMEOUT());
    }, r.prototype.doBinds = function(i) {
        var r = this;
        tls_1.TLSSocket;
        return i.setNoDelay(!0), i.setKeepAlive(!0), i.on("error", function(t) {
            r._handleNetWorkError(error_1.DBError.ECJS_COMMUNICATION_ERROR(t.message));
        }), i.on("end", function() {
            r.access.Y(), r.closed = !0;
        }), i.on("close", function() {
            r.closed = !0;
        }), i.on("data", function(t) {
            r.access.z(t);
        }), this.access.on("data", function(t) {
            i.write(t);
        }), this.access.on("end", function() {
            i.end();
        }), i;
    }, r.prototype.doUnbinds = function(t) {
        t.removeAllListeners(), this.access.removeAllListeners();
    }, r.prototype.startTLS = function(n) {
        var e = this;
        return new Promise(function(t, i) {
            var r = e._createSecureContext();
            n ? (e.tlsSocket = tls_1.connect({
                secureContext: r,
                checkServerIdentity: function() {},
                socket: e.socket,
                rejectUnauthorized: !1,
                requestCert: !0
            }), e.tlsSocket.on("secureConnect", function() {
                t();
            }), e.tlsSocket.on("error", function(t) {
                i(t);
            }), e.doUnbinds(e.socket), e.socket = e.doBinds(e.tlsSocket)) : i(new Error("SSL Authentication not support"));
        });
    }, r.prototype._createSecureContext = function() {
        try {
            if (!this.conn_prop_sslPath) {
                var t = process.env.DM_HOME;
                if (!t) throw new Error("sslPath can not be empty.");
                this.conn_prop_sslPath = "" + t + path_1.sep + "bin" + path_1.sep + "client_ssl" + path_1.sep + this.conn_prop_user + path_1.sep;
            }
            return tls_1.createSecureContext({
                cert: fs_1.readFileSync(path_1.join(this.conn_prop_sslPath, "client-cert.pem")),
                key: fs_1.readFileSync(path_1.join(this.conn_prop_sslPath, "client-key.pem")),
                ca: [ fs_1.readFileSync(path_1.join(this.conn_prop_sslPath, "ca-cert.pem")) ],
                honorCipherOrder: !0,
                minVersion: "TLSv1",
                maxVersion: "TLSv1.3"
            });
        } catch (t) {
            throw error_1.DBError.ECJS_INIT_SSL_FAILED(t + "");
        }
    }, r.nextID = 0, r;
}(events_1.EventEmitter), PoolConnection = function(n) {
    function e(t, i) {
        var r = n.call(this, t) || this;
        return r.pool = i, r.on("close", function(t) {
            var i = void 0;
            try {
                r.pool.releaseConnection(r);
            } catch (t) {
                i = t;
            } finally {
                return void process.nextTick(t, i);
            }
        }), r;
    }
    return __extends(e, n), e.newInstanceOfPoolConnection = function(t, i) {
        var r = new e(t, i);
        return r.newMiddlewares(), r.idleCloseTimer = new timer_1.Timer(function() {
            r.pool.eliminateConnection(r);
        }), r;
    }, e.prototype.newMiddlewares = function() {
        this.middlewares = new middleware_1.Middlewares(this, this.props, e.nextID++, this.pool.middlewares);
    }, e.prototype.return = function() {
        return __awaiter(this, void 0, void 0, function() {
            var n = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(i, r) {
                    n.emit("close", function(t) {
                        t ? r(t) : i();
                    });
                }) ];
            });
        });
    }, e.prototype.startTimer = function() {
        0 < this.pool.poolTimeout && this.idleCloseTimer && this.idleCloseTimer.start(1e3 * this.pool.poolTimeout);
    }, e.prototype.refreshTimer = function() {
        this.idleCloseTimer && this.idleCloseTimer.refresh();
    }, e.prototype.stopTimer = function() {
        this.idleCloseTimer && this.idleCloseTimer.stop();
    }, e;
}(exports.Connection = Connection);

exports.PoolConnection = PoolConnection;
