// dmdb/src/driver/numeric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var error_1 = require("./error"), bufferUtil_1 = require("../utils/bufferUtil"), Numeric = function() {
    function c(t, i, r) {
        if (this.sign = 0, this.weight = 0, this.prec = 0, this.scale = 0, this.hasScale = !1, 
        this.digits = "", this.integerPart = 0, this.prec = t, this.scale = i, this.hasScale = !(-1 === i || t <= 0 && i <= 0), 
        void 0 !== r) {
            if (null === r || 0 === r.length || r.length > c.XDEC_SIZE) throw error_1.DBError.ECJS_FATAL_ERROR();
            if (r[0] !== c.FLAG_ZERO && 1 !== r.length) {
                this.sign = 0 != (r[0] & c.FLAG_ZERO) ? 1 : -1;
                for (var h, s = bufferUtil_1.BufferUtil.H(r, 0), e = 0 < this.sign ? s - c.FLAG_POSITIVE : c.FLAG_NEGTIVE - s, u = "", f = 1; f < r.length && !((h = 0 < this.sign ? r[f] - c.NUM_POSITIVE : c.NUM_NEGTIVE - r[f]) < 0 || 99 < h); f++) h < 10 && (u += "0"), 
                u += h;
                this.digits = u, this.weight = 2 * e - (this.digits.length - 2);
            }
        }
    }
    return c.fromNumber = function(t, i, r) {
        var h = new c(i, r);
        if (h.sign = Math.sign(t), 0 === t) return h;
        var s = Math.abs(t).toLocaleString("en-IN", {
            useGrouping: !1,
            maximumFractionDigits: 20
        });
        return "0" !== s && "-0" !== s || "0.00000000000000000000000000000000000000" !== (s = Math.abs(t).toFixed(38)) ? this.fromString((-1 === h.sign ? "-" : "") + s, i, r) : (h.sign = 0, 
        h);
    }, c.fromString = function(t, i, r) {
        if (Number.isNaN(Number(t))) throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
        var h = new c(i, r);
        return h.parse(t.toLowerCase()), h;
    }, c.prototype.parse = function(t) {
        this.sign = Math.sign(Number(t)), -1 === this.sign && (t = t.substr(1));
        var i, r = t.indexOf("e");
        -1 !== r && (i = Number(t.substr(r + 1)), this.weight += i, t = t.substring(0, r));
        var h = t.indexOf("."), s = 0, e = 0, u = t.length;
        if (-1 !== h) if ("0" === t[0]) {
            for (s = e = 2; s < u && "0" === t[s]; s++) ;
            t = t.substring(s), this.weight -= s - e + t.length;
        } else t = t.substring(0, h) + t.substring(h + 1), this.weight -= u - h - 1;
        for (s = e = (u = t.length) - 1; 0 < s && "0" === t[s]; s--) ;
        t = t.substring(0, s + 1) + t.substring(u), this.weight += e - s, this.digits = t, 
        this.calcIntegerPart();
    }, c.prototype.encode = function(t) {
        if (this.isZero()) return Buffer.from([ c.FLAG_ZERO ]);
        this.checkPrec(t), this.weight % 2 != 0 && (this.digits += "0", this.weight--), 
        this.digits.length % 2 != 0 && (this.digits = "0" + this.digits, this.integerPart++);
        var i = (this.weight + this.digits.length) / 2 - 1;
        this.checkExp(0 <= this.sign, i);
        var r = this.digits.length, h = this.digits.length / 2 + 1, s = 0;
        if (c.XDEC_SIZE < h) {
            if (!t) throw error_1.DBError.EC_DATA_OVERFLOW();
            r = 2 * ((h = c.XDEC_SIZE) - 1), s = 5 <= Number(this.digits.charAt(r)) ? 1 : 0;
        }
        for (var e = this.sign < 0 && h < c.XDEC_SIZE ? h + 1 : h, u = Buffer.alloc(e), f = h - 1, n = !0, o = r - 1; 0 < o && 0 < f; f--) {
            var a, s = 99 < (a = Number(this.digits.charAt(o--)) + 10 * Number(this.digits.charAt(o--)) + s) ? 1 : 0;
            99 < a && (a %= 100), (n = n && 0 === a) && h--, u[f] = 0 < this.sign ? a + c.NUM_POSITIVE : c.NUM_NEGTIVE - a;
        }
        return n && 0 < s && (i += 1, this.checkExp(0 <= this.sign, i), a = s, u[h] = 0 < this.sign ? a + c.NUM_POSITIVE : c.NUM_NEGTIVE - a, 
        h++), u[0] = 0 < this.sign ? i + c.FLAG_POSITIVE : c.FLAG_NEGTIVE - i, this.sign < 0 && h < c.XDEC_SIZE && (u[h++] = 102), 
        h < u.length && (u = u.subarray(0, h)), u;
    }, c.prototype.toNumber = function() {
        if (this.isZero()) return Number(0);
        var t = (this.sign < 0 ? "-" : "") + this.digits, i = Number(t);
        return 0 < this.weight ? i *= Number("1" + "0".repeat(this.weight)) : this.weight < 0 && (i /= Number("1" + "0".repeat(-this.weight))), 
        i;
    }, c.prototype.toString = function() {
        if (this.isZero()) return "0";
        var t = this.digits;
        return 0 < this.weight ? t += "0".repeat(this.weight) : this.weight < 0 && (t = (t = t.padStart(1 - this.weight, "0")).substring(0, t.length + this.weight) + "." + t.substring(t.length + this.weight)), 
        "0" === t[0] && "." !== t[1] && (t = t.substr(1)), "0" === t[t.length - 1] && 0 <= t.indexOf(".") && (t = t.substr(0, t.length - 1)), 
        t = (this.sign < 0 ? "-" : "") + t;
    }, c.prototype.isZero = function() {
        return 0 === this.sign;
    }, c.prototype.calcIntegerPart = function() {
        var t, i, r, h;
        this.digits.length < c.XDEC_MAX_PREC ? this.integerPart = this.digits.length : ("-" === (t = this.toString()).charAt(0) && (t = t.substr(1)), 
        r = 0 < (i = t.indexOf(".")) ? t.length - 1 : t.length, h = 0 < i ? t.length - 1 - i : 0, 
        this.integerPart = 0 == h ? r - this.weight : 0 < h ? Math.max(1, r - h) : 1);
    }, c.prototype.checkPrec = function(t) {
        var i = this.digits.length;
        if (this.hasScale && (i > this.prec || c.XDEC_MAX_PREC < i)) throw error_1.DBError.EC_DATA_OVERFLOW();
        if (!t && this.integerPart > c.XDEC_MAX_PREC) throw error_1.DBError.EC_DATA_OVERFLOW();
    }, c.prototype.checkExp = function(t, i) {
        if ((t ? c.POSITIVE_EXP_MAX : c.EXP_MAX) < i || i < c.EXP_MIN) throw error_1.DBError.EC_DATA_OVERFLOW();
    }, c.XDEC_MAX_PREC = 40, c.XDEC_SIZE = 21, c.FLAG_ZERO = 128, c.FLAG_POSITIVE = 193, 
    c.FLAG_NEGTIVE = 62, c.POSITIVE_EXP_MAX = 255 - c.FLAG_POSITIVE, c.EXP_MAX = 254 - c.FLAG_POSITIVE, 
    c.EXP_MIN = c.FLAG_NEGTIVE + 1 - 127, c.NUM_POSITIVE = 1, c.NUM_NEGTIVE = 101, c;
}();

exports.Numeric = Numeric;
