// dmdb/src/driver/pool.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, o, s, c) {
    return new (s = s || Promise)(function(r, e) {
        function i(t) {
            try {
                u(c.next(t));
            } catch (t) {
                e(t);
            }
        }
        function n(t) {
            try {
                u(c.throw(t));
            } catch (t) {
                e(t);
            }
        }
        function u(t) {
            var e;
            t.done ? r(t.value) : ((e = t.value) instanceof s ? e : new s(function(t) {
                t(e);
            })).then(i, n);
        }
        u((c = c.apply(t, o || [])).next());
    });
}, __generator = this && this.__generator || function(r, i) {
    var n, u, o, s = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: e(0),
        throw: e(1),
        return: e(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function e(e) {
        return function(t) {
            return function(e) {
                if (n) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (n = 1, u && (o = 2 & e[0] ? u.return : e[0] ? u.throw || ((o = u.return) && o.call(u), 
                    0) : u.next) && !(o = o.call(u, e[1])).done) return o;
                    switch (u = 0, o && (e = [ 2 & e[0], o.value ]), e[0]) {
                      case 0:
                      case 1:
                        o = e;
                        break;

                      case 4:
                        return s.label++, {
                            value: e[1],
                            done: !1
                        };

                      case 5:
                        s.label++, u = e[1], e = [ 0 ];
                        continue;

                      case 7:
                        e = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = s.trys).length && o[o.length - 1]) && (6 === e[0] || 2 === e[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === e[0] && (!o || e[1] > o[0] && e[1] < o[3])) {
                            s.label = e[1];
                            break;
                        }
                        if (6 === e[0] && s.label < o[1]) {
                            s.label = o[1], o = e;
                            break;
                        }
                        if (o && s.label < o[2]) {
                            s.label = o[2], s.ops.push(e);
                            break;
                        }
                        o[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    e = i.call(r, s);
                } catch (t) {
                    e = [ 6, t ], u = 0;
                } finally {
                    n = o = 0;
                }
                if (5 & e[0]) throw e[1];
                return {
                    value: e[0] ? e[1] : void 0,
                    done: !0
                };
            }([ e, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var events_1 = require("events"), error_1 = require("./error"), connection_1 = require("./connection"), dm_1 = require("../dm"), middleware_1 = require("../middleware/middleware"), unenumerable_1 = require("../utils/unenumerable"), deque_1 = require("../utils/deque"), Pool = function() {
    function t(t) {
        if (this._poolMax = dm_1.dmdb.poolMax, this._poolMin = dm_1.dmdb.poolMin, this._poolTimeout = dm_1.dmdb.poolTimeout, 
        this._queueMax = dm_1.dmdb.queueMax, this._queueRequests = dm_1.dmdb.queueRequests, 
        this._queueTimeout = dm_1.dmdb.queueTimeout, this._testOnBorrow = !1, this._validationQuery = "select 1;", 
        this._connectString = "", this._status = dm_1.dmdb.POOL_STATUS_OPEN, this._acquiringConnections = new deque_1.Deque(), 
        this._freeConnections = new deque_1.Deque(), this._connectionQueue = new deque_1.Deque(), 
        this._connectString = t.connectString || t.connectionString || this._connectString, 
        void 0 !== t.poolMax && (error_1.DBError.assertParamValue(Number.isInteger(t.poolMax) && 0 < t.poolMax, "poolMax", t.poolMax), 
        this._poolMax = t.poolMax), void 0 !== t.poolMin && (error_1.DBError.assertParamValue(Number.isInteger(t.poolMin) && 0 <= t.poolMin, "poolMin", t.poolMin), 
        this._poolMin = t.poolMin), this._poolMax < this._poolMin) throw error_1.DBError.ECJS_INVALID_POOL_NUM();
        void 0 !== t.poolTimeout && (error_1.DBError.assertParamValue(Number.isInteger(t.poolTimeout) && 0 <= t.poolTimeout, "poolTimeout", t.poolTimeout), 
        this._poolTimeout = t.poolTimeout), void 0 !== t.queueMax && (error_1.DBError.assertParamValue(Number.isInteger(t.queueMax) && -1 <= t.queueMax, "queueMax", t.queueMax), 
        this._queueMax = t.queueMax), void 0 !== t.queueRequests && (error_1.DBError.assertParamValue("boolean" == typeof t.queueRequests, "queueRequests", t.queueRequests), 
        this._queueRequests = t.queueRequests), void 0 !== t.queueTimeout && (error_1.DBError.assertParamValue(Number.isInteger(t.queueTimeout) && 0 <= t.queueTimeout, "queueTimeout", t.queueTimeout), 
        this._queueTimeout = t.queueTimeout), void 0 !== t.testOnBorrow && (error_1.DBError.assertParamValue("boolean" == typeof t.testOnBorrow, "testOnBorrow", t.testOnBorrow), 
        this._testOnBorrow = t.testOnBorrow), void 0 !== t.validationQuery && (error_1.DBError.assertParamValue("string" == typeof t.validationQuery, "validationQuery", t.validationQuery), 
        this._validationQuery = t.validationQuery), void 0 !== t.poolAlias && (error_1.DBError.assertParamValue("string" == typeof t.poolAlias, "poolAlias", t.poolAlias), 
        this._poolAlias = t.poolAlias);
    }
    return t.prototype.newMiddlewares = function() {
        this.middlewares = new middleware_1.Middlewares(null, this._connectProps, t.nextID++, null);
    }, t.prototype.openPool = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 6, , 7 ]), e = this, [ 4, dm_1.dmdb.parseUrl(this._connectString) ];

                  case 1:
                    e._connectProps = t.sent(), this.newMiddlewares(), unenumerable_1.setUnenumerable(this, [ "middlewares", "nextID" ]), 
                    r = 0, t.label = 2;

                  case 2:
                    return r < this._poolMin ? [ 4, connection_1.PoolConnection.newInstanceOfPoolConnection(this._connectProps, this).openConnection(this._connectProps.getAny("epGroup")) ] : [ 3, 5 ];

                  case 3:
                    i = t.sent(), this._freeConnections.push(i), t.label = 4;

                  case 4:
                    return r++, [ 3, 2 ];

                  case 5:
                    return [ 2, this ];

                  case 6:
                    throw n = t.sent(), this.do_close(), n;

                  case 7:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.getConnection = function(e) {
        if (void 0 === e) return this.middlewares.valid ? this.middlewares.reset().pool_getConnection(this) : this.do_getConnection();
        if ("function" != typeof e) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getConnection().then(function(t) {
            e(null, t);
        }).catch(function(t) {
            e(t, void 0);
        });
    }, t.prototype.close = function(e, r) {
        if (void 0 === e && void 0 === r) return this.middlewares.valid ? this.middlewares.reset().pool_close(this) : this.do_close();
        if ("function" == typeof e && void 0 === r) this.close().then(function() {
            e(null);
        }).catch(function(t) {
            e(t);
        }); else {
            if ("number" == typeof e && void 0 === r) return this.middlewares.valid ? this.middlewares.reset().pool_close(this, e) : this.do_close(e);
            if ("number" != typeof e || "function" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
            this.close(e).then(function() {
                r(null);
            }).catch(function(t) {
                r(t);
            });
        }
    }, t.prototype.terminate = function(e, r) {
        if (void 0 === e && void 0 === r) return this.close();
        if ("function" == typeof e && void 0 === r) this.close().then(function() {
            e(null);
        }).catch(function(t) {
            e(t);
        }); else {
            if ("number" == typeof e && void 0 === r) return this.close(e);
            if ("number" != typeof e || "function" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
            this.close(e).then(function() {
                r(null);
            }).catch(function(t) {
                r(t);
            });
        }
    }, t.prototype.do_getConnection = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, u = this;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return this._checkPoolStatus(), [ 4, this.getFreeConnection() ];

                  case 1:
                    return (e = t.sent()) ? [ 2, e ] : [ 4, this.makeNewConnection() ];

                  case 2:
                    if (r = t.sent()) return [ 2, r ];
                    if (this._queueRequests && 0 !== this._queueMax) {
                        if (this._queueMax < 0 || this._connectionQueue.length < this._queueMax) return [ 2, new Promise(function(e, r) {
                            var i, n = new events_1.EventEmitter();
                            u._connectionQueue.push(n), 0 < u._queueTimeout && (i = setTimeout(function() {
                                u._connectionQueue.delete(n), r(error_1.DBError.ECJS_POOL_QUEUE_TIMEOUT());
                            }, u._queueTimeout)), n.on("turn", function(t) {
                                clearTimeout(i), process.nextTick(function() {
                                    e(t);
                                });
                            }), n.on("error", function(t) {
                                clearTimeout(i), u._connectionQueue.delete(n), process.nextTick(function() {
                                    r(t);
                                });
                            });
                        }) ];
                        throw error_1.DBError.ECJS_POOL_QUEUE_FULL();
                    }
                    throw error_1.DBError.ECJS_POOL_NO_MORE_CONNECTION();
                }
            });
        });
    }, t.prototype.do_close = function(t) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    this._checkPoolStatus(), this._status = dm_1.dmdb.POOL_STATUS_DRAINING, t.label = 1;

                  case 1:
                    t.trys.push([ 1, 7, , 8 ]), e = void 0, t.label = 2;

                  case 2:
                    return (e = this._freeConnections.shift()) ? (e.stopTimer(), [ 4, e._innerFree() ]) : [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 2 ];

                  case 4:
                    return (e = this._acquiringConnections.shift()) ? (e.stopTimer(), [ 4, e._innerFree() ]) : [ 3, 6 ];

                  case 5:
                    return t.sent(), [ 3, 4 ];

                  case 6:
                    for (this._status = dm_1.dmdb.POOL_STATUS_CLOSED, r = void 0; r = this._connectionQueue.shift(); ) r.emit("error", error_1.DBError.ECJS_POOL_CLOSED());
                    return this.middlewares.log && this.middlewares.log.logWriter.end(), this._poolAlias && dm_1.dmdb.pools.delete(this._poolAlias), 
                    [ 3, 8 ];

                  case 7:
                    throw i = t.sent(), this._status = dm_1.dmdb.POOL_STATUS_OPEN, i;

                  case 8:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.releaseConnection = function(t) {
        if (this._status !== dm_1.dmdb.POOL_STATUS_CLOSED && this._acquiringConnections.delete(t)) {
            if (0 < this._connectionQueue.length) return this._acquiringConnections.push(t), 
            void this._connectionQueue.shift().emit("turn", t);
            this._freeConnections.push(t), t.startTimer();
        }
    }, t.prototype.eliminateConnection = function(e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return this.connectionsOpen <= this._poolMin || this._acquiringConnections.includes(e) ? e.refreshTimer() : (this._freeConnections.delete(e), 
                e._innerFree()), [ 2 ];
            });
        });
    }, t.prototype.testConnection = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 6, , 7 ]), [ 4, n.do_execute(this._validationQuery, [], {
                        resultSet: !0
                    }) ];

                  case 1:
                    return (e = t.sent()).resultSet ? [ 4, e.resultSet.do_getRow() ] : [ 3, 4 ];

                  case 2:
                    return r = t.sent(), [ 4, e.resultSet.do_close() ];

                  case 3:
                    return t.sent(), r ? [ 2, !0 ] : (this.middlewares.logWarn("连接池连接有效性检查失败:没有结果集."), 
                    [ 2, !1 ]);

                  case 4:
                    return this.middlewares.logWarn("连接池连接有效性检查失败:没有结果集."), [ 2, !1 ];

                  case 5:
                    return [ 3, 7 ];

                  case 6:
                    return i = t.sent(), this.middlewares.logWarn("连接池连接有效性检查失败.\n" + i), [ 2, !1 ];

                  case 7:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.getFreeConnection = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!(e = this._freeConnections.shift())) return [ 3, 11 ];
                    e.stopTimer(), this._acquiringConnections.push(e), t.label = 1;

                  case 1:
                    return (t.trys.push([ 1, 10, , 11 ]), r = e.closed) ? [ 3, 4 ] : (i = this._testOnBorrow) ? [ 4, this.testConnection(e) ] : [ 3, 3 ];

                  case 2:
                    i = !t.sent(), t.label = 3;

                  case 3:
                    r = i, t.label = 4;

                  case 4:
                    return r ? (this._acquiringConnections.delete(e), e._innerFree(), this.connectionsOpen >= this._poolMin ? [ 4, this.getFreeConnection() ] : [ 3, 6 ]) : [ 3, 9 ];

                  case 5:
                    return n = t.sent(), [ 3, 8 ];

                  case 6:
                    return [ 4, this.makeNewConnection() ];

                  case 7:
                    n = t.sent(), t.label = 8;

                  case 8:
                    return [ 2, n ];

                  case 9:
                    return [ 3, 11 ];

                  case 10:
                    return t.sent(), this._acquiringConnections.delete(e), [ 2, void 0 ];

                  case 11:
                    return [ 2, e ];
                }
            });
        });
    }, t.prototype.makeNewConnection = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!(this.connectionsOpen < this._poolMax)) return [ 3, 5 ];
                    e = null, t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, , 4 ]), e = connection_1.PoolConnection.newInstanceOfPoolConnection(this._connectProps, this), 
                    this._acquiringConnections.push(e), [ 4, e.openConnection(this._connectProps.getAny("epGroup")) ];

                  case 2:
                    return t.sent(), [ 2, e ];

                  case 3:
                    return t.sent(), e && this._acquiringConnections.delete(e), [ 2, void 0 ];

                  case 4:
                    return [ 3, 6 ];

                  case 5:
                    return [ 2, void 0 ];

                  case 6:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype._checkPoolStatus = function() {
        if (this._status === dm_1.dmdb.POOL_STATUS_CLOSED) throw error_1.DBError.ECJS_POOL_CLOSED();
        if (this._status === dm_1.dmdb.POOL_STATUS_DRAINING) throw error_1.DBError.ECJS_POOL_DRAINING();
    }, Object.defineProperty(t.prototype, "connectionsInUse", {
        get: function() {
            return this._acquiringConnections.length;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "connectionsOpen", {
        get: function() {
            return this._acquiringConnections.length + this._freeConnections.length;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "connectionsInQueue", {
        get: function() {
            return this._connectionQueue.length;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "poolMax", {
        get: function() {
            return this._poolMax;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "poolMin", {
        get: function() {
            return this._poolMin;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "status", {
        get: function() {
            return this._status;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "poolAlias", {
        get: function() {
            return this._poolAlias;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "poolTimeout", {
        get: function() {
            return this._poolTimeout;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "queueMax", {
        get: function() {
            return this._queueMax;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "queueRequests", {
        get: function() {
            return this._queueRequests;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "queueTimeout", {
        get: function() {
            return this._queueTimeout;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "testOnBorrow", {
        get: function() {
            return this._testOnBorrow;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "validationQuery", {
        get: function() {
            return this._validationQuery;
        },
        enumerable: !0,
        configurable: !0
    }), t.nextID = 0, t;
}();

exports.Pool = Pool;
