// dmdb/src/driver/queryStream.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, n) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, n) {
            t.__proto__ = n;
        } || function(t, n) {
            for (var e in n) n.hasOwnProperty(e) && (t[e] = n[e]);
        })(t, n);
    };
    return function(t, n) {
        function e() {
            this.constructor = t;
        }
        r(t, n), t.prototype = null === n ? Object.create(n) : (e.prototype = n.prototype, 
        new e());
    };
}(), __awaiter = this && this.__awaiter || function(t, u, o, c) {
    return new (o = o || Promise)(function(e, n) {
        function r(t) {
            try {
                s(c.next(t));
            } catch (t) {
                n(t);
            }
        }
        function i(t) {
            try {
                s(c.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function s(t) {
            var n;
            t.done ? e(t.value) : ((n = t.value) instanceof o ? n : new o(function(t) {
                t(n);
            })).then(r, i);
        }
        s((c = c.apply(t, u || [])).next());
    });
}, __generator = this && this.__generator || function(e, r) {
    var i, s, u, o = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (i = 1, s && (u = 2 & n[0] ? s.return : n[0] ? s.throw || ((u = s.return) && u.call(s), 
                    0) : s.next) && !(u = u.call(s, n[1])).done) return u;
                    switch (s = 0, u && (n = [ 2 & n[0], u.value ]), n[0]) {
                      case 0:
                      case 1:
                        u = n;
                        break;

                      case 4:
                        return o.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        o.label++, s = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = o.trys).length && u[u.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === n[0] && (!u || n[1] > u[0] && n[1] < u[3])) {
                            o.label = n[1];
                            break;
                        }
                        if (6 === n[0] && o.label < u[1]) {
                            o.label = u[1], u = n;
                            break;
                        }
                        if (u && o.label < u[2]) {
                            o.label = u[2], o.ops.push(n);
                            break;
                        }
                        u[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    n = r.call(e, o);
                } catch (t) {
                    n = [ 6, t ], s = 0;
                } finally {
                    i = u = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var stream_1 = require("stream"), QueryStream = function(e) {
    function t(t) {
        var n = e.call(this, {
            objectMode: !0
        }) || this;
        return n.fetching = !1, t && process.nextTick(function() {
            n._open(t);
        }), n;
    }
    return __extends(t, e), t.prototype._open = function(t) {
        this.rs = t, this.emit("open"), this.emit("metadata", this.rs.metaData);
    }, t.prototype._destroy = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, r = this;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return this.rs ? (n = this.rs, this.rs = void 0, this.fetching ? [ 4, new Promise(function(t) {
                        return r.once("doneFetching", t);
                    }) ] : [ 3, 2 ]) : [ 3, 6 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return t.trys.push([ 2, 4, 5, 6 ]), [ 4, n.do_close() ];

                  case 3:
                    return t.sent(), [ 3, 6 ];

                  case 4:
                    return e = t.sent(), s(e), [ 2 ];

                  case 5:
                    return this.rs = void 0, [ 7 ];

                  case 6:
                    return s(i), [ 2 ];
                }
            });
        });
    }, t.prototype._read = function() {
        return __awaiter(this, void 0, void 0, function() {
            var n, e;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!this.rs) return this.once("open", this._read), [ 2 ];
                    t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, 4, 5 ]), this.fetching = !0, this.rs._allowGetRowCall = !0, 
                    [ 4, this.rs.getRow() ];

                  case 2:
                    return n = t.sent(), this.push(n || null), [ 3, 5 ];

                  case 3:
                    return e = t.sent(), this.destroy(e), [ 3, 5 ];

                  case 4:
                    return this.fetching = !1, this.rs ? this.rs._allowGetRowCall = !1 : this.emit("doneFetching"), 
                    [ 7 ];

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, t;
}(stream_1.Readable);

exports.QueryStream = QueryStream;
