// dmdb/src/driver/resultSet.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, h, o, u) {
    return new (o = o || Promise)(function(r, i) {
        function e(t) {
            try {
                s(u.next(t));
            } catch (t) {
                i(t);
            }
        }
        function n(t) {
            try {
                s(u.throw(t));
            } catch (t) {
                i(t);
            }
        }
        function s(t) {
            var i;
            t.done ? r(t.value) : ((i = t.value) instanceof o ? i : new o(function(t) {
                t(i);
            })).then(e, n);
        }
        s((u = u.apply(t, h || [])).next());
    });
}, __generator = this && this.__generator || function(r, e) {
    var n, s, h, o = {
        label: 0,
        sent: function() {
            if (1 & h[0]) throw h[1];
            return h[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: i(0),
        throw: i(1),
        return: i(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function i(i) {
        return function(t) {
            return function(i) {
                if (n) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (n = 1, s && (h = 2 & i[0] ? s.return : i[0] ? s.throw || ((h = s.return) && h.call(s), 
                    0) : s.next) && !(h = h.call(s, i[1])).done) return h;
                    switch (s = 0, h && (i = [ 2 & i[0], h.value ]), i[0]) {
                      case 0:
                      case 1:
                        h = i;
                        break;

                      case 4:
                        return o.label++, {
                            value: i[1],
                            done: !1
                        };

                      case 5:
                        o.label++, s = i[1], i = [ 0 ];
                        continue;

                      case 7:
                        i = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(h = 0 < (h = o.trys).length && h[h.length - 1]) && (6 === i[0] || 2 === i[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === i[0] && (!h || i[1] > h[0] && i[1] < h[3])) {
                            o.label = i[1];
                            break;
                        }
                        if (6 === i[0] && o.label < h[1]) {
                            o.label = h[1], h = i;
                            break;
                        }
                        if (h && o.label < h[2]) {
                            o.label = h[2], o.ops.push(i);
                            break;
                        }
                        h[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    i = e.call(r, o);
                } catch (t) {
                    i = [ 6, t ], s = 0;
                } finally {
                    n = h = 0;
                }
                if (5 & i[0]) throw i[1];
                return {
                    value: i[0] ? i[1] : void 0,
                    done: !0
                };
            }([ i, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var statement_1 = require("./statement"), error_1 = require("./error"), db2j_1 = require("../convert/db2j"), middleware_1 = require("../middleware/middleware"), unenumerable_1 = require("../utils/unenumerable"), const_1 = require("../desc/const"), dm_1 = require("../dm"), queryStream_1 = require("./queryStream"), ResultSet = function() {
    function t(t, i, r, e) {
        var n;
        this.datas = new Array(0), this.datasOffset = -1, this.datasStartPos = 0, this.currentPos = -1, 
        this.totalRowCount = BigInt(0), this.updatable = !1, this.bdta = !1, this.wasNull = !1, 
        this.closed = !1, this.fetchSize = 0, this._inUsing = !1, this._convertedToStream = !1, 
        this._allowGetRowCall = !1, this._locked = !1, this.metaData = r.result.metaData, 
        "number" == typeof t && i instanceof statement_1.Statement ? (this.constructor_inner(i, i.columns, r.rsDatas), 
        this.id = t, this.totalRowCount = r.updateCount, this.updatable = r.rsUpdatable, 
        this.bdta = r.rsBdta, n = this.datas.length, this.sizeOfRow = 0 < n ? r.rsSizeof / n : 0, 
        this.updatable) : t instanceof statement_1.Statement && i instanceof Array && (e instanceof Array || void 0 === e) && this.constructor_inner(t, i, e);
    }
    return t.prototype.newMiddlewares = function() {
        this.middlewares = new middleware_1.Middlewares(this.stmt.connection, null, t.nextID++, this.stmt.connection.middlewares);
    }, t.prototype.close = function(i) {
        if (void 0 === i) {
            if (this.middlewares.valid) return this.middlewares.reset().resultSet_close(this);
            if (this._convertedToStream) throw error_1.DBError.ECJS_RESULTSET_CONVERTED();
            return this._inUsing = !0, this.do_close();
        }
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.close().then(function() {
            i(null);
        }).catch(function(t) {
            i(t);
        });
    }, t.prototype.getRow = function(i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().resultSet_getRow(this) : this.do_getRow();
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getRow().then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, void 0);
        });
    }, t.prototype.getRows = function(t, i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().resultSet_getRows(this, t) : this.do_getRows(t);
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getRows(t).then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, void 0);
        });
    }, t.prototype.getRowCount = function(i) {
        if (void 0 === i) return this.middlewares.valid ? this.middlewares.reset().resultSet_getRowCount(this) : this.do_getRowCount();
        if ("function" != typeof i) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getRowCount().then(function(t) {
            i(null, t);
        }).catch(function(t) {
            i(t, -1);
        });
    }, t.prototype.toQueryStream = function() {
        if (this._inUsing) throw error_1.DBError.ECJS_RESULTSET_IN_USING();
        if (this._convertedToStream) throw error_1.DBError.ECJS_RESULTSET_CONVERTED();
        return this._convertedToStream = !0, new queryStream_1.QueryStream(this);
    }, t.prototype.do_close = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return delete this.conn, this.stmt ? [ 4, this.stmt.close() ] : [ 3, 2 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return delete this.stmt, delete this.columns, delete this.datas, this.closed = !0, 
                    [ 2 ];
                }
            });
        });
    }, t.prototype.do_getRow = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this._convertedToStream && !this._allowGetRowCall) throw error_1.DBError.ECJS_RESULTSET_CONVERTED();
                    return this._allowGetRowCall = !1, this._inUsing = !0, [ 4, this.next() ];

                  case 1:
                    return [ 2, t.sent() ? this.getRowData() : void 0 ];
                }
            });
        });
    }, t.prototype.do_getRows = function(s) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, e, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!Number.isSafeInteger(s)) throw error_1.DBError.ECJS_INVALID_PARAM();
                    if (this._convertedToStream) throw error_1.DBError.ECJS_RESULTSET_CONVERTED();
                    this._inUsing = !0, i = new Array(0), t.label = 1;

                  case 1:
                    return (r = 0 < s--) ? [ 4, this.next() ] : [ 3, 3 ];

                  case 2:
                    r = t.sent(), t.label = 3;

                  case 3:
                    return r ? (n = (e = i).push, [ 4, this.getRowData() ]) : [ 3, 5 ];

                  case 4:
                    return n.apply(e, [ t.sent() ]), [ 3, 1 ];

                  case 5:
                    return [ 2, i ];
                }
            });
        });
    }, t.prototype.do_getRowCount = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return (this.checkClosed(), this.totalRowCount !== const_1.Const.INT64_MAX) ? [ 3, 2 ] : [ 4, (i = this).conn.access.J(this) ];

                  case 1:
                    i.totalRowCount = t.sent(), t.label = 2;

                  case 2:
                    return [ 2, Number(this.totalRowCount) ];
                }
            });
        });
    }, t.prototype.checkClosed = function() {
        if (this.closed) throw error_1.DBError.ECJS_RESULTSET_CLOSED();
    }, t.prototype.constructor_inner = function(t, i, r) {
        this.id = 0, this.stmt = t, this.conn = t.connection, this.columns = i || new Array(0), 
        this.fetchInfo = t.execOpts.fetchInfo, this.datas = r || new Array(0), this.totalRowCount = void 0 === r ? BigInt(0) : BigInt(r.length), 
        this.updatable = !1, this.bdta = !1, this.newMiddlewares(), unenumerable_1.setUnenumerable(this, [ "middlewares", "nextID", "_locked" ]);
    }, t.prototype.next = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return this.checkClosed(), this.totalRowCount === BigInt(0) || BigInt(this.currentPos) >= this.totalRowCount ? [ 2, !1 ] : BigInt(this.currentPos) === this.totalRowCount - BigInt(1) ? (this.currentPos++, 
                    this.datasOffset++, [ 2, !1 ]) : this.currentPos + 1 < this.datasStartPos || this.currentPos + 1 >= this.datasStartPos + this.datas.length ? [ 4, this.fetchData(this.currentPos + 1) ] : [ 3, 2 ];

                  case 1:
                    return t.sent() ? (this.currentPos++, [ 2, !0 ]) : (this.currentPos++, this.datasOffset++, 
                    [ 2, !1 ]);

                  case 2:
                    return this.datasOffset++, this.currentPos++, [ 2, !0 ];

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.fetchData = function(r) {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.conn.access.K(this, r) ];

                  case 1:
                    return i = t.sent(), this.totalRowCount = i.updateCount, i.rsDatas ? (this.datas = i.rsDatas, 
                    this.datasStartPos = r, [ 2, !(this.datasOffset = 0) ]) : [ 2, !1 ];
                }
            });
        });
    }, t.prototype.getRowData = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, e, n, s, h;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.stmt.execOpts.outFormat !== dm_1.dmdb.OUT_FORMAT_OBJECT) return [ 3, 5 ];
                    e = new Object(), n = 0, t.label = 1;

                  case 1:
                    return n < this.columns.length ? (i = e, r = this.metaData[n].name, [ 4, this.getColumnData(n) ]) : [ 3, 4 ];

                  case 2:
                    i[r] = t.sent(), t.label = 3;

                  case 3:
                    return n++, [ 3, 1 ];

                  case 4:
                    return [ 2, e ];

                  case 5:
                    e = new Array(this.columns.length), n = 0, t.label = 6;

                  case 6:
                    return n < this.columns.length ? (s = e, h = n, [ 4, this.getColumnData(n) ]) : [ 3, 9 ];

                  case 7:
                    s[h] = t.sent(), t.label = 8;

                  case 8:
                    return n++, [ 3, 6 ];

                  case 9:
                    return [ 2, e ];
                }
            });
        });
    }, t.prototype.getColumnData = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            var i, r, e;
            return __generator(this, function(t) {
                return i = this.datas[this.datasOffset][n + 1], r = this.columns[n], e = this.fetchInfo && this.fetchInfo[r.name] ? this.fetchInfo[r.name].type : void 0, 
                [ 2, db2j_1.DB2J.toNaturalJS(i, r, this.conn, e) ];
            });
        });
    }, t.nextID = 0, t;
}();

exports.ResultSet = ResultSet;
