// dmdb/src/driver/statement.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, e) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, e) {
            t.__proto__ = e;
        } || function(t, e) {
            for (var r in e) e.hasOwnProperty(r) && (t[r] = e[r]);
        })(t, e);
    };
    return function(t, e) {
        function r() {
            this.constructor = t;
        }
        i(t, e), t.prototype = null === e ? Object.create(e) : (r.prototype = e.prototype, 
        new r());
    };
}(), __awaiter = this && this.__awaiter || function(t, h, a, c) {
    return new (a = a || Promise)(function(r, e) {
        function i(t) {
            try {
                n(c.next(t));
            } catch (t) {
                e(t);
            }
        }
        function s(t) {
            try {
                n(c.throw(t));
            } catch (t) {
                e(t);
            }
        }
        function n(t) {
            var e;
            t.done ? r(t.value) : ((e = t.value) instanceof a ? e : new a(function(t) {
                t(e);
            })).then(i, s);
        }
        n((c = c.apply(t, h || [])).next());
    });
}, __generator = this && this.__generator || function(r, i) {
    var s, n, h, a = {
        label: 0,
        sent: function() {
            if (1 & h[0]) throw h[1];
            return h[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: e(0),
        throw: e(1),
        return: e(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function e(e) {
        return function(t) {
            return function(e) {
                if (s) throw new TypeError("Generator is already executing.");
                for (;a; ) try {
                    if (s = 1, n && (h = 2 & e[0] ? n.return : e[0] ? n.throw || ((h = n.return) && h.call(n), 
                    0) : n.next) && !(h = h.call(n, e[1])).done) return h;
                    switch (n = 0, h && (e = [ 2 & e[0], h.value ]), e[0]) {
                      case 0:
                      case 1:
                        h = e;
                        break;

                      case 4:
                        return a.label++, {
                            value: e[1],
                            done: !1
                        };

                      case 5:
                        a.label++, n = e[1], e = [ 0 ];
                        continue;

                      case 7:
                        e = a.ops.pop(), a.trys.pop();
                        continue;

                      default:
                        if (!(h = 0 < (h = a.trys).length && h[h.length - 1]) && (6 === e[0] || 2 === e[0])) {
                            a = 0;
                            continue;
                        }
                        if (3 === e[0] && (!h || e[1] > h[0] && e[1] < h[3])) {
                            a.label = e[1];
                            break;
                        }
                        if (6 === e[0] && a.label < h[1]) {
                            a.label = h[1], h = e;
                            break;
                        }
                        if (h && a.label < h[2]) {
                            a.label = h[2], a.ops.push(e);
                            break;
                        }
                        h[2] && a.ops.pop(), a.trys.pop();
                        continue;
                    }
                    e = i.call(r, a);
                } catch (t) {
                    e = [ 6, t ], n = 0;
                } finally {
                    s = h = 0;
                }
                if (5 & e[0]) throw e[1];
                return {
                    value: e[0] ? e[1] : void 0,
                    done: !0
                };
            }([ e, t ]);
        };
    }
}, __read = this && this.__read || function(t, e) {
    var r = "function" == typeof Symbol && t[Symbol.iterator];
    if (!r) return t;
    var i, s, n = r.call(t), h = [];
    try {
        for (;(void 0 === e || 0 < e--) && !(i = n.next()).done; ) h.push(i.value);
    } catch (t) {
        s = {
            error: t
        };
    } finally {
        try {
            i && !i.done && (r = n.return) && r.call(n);
        } finally {
            if (s) throw s.error;
        }
    }
    return h;
}, __values = this && this.__values || function(t) {
    var e = "function" == typeof Symbol && Symbol.iterator, r = e && t[e], i = 0;
    if (r) return r.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && i >= t.length && (t = void 0), {
                value: t && t[i++],
                done: !t
            };
        }
    };
    throw new TypeError(e ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var executeRetInfo_1 = require("../desc/executeRetInfo"), error_1 = require("./error"), parameter_1 = require("../desc/parameter"), sqlProcessor_1 = require("../parser/sqlProcessor"), resultSet_1 = require("./resultSet"), dm_1 = require("../dm"), stmtInfo_1 = require("../desc/stmtInfo"), resultSetCachePool_1 = require("./resultSetCachePool"), dbtype_1 = require("./dbtype"), j2db_1 = require("../convert/j2db"), lob_1 = require("./lob"), db2j_1 = require("../convert/db2j"), stream_1 = require("stream"), const_1 = require("../desc/const"), driverUtil_1 = require("../utils/driverUtil"), Statement = function() {
    function s(t, e, r) {
        var i;
        this.handle = 0, this.cursorUpdateRow = 0, this.closeOnCompletion = !1, this.closed = !0, 
        this.poolable = !0, this.columns = new Array(0), this.running = !1, this.isBindArray = !0, 
        this.isExecuteMany = !1, this.multiRowsError = !1, this.type = s.STATEMENT, this.connection = t, 
        this.escapeProcessing = t.conn_prop_escapeProcess, r || (i = e.maxRows), void 0 === i && (i = dm_1.dmdb.maxRows), 
        void 0 === i && (i = t.conn_prop_maxRows), this.maxRows = BigInt(i), this.queryTimeout = t.conn_prop_sessionTimeout, 
        this.execOpts = e, this.innerExec = !r && e.innerExec || !1, this.isExecuteMany = r;
    }
    return s.prototype.execute = function(e, t) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.executeInner(e, 0) ];
            });
        });
    }, s.prototype.allocateHandle = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return (e = this.connection.stmtPool && 0 < this.connection.stmtPool.size() ? this.connection.stmtPool.get() : void 0) ? (this.handle = e.handle, 
                    this.cursorName = e.cursorName, this.readBaseColName = e.readBaseColName, [ 3, 3 ]) : [ 3, 1 ];

                  case 1:
                    return [ 4, this.connection.access.tt(this) ];

                  case 2:
                    t.sent(), t.label = 3;

                  case 3:
                    return this.closed = !1, [ 2 ];
                }
            });
        });
    }, s.prototype.close = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return !this.closed && this.poolable && 0 < this.connection.conn_prop_stmtPoolSize ? [ 2, this.connection.stmtPool.put(new stmtInfo_1.StmtInfo(this.handle, this.cursorName, this.readBaseColName)) ] : [ 2, this.free() ];
            });
        });
    }, s.prototype.free = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.closed || !this.connection || !this.connection.access || this.connection.closed) return [ 3, 4 ];
                    t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, , 4 ]), [ 4, this.connection.access.P(this.handle) ];

                  case 2:
                    return t.sent(), [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 4 ];

                  case 4:
                    return this.closed = !0, [ 2 ];
                }
            });
        });
    }, s.prototype.getMetaData = function(r) {
        var i = this, s = new Array(this.columns.length);
        return this.columns.forEach(function(t, e) {
            s[e] = {
                name: i.columnNameCaseCvt(i.connection.conn_prop_columnNameCase, t.name)
            }, r && (s[e].dbType = t.type, s[e].dbTypeName = t.typeName, s[e].precision = t.prec, 
            s[e].scale = t.scale, s[e].nullable = t.nullable);
        }), s;
    }, s.prototype.executeInner = function(c, u) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, s, n, h, a;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!c) throw error_1.DBError.ECJS_SQL_IS_EMPTY();
                    if (e = c, this.escapeProcessing) try {
                        e = sqlProcessor_1.SQLProcessor.escape(e, this.connection.conn_prop_keyWords);
                    } catch (t) {}
                    if (this.nativeSql = e, r = null, this.connection.constParaOpt) {
                        r = new Array();
                        try {
                            this.nativeSql = sqlProcessor_1.SQLProcessor.execOpt(this.nativeSql, r, this.connection.getServerEncoding(), this.connection.backSlashFlag);
                        } catch (t) {}
                    }
                    return this.connection.conn_prop_enRsCache ? (i = resultSetCachePool_1.ResultSetCachePool.get(this.connection.guid || "", this.connection.schema || "", this.nativeSql, this)) ? (this.execInfo = i.execInfo, 
                    this.connection.lastExecInfo = i.execInfo, [ 4, this.genResult(i.getResultSet(this)) ]) : [ 3, 2 ] : [ 3, 2 ];

                  case 1:
                    return t.sent(), [ 2, this.execInfo.result ];

                  case 2:
                    if (!(r && 0 < r.length)) return [ 3, 8 ];
                    t.label = 3;

                  case 3:
                    return t.trys.push([ 3, 5, , 7 ]), [ 4, (s = this).connection.access.et(this, r) ];

                  case 4:
                    return s.execInfo = t.sent(), [ 3, 7 ];

                  case 5:
                    if ((n = t.sent()) instanceof error_1.DBError && (n.errCode === error_1.DBError.ECJS_COMMUNICATION_ERROR().errCode || n.errCode === error_1.DBError.EC_RN_OPERATION_CANCELED().errCode || n.errCode === error_1.DBError.EC_EXCEED_SESSION_REQUEST_LIMIT().errCode)) throw n;
                    return this.nativeSql = e, [ 4, (h = this).connection.access.rt(this, u) ];

                  case 6:
                    return h.execInfo = t.sent(), [ 3, 7 ];

                  case 7:
                    return [ 3, 10 ];

                  case 8:
                    return [ 4, (a = this).connection.access.rt(this, u) ];

                  case 9:
                    a.execInfo = t.sent(), t.label = 10;

                  case 10:
                    return this.connection.lastExecInfo = this.execInfo, this.execInfo.hasResultSet ? [ 4, this.genResult() ] : [ 3, 12 ];

                  case 11:
                    return t.sent(), [ 3, 13 ];

                  case 12:
                    this.execInfo.result = new executeRetInfo_1.Result(this.connection), this.execInfo.result.rowsAffected = Number(this.execInfo.updateCount), 
                    t.label = 13;

                  case 13:
                    return [ 2, this.execInfo.result ];
                }
            });
        });
    }, s.prototype.genResult = function(e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return this.isExecuteMany ? [ 3, 3 ] : [ 4, this.loadResult(this.execInfo, e) ];

                  case 1:
                    return t.sent(), [ 4, this.loadMoreResult(this.execInfo) ];

                  case 2:
                    t.sent(), !e && this.connection.conn_prop_enRsCache && 0 < this.execInfo.rsCacheOffset && this.execInfo.rsDatas && this.execInfo.rsDatas.length === Number(this.execInfo.updateCount) && resultSetCachePool_1.ResultSetCachePool.put(this, this.nativeSql, this.execInfo), 
                    t.label = 3;

                  case 3:
                    return [ 2 ];
                }
            });
        });
    }, s.prototype.loadResult = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return (i.result = new executeRetInfo_1.Result(this.connection), i.result.metaData = this.getMetaData(this.execOpts.extendedMetaData), 
                    e = s || new resultSet_1.ResultSet(1, this, i), this.execOpts.resultSet) ? (delete i.result.rows, 
                    i.result.resultSet = e, [ 3, 6 ]) : [ 3, 1 ];

                  case 1:
                    return delete i.result.resultSet, i.result.rows = new Array(), [ 4, e.getRow() ];

                  case 2:
                    r = t.sent(), t.label = 3;

                  case 3:
                    if (void 0 === r) return [ 3, 6 ];
                    i.result.rows.push(r), t.label = 4;

                  case 4:
                    return [ 4, e.getRow() ];

                  case 5:
                    return r = t.sent(), [ 3, 3 ];

                  case 6:
                    return [ 2 ];
                }
            });
        });
    }, s.prototype.loadMoreResult = function(s) {
        var n;
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return r = e = void 0, [ 4, this.connection.access.it(this, 0) ];

                  case 1:
                    i = t.sent(), t.label = 2;

                  case 2:
                    return i.hasResultSet ? [ 4, this.loadResult(i) ] : [ 3, 5 ];

                  case 3:
                    return t.sent(), void 0 === e && (e = []), void 0 === r && (r = []), e.push(i.result.rows || i.result.resultSet), 
                    r.push(i.result.metaData), [ 4, this.connection.access.it(this, 0) ];

                  case 4:
                    return i = t.sent(), [ 3, 2 ];

                  case 5:
                    return void 0 !== e && void 0 !== r && (e.unshift(s.result.rows || (null === (n = s.result) || void 0 === n ? void 0 : n.resultSet)), 
                    r.unshift(s.result.metaData), delete s.result.rows, delete s.result.resultSet, delete s.result.metaData, 
                    s.result.implicitResults = e, s.result.metaData = r), [ 2 ];
                }
            });
        });
    }, s.prototype.columnNameCaseCvt = function(t, e) {
        return t === const_1.Const.COLUMN_NAME_NATURAL_CASE ? e : t === const_1.Const.COLUMN_NAME_UPPER_CASE ? e.toUpperCase() : t === const_1.Const.COLUMN_NAME_LOWER_CASE ? e.toLowerCase() : e;
    }, s.STATEMENT = 1, s.PREPARED_STATEMENT = 2, s.CALLABLE_STATEMENT = 3, s;
}(), PreparedStatement = function(s) {
    function t(t, e, r) {
        var i = s.call(this, t, e, r) || this;
        return i.serverParams = new Array(0), i.bindParams = new Array(0), i.paramCount = 0, 
        i.rowIndex = 0, i.preExec = !1, i.type = Statement.PREPARED_STATEMENT, i.paramObjectsList = new Array(), 
        i.curRowObjects = new Array(i.paramCount), i;
    }
    return __extends(t, s), t.prototype.prepareSql = function(i, s) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (!i) throw error_1.DBError.ECJS_SQL_IS_EMPTY();
                    if (this.nativeSql = i, this.escapeProcessing) try {
                        this.nativeSql = sqlProcessor_1.SQLProcessor.escape(i, this.connection.conn_prop_keyWords);
                    } catch (t) {}
                    if (this.connection.conn_prop_injectArray && s) try {
                        r = __read(sqlProcessor_1.SQLProcessor.arrayConstant(i, s), 2), this.nativeSql = r[0], 
                        s = r[1];
                    } catch (t) {}
                    return [ 4, this.connection.access.st(this) ];

                  case 1:
                    return e = t.sent(), this.execInfo = e, [ 2, s ];
                }
            });
        });
    }, t.prototype.executePstmt = function(s) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this._encodeArgs(s, !0) ];

                  case 1:
                    return e = t.sent(), [ 4, (r = this).connection.access.nt(this, e, !1) ];

                  case 2:
                    return r.execInfo = t.sent(), this.connection.lastExecInfo = this.execInfo, this.execInfo.hasResultSet ? [ 4, this.genResult() ] : [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 5 ];

                  case 4:
                    this.execInfo.result = new executeRetInfo_1.Result(this.connection), this.execInfo.result.rowsAffected = Number(this.execInfo.updateCount), 
                    t.label = 5;

                  case 5:
                    return this.execInfo.outParamDatas && 0 < this.execInfo.outParamDatas.length ? (this.execInfo.result || (this.execInfo.result = new executeRetInfo_1.Result(this.connection)), 
                    i = this.execInfo.result, [ 4, this.genOutParamDatas1() ]) : [ 3, 7 ];

                  case 6:
                    i.outBinds = t.sent(), t.label = 7;

                  case 7:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.executeManyPstmt = function(b) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, s, n, h, a, c, u, o, _, f, d;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (e = new Array(b.length), !(0 < this.paramCount)) return [ 3, 4 ];
                    r = 0, t.label = 1;

                  case 1:
                    return r < b.length ? (i = b[r], s = e, n = r, [ 4, this._encodeArgs(i, 0 === r) ]) : [ 3, 4 ];

                  case 2:
                    s[n] = t.sent(), t.label = 3;

                  case 3:
                    return r++, [ 3, 1 ];

                  case 4:
                    if (this.execInfo.retSqlType === const_1.Const.SQL_TYPE_DML_SELECT || this.execInfo.retSqlType === const_1.Const.SQL_TYPE_DML_CURSOR_OPEN) throw error_1.DBError.ECJS_INVALID_SQL_TYPE();
                    t.label = 5;

                  case 5:
                    return (t.trys.push([ 5, 12, 21, 22 ]), 0 !== e.length) ? [ 3, 7 ] : [ 4, (h = this).connection.access.ht(this, []) ];

                  case 6:
                    return h.execInfo = t.sent(), [ 3, 11 ];

                  case 7:
                    return 1 === e.length || 2 === this.connection.conn_prop_batchType || this.connection.conn_prop_batchNotOnCall && this.execInfo.retSqlType === const_1.Const.SQL_TYPE_DML_CALL ? [ 4, (a = this).connection.access.ct(this, e, this.preExec) ] : [ 3, 9 ];

                  case 8:
                    return a.execInfo = t.sent(), [ 3, 11 ];

                  case 9:
                    return [ 4, (c = this).connection.access.ut(this, e, this.preExec) ];

                  case 10:
                    c.execInfo = t.sent(), t.label = 11;

                  case 11:
                    return [ 3, 22 ];

                  case 12:
                    return (u = t.sent()) instanceof error_1.DBError && u.errCode === error_1.DBError.EC_SRC_MULTI_ROWS().errCode && !this.multiRowsError && this.execInfo.retSqlType !== const_1.Const.SQL_TYPE_DML_MERGE ? (this.multiRowsError = !0, 
                    0 !== e.length ? [ 3, 14 ] : [ 4, (o = this).connection.access.ht(this, []) ]) : [ 3, 19 ];

                  case 13:
                    return o.execInfo = t.sent(), [ 3, 18 ];

                  case 14:
                    return 1 === e.length || 2 === this.connection.conn_prop_batchType || this.connection.conn_prop_batchNotOnCall && this.execInfo.retSqlType === const_1.Const.SQL_TYPE_DML_CALL ? [ 4, (_ = this).connection.access.ct(this, e, this.preExec) ] : [ 3, 16 ];

                  case 15:
                    return _.execInfo = t.sent(), [ 3, 18 ];

                  case 16:
                    return [ 4, (f = this).connection.access.ut(this, e, this.preExec) ];

                  case 17:
                    f.execInfo = t.sent(), t.label = 18;

                  case 18:
                    return [ 3, 20 ];

                  case 19:
                    throw u;

                  case 20:
                    return [ 3, 22 ];

                  case 21:
                    return this.multiRowsError = !1, [ 7 ];

                  case 22:
                    return (this.execInfo.results = new executeRetInfo_1.Results(this.connection), this.execInfo.results.batchErrors = this.execInfo.batchErrors, 
                    this.execOpts.dmlRowCounts && (this.execInfo.results.dmlRowCounts = this.execInfo.updateCounts), 
                    this.execInfo.updateCounts && 0 != this.execInfo.updateCounts.length ? this.execInfo.results.rowsAffected = this.execInfo.updateCounts.reduce(function(t, e) {
                        return (t < 0 ? 0 : t) + (e < 0 ? 0 : e);
                    }) : this.execInfo.results.rowsAffected = 0, this.execInfo.outParamDatas && 0 < this.execInfo.outParamDatas.length) ? (d = this.execInfo.results, 
                    [ 4, this.genOutParamDatas2() ]) : [ 3, 24 ];

                  case 23:
                    d.outBinds = t.sent(), t.label = 24;

                  case 24:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype._bindInParam = function(t, e, r) {
        var i, s, n;
        r && (i = e === dbtype_1.DBType.NULL, s = this.serverParams[t], n = this.bindParams[t], 
        s ? (n.name = s.name, n.typeDescriptor = s.typeDescriptor, n.mask = s.mask, n.typeFlag = s.typeFlag, 
        s.type != dbtype_1.DBType.UNKNOWN && (i || s.typeFlag === parameter_1.Parameter.TYPE_FLAG_EXACT) || 0 !== s.mask ? (n.type = s.type, 
        n.prec = s.prec, n.scale = s.scale, n.mask = s.mask) : n.resetType(e)) : n.resetType(e), 
        n.ioType = n.ioType === parameter_1.Parameter.IO_TYPE_OUT || n.ioType === parameter_1.Parameter.IO_TYPE_INOUT ? parameter_1.Parameter.IO_TYPE_INOUT : parameter_1.Parameter.IO_TYPE_IN, 
        n._bindName = s._bindName);
    }, t.prototype._checkBindParameters = function(t) {
        for (var e = 0; e < this.paramCount; e++) {
            if (this.bindParams[e].ioType === parameter_1.Parameter.IO_TYPE_UNKNOWN) {
                if (!this.serverParams[e] || this.serverParams[e].ioType !== parameter_1.Parameter.IO_TYPE_OUT) throw error_1.DBError.ECJS_UNBINDED_PARAMETER();
                t[e] = null;
            }
            this.bindParams[e].type !== dbtype_1.DBType.CURSOR ? this.serverParams[e] && this.serverParams[e].ioType !== this.bindParams[e].ioType && (this.bindParams[e].ioType = this.serverParams[e].ioType) : this.bindParams[e].ioType = parameter_1.Parameter.IO_TYPE_INOUT;
        }
        for (e = 0; e < this.paramCount; e++) switch (this.bindParams[e].type) {
          case dbtype_1.DBType.CHAR:
          case dbtype_1.DBType.VARCHAR:
          case dbtype_1.DBType.VARCHAR2:
            var r = t[e] instanceof Buffer ? t[e].length : -1;
            if (r > dbtype_1.DBType.VARCHAR_PREC) throw error_1.DBError.ECJS_STRING_CUT();
            r > this.bindParams[e].prec && (r < dbtype_1.DBType.VARCHAR_PREC / 4 ? this.bindParams[e].prec = dbtype_1.DBType.VARCHAR_PREC / 4 : r < dbtype_1.DBType.VARCHAR_PREC / 2 ? this.bindParams[e].prec = dbtype_1.DBType.VARCHAR_PREC / 2 : r < 3 * dbtype_1.DBType.VARCHAR_PREC / 4 ? this.bindParams[e].prec = 3 * dbtype_1.DBType.VARCHAR_PREC / 4 : this.bindParams[e].prec = dbtype_1.DBType.VARCHAR_PREC);
        }
    }, t.prototype._encodeArgs = function(y, l) {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, s, n, h, a, c, u, o, _, f, d, b, v;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    e = new Array(y.length), r = 0, t.label = 1;

                  case 1:
                    return r < this.serverParams.length ? (i = y[r], this.bindParams[r].outJType = i.type, 
                    this.serverParams[r].type !== dbtype_1.DBType.CURSOR ? [ 3, 4 ] : (this._bindInParam(r, dbtype_1.DBType.CURSOR, l), 
                    this.bindParams[r].cursorStmt ? [ 3, 3 ] : (this.bindParams[r].cursorStmt = new Statement(this.connection, this.execOpts, !1), 
                    [ 4, this.bindParams[r].cursorStmt.allocateHandle() ]))) : [ 3, 24 ];

                  case 2:
                    t.sent(), t.label = 3;

                  case 3:
                    return this.bindParams[r].ioType = parameter_1.Parameter.IO_TYPE_INOUT, [ 3, 23 ];

                  case 4:
                    return null != i && void 0 !== i.val && null !== i.val ? [ 3, 5 ] : (s = i ? driverUtil_1.DriverUtil.toRealBindType(i.type) : dbtype_1.DBType.NULL, 
                    this._bindInParam(r, s, l), e[r] = null, [ 3, 23 ]);

                  case 5:
                    if ("number" != typeof i.val) return [ 3, 6 ];
                    if (Number.isNaN(i.val) || !Number.isFinite(i.val)) throw error_1.DBError.ECJS_INVALID_NUMBER(i.val.toString());
                    return Number.isInteger(i.val) ? (this._bindInParam(r, dbtype_1.DBType.BIGINT, l), 
                    e[r] = j2db_1.J2DB.fromBigInt(BigInt(i.val), this.bindParams[r], this.connection)) : (this._bindInParam(r, dbtype_1.DBType.DOUBLE, l), 
                    e[r] = j2db_1.J2DB.fromNumber(i.val, this.bindParams[r], this.connection)), [ 3, 23 ];

                  case 6:
                    return "bigint" != typeof i.val ? [ 3, 7 ] : (this._bindInParam(r, dbtype_1.DBType.BIGINT, l), 
                    e[r] = j2db_1.J2DB.fromBigInt(i.val, this.bindParams[r], this.connection), [ 3, 23 ]);

                  case 7:
                    return "string" != typeof i.val ? [ 3, 11 ] : "" === i.val && dbtype_1.DBType.emptyStringToNull(this.serverParams[r].type) ? (this._bindInParam(r, dbtype_1.DBType.NULL, l), 
                    e[r] = null, [ 3, 10 ]) : [ 3, 8 ];

                  case 8:
                    return a = i.val.length >= dbtype_1.DBType.VARCHAR_PREC ? dbtype_1.DBType.CLOB : dbtype_1.DBType.VARCHAR, 
                    this._bindInParam(r, a, l), n = e, h = r, [ 4, j2db_1.J2DB.fromString(i.val, this.bindParams[r], this.connection) ];

                  case 9:
                    n[h] = t.sent(), t.label = 10;

                  case 10:
                    return [ 3, 23 ];

                  case 11:
                    return "boolean" != typeof i.val ? [ 3, 12 ] : (this._bindInParam(r, dbtype_1.DBType.TINYINT, l), 
                    e[r] = j2db_1.J2DB.fromBoolean(i.val, this.bindParams[r], this.connection), [ 3, 23 ]);

                  case 12:
                    return i.val instanceof Buffer ? (a = i.val.length >= dbtype_1.DBType.VARBINARY_PREC ? dbtype_1.DBType.BLOB : dbtype_1.DBType.VARBINARY, 
                    this._bindInParam(r, a, l), c = e, u = r, [ 4, j2db_1.J2DB.fromBuffer(i.val, this.bindParams[r], this.connection) ]) : [ 3, 14 ];

                  case 13:
                    return c[u] = t.sent(), [ 3, 23 ];

                  case 14:
                    return i.val instanceof lob_1.Lob ? i.val.type !== dm_1.dmdb.BLOB ? [ 3, 16 ] : (this._bindInParam(r, dbtype_1.DBType.BLOB, l), 
                    o = e, _ = r, [ 4, j2db_1.J2DB.fromBlob(i.val.iLob, this.bindParams[r], this.connection) ]) : [ 3, 19 ];

                  case 15:
                    return o[_] = t.sent(), [ 3, 18 ];

                  case 16:
                    return this._bindInParam(r, dbtype_1.DBType.CLOB, l), f = e, d = r, [ 4, j2db_1.J2DB.fromClob(i.val.iLob, this.bindParams[r], this.connection) ];

                  case 17:
                    f[d] = t.sent(), t.label = 18;

                  case 18:
                    return [ 3, 23 ];

                  case 19:
                    return i.val instanceof Date ? (this._bindInParam(r, dbtype_1.DBType.DATETIME, l), 
                    e[r] = j2db_1.J2DB.fromDate(i.val, this.bindParams[r], this.connection), [ 3, 23 ]) : [ 3, 20 ];

                  case 20:
                    return i.val instanceof stream_1.Readable ? (this._bindInParam(r, this.serverParams[r].type, l), 
                    b = e, v = r, [ 4, j2db_1.J2DB.fromReadable(i.val, this.bindParams[r], this.connection) ]) : [ 3, 22 ];

                  case 21:
                    return b[v] = t.sent(), [ 3, 23 ];

                  case 22:
                    throw error_1.DBError.ECJS_UNSUPPORTED_TYPE();

                  case 23:
                    return r++, [ 3, 1 ];

                  case 24:
                    return this._checkBindParameters(e), [ 2, e ];
                }
            });
        });
    }, t.prototype.genOutParamDatas1 = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, s, n, h, a, c, u, o, _, f;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    e = dm_1.dmdb.outBindFormat === dm_1.dmdb.OUT_FORMAT_AUTO ? !this.isBindArray : dm_1.dmdb.outBindFormat === dm_1.dmdb.OUT_FORMAT_OBJECT, 
                    r = e ? new Object() : new Array(0), i = 0, t.label = 1;

                  case 1:
                    if (!(i < this.execInfo.outParamDatas.length)) return [ 3, 9 ];
                    s = this.execInfo.outParamDatas[i], h = n = 0, t.label = 2;

                  case 2:
                    return h < s.length ? (a = s[h], this.bindParams[h].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.bindParams[h].ioType !== parameter_1.Parameter.IO_TYPE_INOUT ? [ 3, 7 ] : this.bindParams[h].type !== dbtype_1.DBType.CURSOR ? [ 3, 5 ] : (c = this.bindParams[h].cursorStmt, 
                    [ 4, this.connection.access.it(c, 1) ])) : [ 3, 8 ];

                  case 3:
                    return (u = t.sent()).hasResultSet ? [ 4, c.loadResult(u) ] : [ 3, 5 ];

                  case 4:
                    return t.sent(), o = u.result.rows || u.result.resultSet, e ? (f = void 0 !== this.bindParams[h]._bindName ? this.bindParams[h]._bindName : this.bindParams[h].name, 
                    0 === i ? r[f] = o : (1 === i && (r[f] = [ r[f] ]), r[f].push(o))) : (0 === i ? r.push(o) : (1 === i && (r[n] = [ r[n] ]), 
                    r[n].push(o)), n++), [ 3, 7 ];

                  case 5:
                    return [ 4, db2j_1.DB2J.toNaturalJS4Bind(a, this.bindParams[h], this.connection, this.bindParams[h].outJType) ];

                  case 6:
                    _ = t.sent(), e ? (f = void 0 !== this.bindParams[h]._bindName ? this.bindParams[h]._bindName : this.bindParams[h].name, 
                    0 === i ? r[f] = _ : (1 === i && (r[f] = [ r[f] ]), r[f].push(_))) : (0 === i ? r.push(_) : (1 === i && (r[n] = [ r[n] ]), 
                    r[n].push(_)), n++), t.label = 7;

                  case 7:
                    return h++, [ 3, 2 ];

                  case 8:
                    return i++, [ 3, 1 ];

                  case 9:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.genOutParamDatas2 = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e, r, i, s, n, h, a, c, u, o, _, f, d, b, v;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    e = dm_1.dmdb.outBindFormat === dm_1.dmdb.OUT_FORMAT_AUTO ? !this.isBindArray : dm_1.dmdb.outBindFormat === dm_1.dmdb.OUT_FORMAT_OBJECT, 
                    r = new Array(), t.label = 1;

                  case 1:
                    t.trys.push([ 1, 12, 13, 14 ]), i = __values(this.execInfo.outParamDatas), s = i.next(), 
                    t.label = 2;

                  case 2:
                    if (s.done) return [ 3, 11 ];
                    n = s.value, h = e ? new Object() : new Array(0), a = 0, t.label = 3;

                  case 3:
                    return a < n.length ? (c = n[a], this.bindParams[a].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.bindParams[a].ioType !== parameter_1.Parameter.IO_TYPE_INOUT ? [ 3, 8 ] : this.bindParams[a].type !== dbtype_1.DBType.CURSOR ? [ 3, 6 ] : (u = this.bindParams[a].cursorStmt, 
                    [ 4, this.connection.access.it(u, 1) ])) : [ 3, 9 ];

                  case 4:
                    return (o = t.sent()).hasResultSet ? [ 4, u.loadResult(o) ] : [ 3, 6 ];

                  case 5:
                    return t.sent(), _ = o.result.rows || o.result.resultSet, e ? h[void 0 !== this.bindParams[a]._bindName ? this.bindParams[a]._bindName : this.bindParams[a].name] = _ : h.push(_), 
                    [ 3, 8 ];

                  case 6:
                    return [ 4, db2j_1.DB2J.toNaturalJS4Bind(c, this.bindParams[a], this.connection, this.bindParams[a].outJType) ];

                  case 7:
                    f = t.sent(), e ? h[void 0 !== this.bindParams[a]._bindName ? this.bindParams[a]._bindName : this.bindParams[a].name] = f : h.push(f), 
                    t.label = 8;

                  case 8:
                    return a++, [ 3, 3 ];

                  case 9:
                    r.push(h), t.label = 10;

                  case 10:
                    return s = i.next(), [ 3, 2 ];

                  case 11:
                    return [ 3, 14 ];

                  case 12:
                    return d = t.sent(), b = {
                        error: d
                    }, [ 3, 14 ];

                  case 13:
                    try {
                        s && !s.done && (v = i.return) && v.call(i);
                    } finally {
                        if (b) throw b.error;
                    }
                    return [ 7 ];

                  case 14:
                    return [ 2, r ];
                }
            });
        });
    }, t;
}(exports.Statement = Statement);

exports.PreparedStatement = PreparedStatement;
