// dmdb/src/middleware/rw/rwCounter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var middleware_1 = require("../middleware"), const_1 = require("../../desc/const"), RWCounter = function() {
    function n(t, i) {
        this.standbyCount = 0, this.primaryIncrement = 0, this.standbyIncrement = 0, this.primaryPercent = 0, 
        this.standbyPercent = 0, this.standbyChooseIndex = 0, this.reset(t, i);
    }
    return n.prototype.reset = function(t, i) {
        this.standbyCount = i;
        var s = new Array(i + 1), h = this.gcd(t * i, 100 - t);
        s[0] = t * i / h, s.fill((100 - t) / h, 1, s.length), this.primaryIncrement = s[0], 
        this.standbyIncrement = 0 < i ? s[1] : 0, 0 < i ? (this.primaryPercent = t / 100, 
        this.standbyPercent = (100 - t) / 100 / i) : (this.primaryPercent = 1, this.standbyPercent = 0), 
        this.primaryCounter = new EPCounter(), this.primaryCounter.balance = this.primaryIncrement, 
        this.standbysCounter = new EPCounter(), this.standbysCounter.balance = this.standbyIncrement * i, 
        this.standbyMap = new Map();
    }, n.getInstance = function(t, i) {
        var s = t.conn_prop_host + "_" + t.conn_prop_port + "_" + t.conn_prop_rwPercent, h = this.rwMap.get(s);
        return h ? h.standbyCount !== i && h.reset(t.conn_prop_rwPercent, i) : (h = new n(t.conn_prop_rwPercent, i), 
        this.rwMap.set(s, h)), h;
    }, n.prototype.countPrimary = function() {
        return this.adjustNtrx(), this.primaryCounter.incrementNtrx(), middleware_1.RWSiteEnum.PRIMARY;
    }, n.prototype.countStandby = function() {
        return this.adjustNtrx(), this.standbysCounter.incrementNtrx(), middleware_1.RWSiteEnum.STANDBY;
    }, n.prototype.count = function(t, i) {
        switch (this.adjustNtrx(), t) {
          case middleware_1.RWSiteEnum.ANY:
            t = 1 === this.primaryPercent || this.primaryCounter.balance > this.getStandbyCounter(i).balance && this.primaryCounter.balance > this.standbysCounter.balance ? (this.primaryCounter.incrementNtrx(), 
            middleware_1.RWSiteEnum.PRIMARY) : (this.getStandbyCounter(i).incrementNtrx(), this.standbysCounter.incrementNtrx(), 
            middleware_1.RWSiteEnum.STANDBY);
            break;

          case middleware_1.RWSiteEnum.STANDBY:
            this.getStandbyCounter(i).incrementNtrx(), this.standbysCounter.incrementNtrx();
            break;

          case middleware_1.RWSiteEnum.PRIMARY:
            this.primaryCounter.incrementNtrx();
            break;

          default:
            throw new Error("Invalid RWSite!");
        }
        return t;
    }, n.prototype.adjustNtrx = function() {
        var i, s, h = this;
        this.primaryCounter.ntrx + this.standbysCounter.ntrx >= const_1.Const.INT64_MAX && (i = 0, 
        this.standbyMap.forEach(function(t) {
            (0 === i || t.ntrx < i) && (i = t.ntrx);
        }), i = i < this.primaryCounter.ntrx ? i : this.primaryCounter.ntrx, this.primaryCounter.ntrx /= i, 
        s = 0, this.standbyMap.forEach(function(t) {
            t.ntrx /= i, s += t.ntrx;
        }), this.standbysCounter.ntrx = s), this.primaryCounter.balance <= 0 && this.standbysCounter.balance <= 0 && (this.primaryCounter.balance += this.primaryIncrement, 
        this.standbyMap.forEach(function(t) {
            t.balance += h.standbyIncrement;
        }));
    }, n.prototype.getStandbyCounter = function(t) {
        var i = t.conn_prop_host + ":" + t.conn_prop_port, s = this.standbyMap.get(i);
        return s || ((s = new EPCounter()).id = i, s.balance = this.standbyIncrement, this.standbyMap.set(i, s)), 
        s;
    }, n.prototype.chooseStandbyIndex = function() {
        return this.standbyCount <= 0 ? 0 : Math.abs(this.standbyChooseIndex++) % this.standbyCount;
    }, n.prototype.toString = function() {
        return "PERCENT(P/S) : " + this.primaryPercent + "/" + this.standbyPercent + "\nNTRX_PRIMARY : " + this.primaryCounter.ntrx + "\nNTRX_STANDBY : " + this.standbysCounter.ntrx;
    }, n.prototype.gcd = function(t, i) {
        if (!Number.isSafeInteger(t) || !Number.isSafeInteger(i)) throw new Error(t + " or " + i + " is not safe integer number: ");
        if (0 === t || 0 === i) return 1;
        var s = t % i;
        return t = i, 0 == (i = s) ? t : this.gcd(t, i);
    }, n.rwMap = new Map(), n;
}();

exports.RWCounter = RWCounter;

var EPCounter = function() {
    function t() {
        this.id = "", this.ntrx = 0, this.balance = 0, this.isStandby = !1;
    }
    return t.prototype.incrementNtrx = function() {
        this.ntrx++, this.balance--;
    }, t;
}();
