// dmdb/src/middleware/rw/rwMiddleware.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, o, c, a) {
    return new (c = c || Promise)(function(r, n) {
        function e(t) {
            try {
                u(a.next(t));
            } catch (t) {
                n(t);
            }
        }
        function i(t) {
            try {
                u(a.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function u(t) {
            var n;
            t.done ? r(t.value) : ((n = t.value) instanceof c ? n : new c(function(t) {
                t(n);
            })).then(e, i);
        }
        u((a = a.apply(t, o || [])).next());
    });
}, __generator = this && this.__generator || function(r, e) {
    var i, u, o, c = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (i = 1, u && (o = 2 & n[0] ? u.return : n[0] ? u.throw || ((o = u.return) && o.call(u), 
                    0) : u.next) && !(o = o.call(u, n[1])).done) return o;
                    switch (u = 0, o && (n = [ 2 & n[0], o.value ]), n[0]) {
                      case 0:
                      case 1:
                        o = n;
                        break;

                      case 4:
                        return c.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        c.label++, u = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = c.trys).length && o[o.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === n[0] && (!o || n[1] > o[0] && n[1] < o[3])) {
                            c.label = n[1];
                            break;
                        }
                        if (6 === n[0] && c.label < o[1]) {
                            c.label = o[1], o = n;
                            break;
                        }
                        if (o && c.label < o[2]) {
                            c.label = o[2], c.ops.push(n);
                            break;
                        }
                        o[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    n = e.call(r, c);
                } catch (t) {
                    n = [ 6, t ], u = 0;
                } finally {
                    i = o = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var connection_1 = require("../../driver/connection"), rwUtil_1 = require("./rwUtil"), rwCounter_1 = require("./rwCounter"), const_1 = require("../../desc/const"), rwUtil2_1 = require("./rwUtil2"), RWMiddleware = function() {
    function n(t) {
        this.rwUtil = t;
    }
    return n.getInstance = function(t) {
        switch (t) {
          case const_1.Const.RW_SEPARATE_OFF:
          case const_1.Const.RW_SEPARATE_CLIENT:
          case const_1.Const.RW_SEPARATE_EP_GROUP:
          case const_1.Const.RW_SEPARATE_USER_DEFINED:
            return new n(new rwUtil2_1.RWUtil2());

          default:
            return new n(new rwUtil_1.RWUtil());
        }
    }, n.prototype.connection_openConnection = function(t, n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.rwUtil.connect(n, r) ];
            });
        });
    }, n.prototype.connection_execute = function(t, n, r, e, i) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.rwUtil.execute(n, r, function() {
                    return n.middlewares.rwInfo.connCurrent.do_execute(r, e, i);
                }, function(t) {
                    return t.do_execute(r, e, i);
                }) ];
            });
        });
    }, n.prototype.connection_executeMany = function(t, n, r, e, i) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.rwUtil.execute(n, r, function() {
                    return n.middlewares.rwInfo.connCurrent.do_executeMany(r, e, i);
                }, function(t) {
                    return t.do_executeMany(r, e, i);
                }) ];
            });
        });
    }, n.prototype.connection_createLob = function(n, r, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.connection_createLob(r, e) ];
            });
        });
    }, n.prototype.connection_close = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 3, , 4 ]), !this.rwUtil.isStandbyAlive(r) || r instanceof connection_1.PoolConnection ? [ 3, 2 ] : [ 4, r.middlewares.rwInfo.connStandby.do_close() ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 4 ];

                  case 4:
                    return [ 2, n.connection_close(r) ];
                }
            });
        });
    }, n.prototype.connection_free = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 3, , 4 ]), this.rwUtil.isStandbyAlive(r) ? [ 4, r.middlewares.rwInfo.connStandby.do_close() ] : [ 3, 2 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return [ 3, 4 ];

                  case 3:
                    return t.sent(), [ 3, 4 ];

                  case 4:
                    return [ 2, n.connection_free(r) ];
                }
            });
        });
    }, n.prototype.connection_commit = function(r, e) {
        return __awaiter(this, void 0, void 0, function() {
            var n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 3, , 5 ]), this.rwUtil.isStandbyAlive(e) ? [ 4, e.middlewares.rwInfo.connStandby.do_commit() ] : [ 3, 2 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return [ 3, 5 ];

                  case 3:
                    return n = t.sent(), [ 4, this.rwUtil.afterExceptionOnStandby(e, n) ];

                  case 4:
                    return t.sent(), [ 3, 5 ];

                  case 5:
                    return [ 2, r.connection_commit(e) ];
                }
            });
        });
    }, n.prototype.connection_getStatementInfo = function(n, r, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.connection_getStatementInfo(r, e) ];
            });
        });
    }, n.prototype.connection_rollback = function(r, e) {
        return __awaiter(this, void 0, void 0, function() {
            var n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return t.trys.push([ 0, 3, , 5 ]), this.rwUtil.isStandbyAlive(e) ? [ 4, e.middlewares.rwInfo.connStandby.do_rollback() ] : [ 3, 2 ];

                  case 1:
                    t.sent(), t.label = 2;

                  case 2:
                    return [ 3, 5 ];

                  case 3:
                    return n = t.sent(), [ 4, this.rwUtil.afterExceptionOnStandby(e, n) ];

                  case 4:
                    return t.sent(), [ 3, 5 ];

                  case 5:
                    return [ 2, r.connection_rollback(e) ];
                }
            });
        });
    }, n.prototype.lob_close = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.lob_close(r) ];
            });
        });
    }, n.prototype.lob_getData = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.lob_getData(r) ];
            });
        });
    }, n.prototype.lob_getLength = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.lob_getLength(r) ];
            });
        });
    }, n.prototype.pool_getConnection = function(r, e) {
        return __awaiter(this, void 0, void 0, function() {
            var n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, r.pool_getConnection(e) ];

                  case 1:
                    return n = t.sent(), r.rwInfo.rwCounter = rwCounter_1.RWCounter.getInstance(n, n.standbyCount), 
                    [ 2, n ];
                }
            });
        });
    }, n.prototype.pool_close = function(n, r, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.pool_close(r, e) ];
            });
        });
    }, n.prototype.resultSet_close = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.resultSet_close(r) ];
            });
        });
    }, n.prototype.resultSet_getRow = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.resultSet_getRow(r) ];
            });
        });
    }, n.prototype.resultSet_getRows = function(n, r, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.resultSet_getRows(r, e) ];
            });
        });
    }, n.prototype.resultSet_getRowCount = function(n, r) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, n.resultSet_getRowCount(r) ];
            });
        });
    }, n;
}();

exports.RWMiddleware = RWMiddleware;
