// dmdb/src/middleware/rw/rwUtil2.js
"use strict";

var __extends = this && this.__extends || function() {
    var t = function(r, n) {
        return (t = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(r, n) {
            r.__proto__ = n;
        } || function(r, n) {
            for (var e in n) n.hasOwnProperty(e) && (r[e] = n[e]);
        })(r, n);
    };
    return function(r, n) {
        function e() {
            this.constructor = r;
        }
        t(r, n), r.prototype = null === n ? Object.create(n) : (e.prototype = n.prototype, 
        new e());
    };
}(), __awaiter = this && this.__awaiter || function(r, a, s, o) {
    return new (s = s || Promise)(function(e, n) {
        function t(r) {
            try {
                c(o.next(r));
            } catch (r) {
                n(r);
            }
        }
        function i(r) {
            try {
                c(o.throw(r));
            } catch (r) {
                n(r);
            }
        }
        function c(r) {
            var n;
            r.done ? e(r.value) : ((n = r.value) instanceof s ? n : new s(function(r) {
                r(n);
            })).then(t, i);
        }
        c((o = o.apply(r, a || [])).next());
    });
}, __generator = this && this.__generator || function(e, t) {
    var i, c, a, s = {
        label: 0,
        sent: function() {
            if (1 & a[0]) throw a[1];
            return a[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function n(n) {
        return function(r) {
            return function(n) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (i = 1, c && (a = 2 & n[0] ? c.return : n[0] ? c.throw || ((a = c.return) && a.call(c), 
                    0) : c.next) && !(a = a.call(c, n[1])).done) return a;
                    switch (c = 0, a && (n = [ 2 & n[0], a.value ]), n[0]) {
                      case 0:
                      case 1:
                        a = n;
                        break;

                      case 4:
                        return s.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        s.label++, c = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(a = 0 < (a = s.trys).length && a[a.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === n[0] && (!a || n[1] > a[0] && n[1] < a[3])) {
                            s.label = n[1];
                            break;
                        }
                        if (6 === n[0] && s.label < a[1]) {
                            s.label = a[1], a = n;
                            break;
                        }
                        if (a && s.label < a[2]) {
                            s.label = a[2], s.ops.push(n);
                            break;
                        }
                        a[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    n = t.call(e, s);
                } catch (r) {
                    n = [ 6, r ], c = 0;
                } finally {
                    i = a = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, r ]);
        };
    }
}, __values = this && this.__values || function(r) {
    var n = "function" == typeof Symbol && Symbol.iterator, e = n && r[n], t = 0;
    if (e) return e.call(r);
    if (r && "number" == typeof r.length) return {
        next: function() {
            return r && t >= r.length && (r = void 0), {
                value: r && r[t++],
                done: !r
            };
        }
    };
    throw new TypeError(n ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var connection_1 = require("../../driver/connection"), rwCounter_1 = require("./rwCounter"), ep_1 = require("../../desc/ep"), properties_1 = require("../../desc/properties"), const_1 = require("../../desc/const"), rwUtil_1 = require("./rwUtil"), RWUtil2 = function(r) {
    function I() {
        return null !== r && r.apply(this, arguments) || this;
    }
    return __extends(I, r), I.prototype.connectStandby = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return [ 4, this.chooseValidStandby(i) ];

                  case 1:
                    if (!(n = r.sent())) return [ 2 ];
                    r.label = 2;

                  case 2:
                    return r.trys.push([ 2, 6, , 7 ]), (e = new properties_1.Properties(i.props)).set("host", n.host), 
                    e.set("port", n.port.toString()), e.set("rwStandby", "true"), e.setAny("epGroup", void 0), 
                    e.set("loginMode", const_1.Const.LOGIN_MODE_STANDBY_ONLY.toString()), e.set("switchTimes", "0"), 
                    t = i.middlewares.rwInfo, [ 4, connection_1.Connection.newInstanceOfConnection(e).do_openConnection() ];

                  case 3:
                    return t.connStandby = r.sent(), i.middlewares.rwInfo.connStandby.svrMode === const_1.Const.SERVER_MODE_STANDBY && i.middlewares.rwInfo.connStandby.svrStat === const_1.Const.SERVER_STATUS_OPEN ? [ 3, 5 ] : [ 4, this.removeStandby(i) ];

                  case 4:
                    r.sent(), r.label = 5;

                  case 5:
                    return [ 3, 7 ];

                  case 6:
                    return r.sent(), [ 3, 7 ];

                  case 7:
                    return [ 2 ];
                }
            });
        });
    }, I.prototype.chooseValidStandby = function(l) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t, i, c, a, s, o, u, _, f, h;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (l.conn_prop_rwSeparate === const_1.Const.RW_SEPARATE_USER_DEFINED) return [ 2, this.chooseStandbyUserDefined(l) ];
                    e = n = null, i = t = "", r.label = 1;

                  case 1:
                    return r.trys.push([ 1, 3, , 8 ]), l.conn_prop_rwSeparate === const_1.Const.RW_SEPARATE_EP_GROUP && 0 < (c = l.epGroup.epList.map(function(r) {
                        return "'" + r.host + ":" + r.port + "'";
                    }).join(",")).length && (t = " and (mailIni.INST_IP || ':'|| mailIni.INST_PORT) in (" + c + ")", 
                    i = " and (mailIni.mal_INST_HOST || ':'|| mailIni.mal_INST_PORT) in (" + c + ")"), 
                    [ 4, l.do_executeInner(l.malini2 ? I.SQL_SELECT_STANDBY2 + i : I.SQL_SELECT_STANDBY + t) ];

                  case 2:
                    return n = r.sent(), [ 3, 8 ];

                  case 3:
                    r.sent(), r.label = 4;

                  case 4:
                    return r.trys.push([ 4, 6, , 7 ]), [ 4, l.do_executeInner(l.malini2 ? I.SQL_SELECT_STANDBY + t : I.SQL_SELECT_STANDBY2 + i) ];

                  case 5:
                    return n = r.sent(), [ 3, 7 ];

                  case 6:
                    return a = r.sent(), e = a, [ 3, 7 ];

                  case 7:
                    return [ 3, 8 ];

                  case 8:
                    return r.trys.push([ 8, 17, 18, 20 ]), n ? [ 4, n.resultSet.do_getRowCount() ] : [ 3, 10 ];

                  case 9:
                    return o = r.sent(), [ 3, 11 ];

                  case 10:
                    o = 0, r.label = 11;

                  case 11:
                    if (!(0 < (s = o))) return [ 3, 16 ];
                    l.middlewares.rwInfo.rwCounter = rwCounter_1.RWCounter.getInstance(l, s), u = 0, 
                    _ = l.middlewares.rwInfo.rwCounter.chooseStandbyIndex(), r.label = 12;

                  case 12:
                    return [ 4, n.resultSet.next() ];

                  case 13:
                    return r.sent() ? u !== _ ? [ 3, 15 ] : [ 4, n.resultSet.getRowData() ] : [ 3, 16 ];

                  case 14:
                    return f = r.sent(), [ 2, new ep_1.EP(f[1], f[2]) ];

                  case 15:
                    return u++, [ 3, 12 ];

                  case 16:
                    return [ 3, 20 ];

                  case 17:
                    return h = r.sent(), e = h, [ 3, 20 ];

                  case 18:
                    return [ 4, n.resultSet.do_close() ];

                  case 19:
                    return r.sent(), [ 7 ];

                  case 20:
                    if (e) throw new Error("choose valid standby error!" + e.message);
                    return [ 2, void 0 ];
                }
            });
        });
    }, I.prototype.chooseStandbyUserDefined = function(o) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t, i, c, a, s;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    n = [], r.label = 1;

                  case 1:
                    r.trys.push([ 1, 6, 7, 8 ]), e = __values(o.epGroup.epList), t = e.next(), r.label = 2;

                  case 2:
                    return t.done ? [ 3, 5 ] : (i = t.value, [ 4, this.isAliveStandby(o.epGroup, i) ]);

                  case 3:
                    r.sent() && n.push(i), r.label = 4;

                  case 4:
                    return t = e.next(), [ 3, 2 ];

                  case 5:
                    return [ 3, 8 ];

                  case 6:
                    return c = r.sent(), a = {
                        error: c
                    }, [ 3, 8 ];

                  case 7:
                    try {
                        t && !t.done && (s = e.return) && s.call(e);
                    } finally {
                        if (a) throw a.error;
                    }
                    return [ 7 ];

                  case 8:
                    return 0 < n.length ? (o.middlewares.rwInfo.rwCounter = rwCounter_1.RWCounter.getInstance(o, n.length), 
                    [ 2, n[o.middlewares.rwInfo.rwCounter.chooseStandbyIndex()] ]) : [ 2, void 0 ];
                }
            });
        });
    }, I.prototype.isAliveStandby = function(e, t) {
        return __awaiter(this, void 0, void 0, function() {
            var n;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return r.trys.push([ 0, 2, 3, 6 ]), n = connection_1.Connection.newInstanceOfConnection(e.props), 
                    [ 4, t.connect(n) ];

                  case 1:
                    return r.sent(), [ 2, n.svrMode === const_1.Const.SERVER_MODE_STANDBY ];

                  case 2:
                    return r.sent(), [ 3, 6 ];

                  case 3:
                    return n ? [ 4, n.close() ] : [ 3, 5 ];

                  case 4:
                    r.sent(), r.label = 5;

                  case 5:
                    return [ 7 ];

                  case 6:
                    return [ 2, !1 ];
                }
            });
        });
    }, I.SQL_SELECT_STANDBY = "select distinct mailIni.inst_name, mailIni.INST_IP, mailIni.INST_PORT, archIni.arch_status from  v$arch_status archIni left join (select * from V$DM_MAL_INI) mailIni on archIni.arch_dest = mailIni.inst_name left join V$MAL_LINK_STATUS on CTL_LINK_STATUS  = 'CONNECTED' AND DATA_LINK_STATUS = 'CONNECTED' where archIni.arch_type in ('TIMELY', 'REALTIME') AND  archIni.arch_status = 'VALID'", 
    I.SQL_SELECT_STANDBY2 = "select distinct mailIni.mal_inst_name, mailIni.mal_INST_HOST, mailIni.mal_INST_PORT, archIni.arch_status from v$arch_status archIni left join (select * from V$DM_MAL_INI) mailIni on archIni.arch_dest = mailIni.mal_inst_name left join V$MAL_LINK_STATUS on CTL_LINK_STATUS  = 'CONNECTED' AND DATA_LINK_STATUS = 'CONNECTED' where archIni.arch_type in ('TIMELY', 'REALTIME') AND  archIni.arch_status = 'VALID'", 
    I;
}(rwUtil_1.RWUtil);

exports.RWUtil2 = RWUtil2;
