// dmdb/src/middleware/rw/rwUtil.js
"use strict";

var __awaiter = this && this.__awaiter || function(r, s, u, o) {
    return new (u = u || Promise)(function(t, e) {
        function n(r) {
            try {
                c(o.next(r));
            } catch (r) {
                e(r);
            }
        }
        function i(r) {
            try {
                c(o.throw(r));
            } catch (r) {
                e(r);
            }
        }
        function c(r) {
            var e;
            r.done ? t(r.value) : ((e = r.value) instanceof u ? e : new u(function(r) {
                r(e);
            })).then(n, i);
        }
        c((o = o.apply(r, s || [])).next());
    });
}, __generator = this && this.__generator || function(t, n) {
    var i, c, s, u = {
        label: 0,
        sent: function() {
            if (1 & s[0]) throw s[1];
            return s[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: e(0),
        throw: e(1),
        return: e(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function e(e) {
        return function(r) {
            return function(e) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;u; ) try {
                    if (i = 1, c && (s = 2 & e[0] ? c.return : e[0] ? c.throw || ((s = c.return) && s.call(c), 
                    0) : c.next) && !(s = s.call(c, e[1])).done) return s;
                    switch (c = 0, s && (e = [ 2 & e[0], s.value ]), e[0]) {
                      case 0:
                      case 1:
                        s = e;
                        break;

                      case 4:
                        return u.label++, {
                            value: e[1],
                            done: !1
                        };

                      case 5:
                        u.label++, c = e[1], e = [ 0 ];
                        continue;

                      case 7:
                        e = u.ops.pop(), u.trys.pop();
                        continue;

                      default:
                        if (!(s = 0 < (s = u.trys).length && s[s.length - 1]) && (6 === e[0] || 2 === e[0])) {
                            u = 0;
                            continue;
                        }
                        if (3 === e[0] && (!s || e[1] > s[0] && e[1] < s[3])) {
                            u.label = e[1];
                            break;
                        }
                        if (6 === e[0] && u.label < s[1]) {
                            u.label = s[1], s = e;
                            break;
                        }
                        if (s && u.label < s[2]) {
                            u.label = s[2], u.ops.push(e);
                            break;
                        }
                        s[2] && u.ops.pop(), u.trys.pop();
                        continue;
                    }
                    e = n.call(t, u);
                } catch (r) {
                    e = [ 6, r ], c = 0;
                } finally {
                    i = s = 0;
                }
                if (5 & e[0]) throw e[1];
                return {
                    value: e[0] ? e[1] : void 0,
                    done: !0
                };
            }([ e, r ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var connection_1 = require("../../driver/connection"), rwCounter_1 = require("./rwCounter"), properties_1 = require("../../desc/properties"), const_1 = require("../../desc/const"), error_1 = require("../../driver/error"), middleware_1 = require("../middleware"), stringUtil_1 = require("../../utils/stringUtil"), RWUtil = function() {
    function r() {}
    return r.prototype.connect = function(e, t) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return e.props.set("loginMode", const_1.Const.LOGIN_MODE_PRIMARY_ONLY + ""), e.conn_prop_loginMode = const_1.Const.LOGIN_MODE_PRIMARY_ONLY, 
                    [ 4, e.do_openConnection(t) ];

                  case 1:
                    return (e = r.sent()).middlewares.rwInfo.rwCounter = rwCounter_1.RWCounter.getInstance(e, e.standbyCount), 
                    [ 4, this.connectStandby(e) ];

                  case 2:
                    return r.sent(), [ 2, e ];
                }
            });
        });
    }, r.prototype.reconnect = function(e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return e.middlewares.rwInfo ? [ 4, this.removeStandby(e) ] : [ 2 ];

                  case 1:
                    return r.sent(), [ 4, e.do_reconnect() ];

                  case 2:
                    return r.sent(), e.middlewares.rwInfo.cleanup(), e.middlewares.rwInfo.rwCounter = rwCounter_1.RWCounter.getInstance(e, e.standbyCount), 
                    [ 4, this.connectStandby(e) ];

                  case 3:
                    return r.sent(), [ 2 ];
                }
            });
        });
    }, r.prototype.recoverStandby = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            var e, t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return n.closed || this.isStandbyAlive(n) ? [ 2 ] : (e = Date.now(), (t = n.conn_prop_rwStandbyRecoverTime) <= 0 || e - n.middlewares.rwInfo.tryRecoverTs < t ? [ 2 ] : [ 4, this.connectStandby(n) ]);

                  case 1:
                    return r.sent(), [ 4, this.checkStatusValid(n) ];

                  case 2:
                    return r.sent() ? [ 3, 4 ] : [ 4, this.removeStandby(n) ];

                  case 3:
                    r.sent(), r.label = 4;

                  case 4:
                    return n.middlewares.rwInfo.tryRecoverTs = e, [ 2 ];
                }
            });
        });
    }, r.prototype.connectStandby = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            var e, t;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (!n.standbyHost || !n.standbyPort || n.standbyPort <= 0) return [ 2 ];
                    r.label = 1;

                  case 1:
                    return r.trys.push([ 1, 3, , 4 ]), (e = new properties_1.Properties(n.props)).set("host", n.standbyHost), 
                    e.set("port", n.standbyPort.toString()), e.set("rwStandby", "true"), e.setAny("epGroup", void 0), 
                    e.set("loginMode", const_1.Const.LOGIN_MODE_STANDBY_ONLY.toString()), e.set("switchTimes", "1"), 
                    t = n.middlewares.rwInfo, [ 4, connection_1.Connection.newInstanceOfConnection(e).do_openConnection() ];

                  case 2:
                    return t.connStandby = r.sent(), [ 3, 4 ];

                  case 3:
                    return r.sent(), [ 3, 4 ];

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.checkStatusValid = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var e, t, n;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (!(e = i.middlewares.rwInfo.connStandby) || e.closed) return [ 2, !1 ];
                    if (e.svrMode !== const_1.Const.SERVER_MODE_STANDBY || e.svrStat !== const_1.Const.SERVER_STATUS_OPEN) return [ 2, !1 ];
                    r.label = 1;

                  case 1:
                    return r.trys.push([ 1, 6, , 7 ]), [ 4, i.do_executeInner("select oguid from v$instance") ];

                  case 2:
                    return [ 4, r.sent().resultSet.do_getRow() ];

                  case 3:
                    return t = r.sent(), n = t[0], [ 4, e.do_executeInner("select oguid from v$instance") ];

                  case 4:
                    return [ 4, r.sent().resultSet.do_getRow() ];

                  case 5:
                    return t = r.sent(), [ 2, n === t[0] ];

                  case 6:
                    return r.sent(), [ 3, 7 ];

                  case 7:
                    return [ 2, !1 ];
                }
            });
        });
    }, r.prototype.afterExceptionOnStandby = function(e, t) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return t instanceof error_1.DBError && t.errCode === error_1.DBError.ECJS_COMMUNICATION_ERROR().errCode ? [ 4, this.removeStandby(e) ] : [ 3, 2 ];

                  case 1:
                    r.sent(), r.label = 2;

                  case 2:
                    return [ 2 ];
                }
            });
        });
    }, r.prototype.execute = function(o, a, _, f) {
        return __awaiter(this, void 0, void 0, function() {
            var e, t, n, i, c, s, u;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return [ 4, this.recoverStandby(o) ];

                  case 1:
                    r.sent(), this.distributeSql(o, a), t = !1, r.label = 2;

                  case 2:
                    return r.trys.push([ 2, 17, , 21 ]), [ 4, _() ];

                  case 3:
                    switch (e = r.sent(), n = o.middlewares.rwInfo.connCurrent, n.lastExecInfo.retSqlType) {
                      case const_1.Const.SQL_TYPE_COMMIT:
                      case const_1.Const.SQL_TYPE_ROLLBACK:
                      case const_1.Const.SQL_TYPE_SET_CURSCH:
                      case const_1.Const.SQL_TYPE_SET_SESS_TRAN:
                      case const_1.Const.SQL_TYPE_SET_TIME_ZONE:
                      case const_1.Const.SQL_TYPE_SAVEPNT:
                        return [ 3, 4 ];

                      case const_1.Const.SQL_TYPE_DML_CALL:
                        return [ 3, 9 ];

                      case const_1.Const.SQL_TYPE_DML_SELECT:
                        return [ 3, 14 ];
                    }
                    return [ 3, 15 ];

                  case 4:
                    if (!(i = n !== o ? o : o.middlewares.rwInfo.connStandby)) return [ 3, 8 ];
                    r.label = 5;

                  case 5:
                    return r.trys.push([ 5, 7, , 8 ]), [ 4, f(i) ];

                  case 6:
                    return r.sent(), [ 3, 8 ];

                  case 7:
                    return r.sent(), [ 3, 8 ];

                  case 8:
                    return [ 3, 16 ];

                  case 9:
                    if ("SP_SET_PARA_VALUE" !== (c = a.trim().split(/[ (]/, 2)[0]).toUpperCase() && "SP_SET_SESSION_READONLY" !== c.toUpperCase()) return [ 3, 13 ];
                    if (!(s = n !== o ? o : o.middlewares.rwInfo.connStandby)) return [ 3, 13 ];
                    r.label = 10;

                  case 10:
                    return r.trys.push([ 10, 12, , 13 ]), [ 4, f(s) ];

                  case 11:
                    return r.sent(), [ 3, 13 ];

                  case 12:
                    return r.sent(), [ 3, 13 ];

                  case 13:
                    return [ 3, 16 ];

                  case 14:
                    return !o.conn_prop_rwHA || n !== o.middlewares.rwInfo.connStandby || o.lastExecInfo.rsDatas && 0 !== o.lastExecInfo.rsDatas.length || (t = !0), 
                    [ 3, 16 ];

                  case 15:
                    return [ 3, 16 ];

                  case 16:
                    return [ 3, 21 ];

                  case 17:
                    return u = r.sent(), o.middlewares.rwInfo.connCurrent !== o.middlewares.rwInfo.connStandby ? [ 3, 19 ] : [ 4, this.afterExceptionOnStandby(o, u) ];

                  case 18:
                    return r.sent(), t = !0, [ 3, 20 ];

                  case 19:
                    throw u;

                  case 20:
                    return [ 3, 21 ];

                  case 21:
                    return t ? (o.middlewares.rwInfo.toPrimary(), o.middlewares.rwInfo.connCurrent = o, 
                    [ 2, f(o) ]) : [ 2, e ];
                }
            });
        });
    }, r.prototype.checkReadonly = function(r, e) {
        var t, n = !0;
        return e && !r.conn_prop_rwIgnoreSql && (t = e.trim().split(" ", 2)[0], n = !(stringUtil_1.StringUtil.equalsIgnoreCase(t, "insert") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "update") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "delete") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "create") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "truncate") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "drop") || stringUtil_1.StringUtil.equalsIgnoreCase(t, "alter"))), 
        n;
    }, r.prototype.distributeSql = function(r, e) {
        var t = this.isStandbyAlive(r) && r.isoLevel !== const_1.Const.CLI_SERIALIZABLE && this.checkReadonly(r, e) ? r.middlewares.rwInfo.distribute === middleware_1.RWSiteEnum.PRIMARY && !r.trxFinish || r.middlewares.rwInfo.distribute === middleware_1.RWSiteEnum.STANDBY && !r.middlewares.rwInfo.connStandby.trxFinish ? r.middlewares.rwInfo.distribute : r.conn_prop_rwAutoDistribute ? r.middlewares.rwInfo.toAny() : r.readOnly ? r.middlewares.rwInfo.toStandby() : r.middlewares.rwInfo.toPrimary() : r.middlewares.rwInfo.toPrimary();
        return r.middlewares.rwInfo.connCurrent = t === middleware_1.RWSiteEnum.PRIMARY ? r : r.middlewares.rwInfo.connStandby, 
        t;
    }, r.prototype.isStandbyAlive = function(r) {
        return void 0 !== r.middlewares.rwInfo.connStandby && !r.middlewares.rwInfo.connStandby.closed;
    }, r.prototype.removeStandby = function(e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    if (!e.middlewares.rwInfo.connStandby) return [ 3, 5 ];
                    r.label = 1;

                  case 1:
                    return r.trys.push([ 1, 3, , 4 ]), [ 4, e.middlewares.rwInfo.connStandby.do_close() ];

                  case 2:
                    return r.sent(), [ 3, 4 ];

                  case 3:
                    return r.sent(), [ 3, 4 ];

                  case 4:
                    e.middlewares.rwInfo.connStandby = void 0, r.label = 5;

                  case 5:
                    return [ 2 ];
                }
            });
        });
    }, r;
}();

exports.RWUtil = RWUtil;
