// dmdb/src/desc/abstractLob.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, h, u, o) {
    return new (u = u || Promise)(function(r, i) {
        function s(t) {
            try {
                e(o.next(t));
            } catch (t) {
                i(t);
            }
        }
        function n(t) {
            try {
                e(o.throw(t));
            } catch (t) {
                i(t);
            }
        }
        function e(t) {
            var i;
            t.done ? r(t.value) : ((i = t.value) instanceof u ? i : new u(function(t) {
                t(i);
            })).then(s, n);
        }
        e((o = o.apply(t, h || [])).next());
    });
}, __generator = this && this.__generator || function(r, s) {
    var n, e, h, u = {
        label: 0,
        sent: function() {
            if (1 & h[0]) throw h[1];
            return h[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: i(0),
        throw: i(1),
        return: i(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function i(i) {
        return function(t) {
            return function(i) {
                if (n) throw new TypeError("Generator is already executing.");
                for (;u; ) try {
                    if (n = 1, e && (h = 2 & i[0] ? e.return : i[0] ? e.throw || ((h = e.return) && h.call(e), 
                    0) : e.next) && !(h = h.call(e, i[1])).done) return h;
                    switch (e = 0, h && (i = [ 2 & i[0], h.value ]), i[0]) {
                      case 0:
                      case 1:
                        h = i;
                        break;

                      case 4:
                        return u.label++, {
                            value: i[1],
                            done: !1
                        };

                      case 5:
                        u.label++, e = i[1], i = [ 0 ];
                        continue;

                      case 7:
                        i = u.ops.pop(), u.trys.pop();
                        continue;

                      default:
                        if (!(h = 0 < (h = u.trys).length && h[h.length - 1]) && (6 === i[0] || 2 === i[0])) {
                            u = 0;
                            continue;
                        }
                        if (3 === i[0] && (!h || i[1] > h[0] && i[1] < h[3])) {
                            u.label = i[1];
                            break;
                        }
                        if (6 === i[0] && u.label < h[1]) {
                            u.label = h[1], h = i;
                            break;
                        }
                        if (h && u.label < h[2]) {
                            u.label = h[2], u.ops.push(i);
                            break;
                        }
                        h[2] && u.ops.pop(), u.trys.pop();
                        continue;
                    }
                    i = s.call(r, u);
                } catch (t) {
                    i = [ 6, t ], e = 0;
                } finally {
                    n = h = 0;
                }
                if (5 & i[0]) throw i[1];
                return {
                    value: i[0] ? i[1] : void 0,
                    done: !0
                };
            }([ i, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var const_1 = require("./const"), bufferUtil_1 = require("../utils/bufferUtil"), error_1 = require("../driver/error"), msg_1 = require("../net/msg/msg"), data_1 = require("../convert/data"), AbstractLob = function() {
    function e(t, i, r, s) {
        if (this.blobId = BigInt(0), this.storageType = 0, this.groupId = -1, this.fileId = -1, 
        this.pageNo = -1, this.tabId = 0, this.colId = 0, this.curFileId = 0, this.curPageNo = 0, 
        this.curOffset = 0, this.totalOffset = BigInt(0), this.readOver = !1, this.local = !0, 
        this.updateable = !0, this.bytesLength = BigInt(-1), this.length = -1, this.compatibleOracle = !1, 
        this.fetchAll = !1, this.freed = !1, this.modify = !1, this.conn = i, this.lobFlag = t, 
        r && s) {
            this.compatibleOracle = i.compatibleOracle(), this.local = !1, this.updateable = !s.readonly, 
            this.tabId = s.lobTabId, this.colId = s.lobColId;
            var n = 0;
            if (this.storageType = bufferUtil_1.BufferUtil.a(r, n), n += const_1.Const.BYTE_SIZE, 
            this.blobId = bufferUtil_1.BufferUtil.i(r, n), n += const_1.Const.LINT64_SIZE, this.bytesLength = BigInt(bufferUtil_1.BufferUtil.g(r, n)), 
            (n += const_1.Const.ULINT_SIZE) === r.length) return;
            if (this.groupId = bufferUtil_1.BufferUtil.U(r, n), n += const_1.Const.USINT_SIZE, 
            this.fileId = bufferUtil_1.BufferUtil.U(r, n), n += const_1.Const.USINT_SIZE, this.pageNo = bufferUtil_1.BufferUtil.s(r, n), 
            n += const_1.Const.ULINT_SIZE, this.resetCurrentInfo(), n === r.length) return;
            if (this.tabId = bufferUtil_1.BufferUtil.s(r, n), n += const_1.Const.ULINT_SIZE, 
            this.colId = bufferUtil_1.BufferUtil.U(r, n), n += const_1.Const.USINT_SIZE, i.msgVersion < msg_1.Msg.VERSION_ROWID_B12 ? (this.rowId = bufferUtil_1.BufferUtil.m(r, n, const_1.Const.DDWORD_SIZE), 
            n += const_1.Const.DDWORD_SIZE) : (this.rowId = bufferUtil_1.BufferUtil.m(r, n, const_1.Const.ROWID_SIZE), 
            n += const_1.Const.ROWID_SIZE), n === r.length) return;
            this.storageType === e.STORAGE_LONG_ROW && (this.bytesLength = bufferUtil_1.BufferUtil.i(r, n), 
            n += const_1.Const.LINT64_SIZE);
        }
    }
    return e.prototype.resetCurrentInfo = function() {
        this.curFileId = this.fileId, this.curPageNo = this.pageNo, this.totalOffset = BigInt(0), 
        this.curOffset = 0;
    }, e.prototype.checkFreed = function() {
        if (this.freed) throw error_1.DBError.ECJS_LOB_FREED();
    }, e.prototype.getLength = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return (this.checkFreed(), -1 !== this.length) ? [ 3, 2 ] : (this.conn.checkClosed(), 
                    [ 4, (i = this).conn.access.q(this) ]);

                  case 1:
                    i.length = t.sent(), t.label = 2;

                  case 2:
                    return [ 2, this.length ];
                }
            });
        });
    }, e.prototype.canOptimized = function(t) {
        return t.lobOptimized && t === this.conn && this.storageType !== e.STORAGE_IN_ROW && !this.fetchAll && !this.local && !this.modify;
    }, e.prototype.getHeadSize = function() {
        return this.conn.newLobFlag ? this.conn.msgVersion < msg_1.Msg.VERSION_ROWID_B12 ? e.NBLOB_HEAD_SIZE_EX : e.NBLOB_HEAD_SIZE_EX_ROWID_12B : this.storageType === e.STORAGE_IN_ROW ? e.NBLOB_HEAD_SIZE_INROW : e.NBLOB_HEAD_SIZE_OUTROW;
    }, e.prototype.buildCtlData = function() {
        return new data_1.LobCtl(this);
    }, e.LOB_FLAG_BYTE = 0, e.LOB_FLAG_CHAR = 1, e.STORAGE_IN_ROW = 1, e.STORAGE_OUT_ROW = 2, 
    e.STORAGE_LONG_ROW = 4, e.NBLOB_HEAD_SIZE_INROW = 13, e.NBLOB_HEAD_SIZE_OUTROW = 21, 
    e.NBLOB_HEAD_SIZE_EX = 43, e.NBLOB_HEAD_SIZE_EX_ROWID_12B = 47, e;
}();

exports.AbstractLob = AbstractLob;
