// dmdb/src/desc/ep.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, i, o, a) {
    return new (o = o || Promise)(function(s, n) {
        function e(t) {
            try {
                c(a.next(t));
            } catch (t) {
                n(t);
            }
        }
        function r(t) {
            try {
                c(a.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function c(t) {
            var n;
            t.done ? s(t.value) : ((n = t.value) instanceof o ? n : new o(function(t) {
                t(n);
            })).then(e, r);
        }
        c((a = a.apply(t, i || [])).next());
    });
}, __generator = this && this.__generator || function(s, e) {
    var r, c, i, o = {
        label: 0,
        sent: function() {
            if (1 & i[0]) throw i[1];
            return i[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (r) throw new TypeError("Generator is already executing.");
                for (;o; ) try {
                    if (r = 1, c && (i = 2 & n[0] ? c.return : n[0] ? c.throw || ((i = c.return) && i.call(c), 
                    0) : c.next) && !(i = i.call(c, n[1])).done) return i;
                    switch (c = 0, i && (n = [ 2 & n[0], i.value ]), n[0]) {
                      case 0:
                      case 1:
                        i = n;
                        break;

                      case 4:
                        return o.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        o.label++, c = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = o.ops.pop(), o.trys.pop();
                        continue;

                      default:
                        if (!(i = 0 < (i = o.trys).length && i[i.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            o = 0;
                            continue;
                        }
                        if (3 === n[0] && (!i || n[1] > i[0] && n[1] < i[3])) {
                            o.label = n[1];
                            break;
                        }
                        if (6 === n[0] && o.label < i[1]) {
                            o.label = i[1], i = n;
                            break;
                        }
                        if (i && o.label < i[2]) {
                            o.label = i[2], o.ops.push(n);
                            break;
                        }
                        i[2] && o.ops.pop(), o.trys.pop();
                        continue;
                    }
                    n = e.call(s, o);
                } catch (t) {
                    n = [ 6, t ], c = 0;
                } finally {
                    r = i = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var const_1 = require("./const"), EP = function() {
    function s(t, n) {
        this.alive = !1, this.statusRefreshTs = 0, this.serverMode = -1, this.serverStatus = -1, 
        this.dscControl = !1, this.sort = s.SORT_UNKNOWN, this.host = t, this.port = n;
    }
    return s.prototype.connect = function(s) {
        return __awaiter(this, void 0, void 0, function() {
            var n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    s.ep = this, s.conn_prop_host = this.host, s.conn_prop_port = this.port, s.props.set("host", this.host), 
                    s.props.set("port", this.port.toString()), t.label = 1;

                  case 1:
                    return t.trys.push([ 1, 3, , 4 ]), [ 4, s.do_openConnection() ];

                  case 2:
                    return t.sent(), this.refreshStatus(s, !0), [ 3, 4 ];

                  case 3:
                    throw n = t.sent(), this.refreshStatus(s, !1), n;

                  case 4:
                    return [ 2 ];
                }
            });
        });
    }, s.prototype.getSort = function(t) {
        return !t || Date.now() - this.statusRefreshTs < s.STATUS_VALID_TIME ? this.sort : s.SORT_UNKNOWN;
    }, s.prototype.getServerModeDesc = function(t) {
        switch (t) {
          case const_1.Const.SERVER_MODE_NORMAL:
            return "NORMAL";

          case const_1.Const.SERVER_MODE_PRIMARY:
            return "PRIMARY";

          case const_1.Const.SERVER_MODE_STANDBY:
            return "STANDBY";

          default:
            return "UNKNOW";
        }
    }, s.prototype.getServerStatusDesc = function(t) {
        switch (t) {
          case const_1.Const.SERVER_STATUS_OPEN:
            return "OPEN";

          case const_1.Const.SERVER_STATUS_MOUNT:
            return "MOUNT";

          case const_1.Const.SERVER_STATUS_SUSPEND:
            return "SUSPEND";

          default:
            return "UNKNOW";
        }
    }, s.prototype.toString = function() {
        return this.host + ":" + this.port + " (" + this.getServerModeDesc(this.serverMode) + ", " + this.getServerStatusDesc(this.serverStatus) + (this.dscControl ? ", DSC CONTROL" : "") + "))";
    }, s.prototype.refreshStatus = function(t, n) {
        this.alive = n, this.statusRefreshTs = Date.now(), this.serverMode = n ? t.svrMode : -1, 
        this.serverStatus = n ? t.svrStat : -1, this.dscControl = !!n && t.dscControl, this.sort = n ? this.calcSort(t.conn_prop_loginMode) : s.SORT_SERVER_NOT_ALIVE;
    }, s.prototype.calcSort = function(t) {
        var n = 0;
        switch (t) {
          case const_1.Const.LOGIN_MODE_PRIMARY_FIRST:
            switch (this.serverMode) {
              case const_1.Const.SERVER_MODE_NORMAL:
                n += 10 * s.SORT_NORMAL;
                break;

              case const_1.Const.SERVER_MODE_PRIMARY:
                n += 100 * s.SORT_PRIMARY;
                break;

              case const_1.Const.SERVER_MODE_STANDBY:
                n += s.SORT_STANDBY;
            }
            break;

          case const_1.Const.LOGIN_MODE_STANDBY_FIRST:
            switch (this.serverMode) {
              case const_1.Const.SERVER_MODE_NORMAL:
                n += s.SORT_NORMAL;
                break;

              case const_1.Const.SERVER_MODE_PRIMARY:
                n += 10 * s.SORT_PRIMARY;
                break;

              case const_1.Const.SERVER_MODE_STANDBY:
                n += 100 * s.SORT_STANDBY;
            }
            break;

          case const_1.Const.LOGIN_MODE_NORMAL_FIRST:
            switch (this.serverMode) {
              case const_1.Const.SERVER_MODE_STANDBY:
                n += s.SORT_STANDBY;
                break;

              case const_1.Const.SERVER_MODE_PRIMARY:
                n += 10 * s.SORT_PRIMARY;
                break;

              case const_1.Const.SERVER_MODE_NORMAL:
                n += 100 * s.SORT_NORMAL;
            }
            break;

          case const_1.Const.LOGIN_MODE_PRIMARY_ONLY:
            if (this.serverMode != const_1.Const.SERVER_MODE_PRIMARY) return s.SORT_SERVER_MODE_INVALID;
            n += s.SORT_PRIMARY;
            break;

          case const_1.Const.LOGIN_MODE_STANDBY_ONLY:
            if (this.serverMode != const_1.Const.SERVER_MODE_STANDBY) return s.SORT_SERVER_MODE_INVALID;
            n += s.SORT_STANDBY;
        }
        switch (this.serverStatus) {
          case const_1.Const.SERVER_STATUS_MOUNT:
            n += s.SORT_MOUNT;
            break;

          case const_1.Const.SERVER_STATUS_OPEN:
            n += s.SORT_OPEN;
            break;

          case const_1.Const.SERVER_STATUS_SUSPEND:
            n += s.SORT_SUSPEND;
        }
        return n;
    }, s.STATUS_VALID_TIME = 2e4, s.SORT_SERVER_MODE_INVALID = -1, s.SORT_SERVER_NOT_ALIVE = -2, 
    s.SORT_UNKNOWN = const_1.Const.INT32_MAX, s.SORT_NORMAL = 30, s.SORT_PRIMARY = 20, 
    s.SORT_STANDBY = 10, s.SORT_OPEN = 3, s.SORT_MOUNT = 2, s.SORT_SUSPEND = 1, s;
}();

exports.EP = EP;
