// dmdb/src/desc/epGroup.js
"use strict";

var __extends = this && this.__extends || function() {
    var n = function(t, r) {
        return (n = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var e in r) r.hasOwnProperty(e) && (t[e] = r[e]);
        })(t, r);
    };
    return function(t, r) {
        function e() {
            this.constructor = t;
        }
        n(t, r), t.prototype = null === r ? Object.create(r) : (e.prototype = r.prototype, 
        new e());
    };
}(), __awaiter = this && this.__awaiter || function(t, o, c, u) {
    return new (c = c || Promise)(function(e, r) {
        function n(t) {
            try {
                s(u.next(t));
            } catch (t) {
                r(t);
            }
        }
        function i(t) {
            try {
                s(u.throw(t));
            } catch (t) {
                r(t);
            }
        }
        function s(t) {
            var r;
            t.done ? e(t.value) : ((r = t.value) instanceof c ? r : new c(function(t) {
                t(r);
            })).then(n, i);
        }
        s((u = u.apply(t, o || [])).next());
    });
}, __generator = this && this.__generator || function(e, n) {
    var i, s, o, c = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: r(0),
        throw: r(1),
        return: r(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function r(r) {
        return function(t) {
            return function(r) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (i = 1, s && (o = 2 & r[0] ? s.return : r[0] ? s.throw || ((o = s.return) && o.call(s), 
                    0) : s.next) && !(o = o.call(s, r[1])).done) return o;
                    switch (s = 0, o && (r = [ 2 & r[0], o.value ]), r[0]) {
                      case 0:
                      case 1:
                        o = r;
                        break;

                      case 4:
                        return c.label++, {
                            value: r[1],
                            done: !1
                        };

                      case 5:
                        c.label++, s = r[1], r = [ 0 ];
                        continue;

                      case 7:
                        r = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = c.trys).length && o[o.length - 1]) && (6 === r[0] || 2 === r[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === r[0] && (!o || r[1] > o[0] && r[1] < o[3])) {
                            c.label = r[1];
                            break;
                        }
                        if (6 === r[0] && c.label < o[1]) {
                            c.label = o[1], o = r;
                            break;
                        }
                        if (o && c.label < o[2]) {
                            c.label = o[2], c.ops.push(r);
                            break;
                        }
                        o[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    r = n.call(e, c);
                } catch (t) {
                    r = [ 6, t ], s = 0;
                } finally {
                    i = o = 0;
                }
                if (5 & r[0]) throw r[1];
                return {
                    value: r[0] ? r[1] : void 0,
                    done: !0
                };
            }([ r, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var os_1 = require("os"), error_1 = require("../driver/error"), const_1 = require("./const"), EPGroup = function() {
    function t(t, r) {
        this.name = t, this.epList = r;
    }
    return t.prototype.connect = function(o) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    this.resetEpSelectorConfig(o.conn_prop_epSelector), e = this.epSelector.start(), 
                    n = 1 === this.epList.length ? o.conn_prop_switchTimes : o.conn_prop_switchTimes + 1, 
                    t.label = 1;

                  case 1:
                    t.trys.push([ 1, , 10, 11 ]), i = 0, t.label = 2;

                  case 2:
                    if (!(i < n)) return [ 3, 9 ];
                    t.label = 3;

                  case 3:
                    return t.trys.push([ 3, 5, , 8 ]), [ 4, this.epSelector.select(e, o, 0 === i, i === n - 1) ];

                  case 4:
                    return t.sent(), [ 2, o ];

                  case 5:
                    return s = t.sent(), r = s, error_1.DBError.isReconnectErrorCode(s.errCode) ? [ 4, new Promise(function(t, r) {
                        setTimeout(function() {
                            t();
                        }, o.conn_prop_switchInterval);
                    }) ] : [ 3, 7 ];

                  case 6:
                    return t.sent(), [ 3, 8 ];

                  case 7:
                    return [ 3, 9 ];

                  case 8:
                    return i++, [ 3, 2 ];

                  case 9:
                    throw r;

                  case 10:
                    return this.resetEpSelectorStart(o.ep.epSeqno, o), [ 7 ];

                  case 11:
                    return [ 2 ];
                }
            });
        });
    }, t.prototype.resetEpSelectorConfig = function(t) {
        this.epSelector ? this.epSelector.value !== t && (t === EPSelector.TYPE_WELL_DISTRIBUTE ? this.epSelector = new WellDistributeEpSelector(this) : this.epSelector instanceof SingleEpSelector ? this.epSelector.reset(t - 1) : this.epSelector = new SingleEpSelector(this, t)) : t === EPSelector.TYPE_WELL_DISTRIBUTE ? this.epSelector = new WellDistributeEpSelector(this) : this.epSelector = new SingleEpSelector(this, t);
    }, t.prototype.resetEpSelectorStart = function(t, r) {
        r.conn_prop_epSelectorDynamic && this.epSelector.value !== EPSelector.TYPE_WELL_DISTRIBUTE && t !== this.epSelector.startPos && this.epSelector.reset(t);
    }, t;
}();

exports.EPGroup = EPGroup;

var EPSelector = function() {
    function t(t, r, e) {
        this.modifyTime = -1, this.group = t, this.value = r, this.startPos = e % t.epList.length;
    }
    return t.prototype.select = function(o, c, u, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r, e, n, i, s;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    if (this.beforeSelect(o, u), r = "", e = null, a && !u && !o.some(function(t) {
                        return t.alive;
                    })) throw error_1.DBError.ECJS_COMMUNICATION_ERROR();
                    n = 0, t.label = 1;

                  case 1:
                    if (!(n < o.length)) return [ 3, 9 ];
                    i = o[n], t.label = 2;

                  case 2:
                    return t.trys.push([ 2, 5, , 8 ]), [ 4, i.connect(c) ];

                  case 3:
                    return t.sent(), [ 4, this.checkServerMode(a, c) ];

                  case 4:
                    if (!t.sent()) throw error_1.DBError.ECJS_INVALID_SERVER_MODE();
                    return [ 2 ];

                  case 5:
                    return s = t.sent(), e = e || s, r += "[" + i.toString() + "]" + s.message + os_1.EOL, 
                    error_1.DBError.isReconnectErrorCode(s.errCode) ? [ 4, c._innerFree() ] : [ 3, 7 ];

                  case 6:
                    return t.sent(), [ 3, 8 ];

                  case 7:
                    return [ 3, 9 ];

                  case 8:
                    return n++, [ 3, 1 ];

                  case 9:
                    if (e) throw e;
                    throw new error_1.DBError(error_1.DBError.ECJS_COMMUNICATION_ERROR().errCode, r, null);
                }
            });
        });
    }, t.prototype.checkServerMode = function(r, e) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return !e.conn_prop_loginDscCtrl || e.dscControl ? [ 3, 2 ] : [ 4, e._innerFree() ];

                  case 1:
                    throw t.sent(), error_1.DBError.ECJS_INVALID_SERVER_MODE();

                  case 2:
                    return 0 < e.conn_prop_loginStatus && e.svrStat !== e.conn_prop_loginStatus ? [ 4, e._innerFree() ] : [ 3, 4 ];

                  case 3:
                    throw t.sent(), error_1.DBError.ECJS_INVALID_SERVER_MODE();

                  case 4:
                    if (r) switch (e.conn_prop_loginMode) {
                      case const_1.Const.LOGIN_MODE_PRIMARY_ONLY:
                        return [ 2, e.svrMode === const_1.Const.SERVER_MODE_PRIMARY ];

                      case const_1.Const.LOGIN_MODE_STANDBY_ONLY:
                        return [ 2, e.svrMode === const_1.Const.SERVER_MODE_STANDBY ];

                      default:
                        return [ 2, !0 ];
                    }
                    switch (e.conn_prop_loginMode) {
                      case const_1.Const.LOGIN_MODE_NORMAL_FIRST:
                        return [ 2, e.svrMode == const_1.Const.SERVER_MODE_NORMAL ];

                      case const_1.Const.LOGIN_MODE_PRIMARY_FIRST:
                      case const_1.Const.LOGIN_MODE_PRIMARY_ONLY:
                        return [ 2, e.svrMode == const_1.Const.SERVER_MODE_PRIMARY ];

                      case const_1.Const.LOGIN_MODE_STANDBY_FIRST:
                      case const_1.Const.LOGIN_MODE_STANDBY_ONLY:
                        return [ 2, e.svrMode == const_1.Const.SERVER_MODE_STANDBY ];
                    }
                    return [ 2, !1 ];
                }
            });
        });
    }, t.TYPE_WELL_DISTRIBUTE = 0, t;
}(), WellDistributeEpSelector = function(r) {
    function t(t) {
        return r.call(this, t, EPSelector.TYPE_WELL_DISTRIBUTE, 0 === t.epList.length ? -1 : Math.floor(Math.random() * t.epList.length) - 1) || this;
    }
    return __extends(t, r), t.prototype.start = function() {
        var t = this.group.epList.length, r = new Array(t);
        this.startPos = (this.startPos + 1) % t;
        for (var e = 0; e < t; e++) r[e] = this.group.epList[(e + this.startPos) % t];
        return r;
    }, t.prototype.beforeSelect = function(t, e) {
        e || t.sort(function(t, r) {
            return r.getSort(e) - t.getSort(e);
        });
    }, t;
}(EPSelector), SingleEpSelector = function(e) {
    function t(t, r) {
        return e.call(this, t, r, r - 1) || this;
    }
    return __extends(t, e), t.prototype.reset = function(t) {
        this.startPos = t % this.group.epList.length, this.modifyTime = new Date().getMilliseconds();
    }, t.prototype.start = function() {
        for (var t = this.group.epList.length, r = new Array(t), e = 0; e < t; e++) r[e] = this.group.epList[(e + this.startPos) % t];
        return r;
    }, t.prototype.beforeSelect = function(t, r) {}, t;
}(EPSelector);
