// dmdb/src/desc/executeRetInfo.js
"use strict";

var __read = this && this.__read || function(t, i) {
    var e = "function" == typeof Symbol && t[Symbol.iterator];
    if (!e) return t;
    var r, s, n = e.call(t), u = [];
    try {
        for (;(void 0 === i || 0 < i--) && !(r = n.next()).done; ) u.push(r.value);
    } catch (t) {
        s = {
            error: t
        };
    } finally {
        try {
            r && !r.done && (e = n.return) && e.call(n);
        } finally {
            if (s) throw s.error;
        }
    }
    return u;
}, __spread = this && this.__spread || function() {
    for (var t = [], i = 0; i < arguments.length; i++) t = t.concat(__read(arguments[i]));
    return t;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var middleware_1 = require("../middleware/middleware"), unenumerable_1 = require("../utils/unenumerable"), ExecuteRetInfo = function() {
    function f() {
        this.outParamDatas = new Array(0), this.hasResultSet = !1, this.rsSizeof = 0, this.rsCacheOffset = 0, 
        this.rsBdta = !1, this.rsUpdatable = !1, this.updateCount = BigInt(0), this.retSqlType = -1;
    }
    return f.prototype.union = function(t, i, e, r) {
        var s, n;
        if (t instanceof f) {
            var u = t;
            if (void 0 === this.updateCounts || this.updateCounts.length !== i) {
                var a = new Array(i).fill(0);
                if (void 0 !== this.updateCounts) for (var h = 0; h < this.updateCounts.length; h++) a[h] = this.updateCounts[h];
                this.updateCounts = a;
            }
            if (void 0 !== u.updateCounts) for (h = e; h < e + r; h++) this.updateCounts[h] = u.updateCounts[h - e];
            u.outParamDatas && (s = this.outParamDatas).push.apply(s, __spread(u.outParamDatas)), 
            u.batchErrors && (this.batchErrors || (this.batchErrors = []), (n = this.batchErrors).push.apply(n, __spread(u.batchErrors)));
        } else {
            var o = t;
            if (void 0 === this.updateCounts || this.updateCounts.length !== i) {
                a = new Array(i).fill(0);
                if (void 0 !== this.updateCounts) for (h = 0; h < this.updateCounts.length; h++) a[h] = this.updateCounts[h];
                this.updateCounts = a;
            }
            for (h = e; h < e + r; h++) this.updateCounts[h] = o;
        }
    }, f;
}();

exports.ExecuteRetInfo = ExecuteRetInfo;

var Result = function() {
    function i(t) {
        this.middlewares = new middleware_1.Middlewares(t, null, i.nextID, t.middlewares), 
        i.nextID++, unenumerable_1.setUnenumerable(this, [ "middlewares", "nextID" ]);
    }
    return i.nextID = 0, i;
}();

exports.Result = Result;

var Results = function() {
    function i(t) {
        this.middlewares = new middleware_1.Middlewares(t, null, i.nextID, t.middlewares), 
        i.nextID++, unenumerable_1.setUnenumerable(this, [ "middlewares", "nextID" ]);
    }
    return i.nextID = 0, i;
}();

exports.Results = Results;
