// dmdb/src/desc/properties.js
"use strict";

var __values = this && this.__values || function(t) {
    var r = "function" == typeof Symbol && Symbol.iterator, i = r && t[r], n = 0;
    if (i) return i.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && n >= t.length && (t = void 0), {
                value: t && t[n++],
                done: !t
            };
        }
    };
    throw new TypeError(r ? "Object is not iterable." : "Symbol.iterator is not defined.");
}, __read = this && this.__read || function(t, r) {
    var i = "function" == typeof Symbol && t[Symbol.iterator];
    if (!i) return t;
    var n, e, o = i.call(t), f = [];
    try {
        for (;(void 0 === r || 0 < r--) && !(n = o.next()).done; ) f.push(n.value);
    } catch (t) {
        e = {
            error: t
        };
    } finally {
        try {
            n && !n.done && (i = o.return) && i.call(o);
        } finally {
            if (e) throw e.error;
        }
    }
    return f;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var convertUtil_1 = require("../utils/convertUtil"), stringUtil_1 = require("../utils/stringUtil"), Properties = function() {
    function h(t) {
        this.innerProps = new Map(), void 0 !== t && this.setProps(t.innerProps);
    }
    return h.prototype.setProps = function(t) {
        var r, i;
        if (void 0 !== t) try {
            for (var n = __values(t), e = n.next(); !e.done; e = n.next()) {
                var o = __read(e.value, 2), f = o[0], a = o[1];
                this.innerProps.set(f, a);
            }
        } catch (t) {
            r = {
                error: t
            };
        } finally {
            try {
                e && !e.done && (i = n.return) && i.call(n);
            } finally {
                if (r) throw r.error;
            }
        }
    }, h.prototype.setDiffProps = function(t) {
        var r, i, n, e;
        if (void 0 !== t) if (t instanceof h) try {
            for (var o = __values(t.innerProps), f = o.next(); !f.done; f = o.next()) {
                var a = __read(f.value, 2), u = a[0], s = a[1];
                this.innerProps.has(u) || this.innerProps.set(u, s);
            }
        } catch (t) {
            r = {
                error: t
            };
        } finally {
            try {
                f && !f.done && (i = o.return) && i.call(o);
            } finally {
                if (r) throw r.error;
            }
        } else if (t instanceof Map) try {
            for (var l = __values(t), v = l.next(); !v.done; v = l.next()) {
                var c = __read(v.value, 2), u = c[0], s = c[1];
                this.innerProps.has(u) || this.innerProps.set(u, s);
            }
        } catch (t) {
            n = {
                error: t
            };
        } finally {
            try {
                v && !v.done && (e = l.return) && e.call(l);
            } finally {
                if (n) throw n.error;
            }
        }
    }, h.prototype.setAny = function(t, r) {
        this.innerProps.set(t.toLowerCase(), r);
    }, h.prototype.set = function(t, r) {
        this.innerProps.set(t.toLowerCase(), r);
    }, h.prototype.setIfNotExist = function(t, r) {
        this.innerProps.has(t.toLowerCase()) || this.set(t, r);
    }, h.prototype.getAny = function(t) {
        return this.innerProps.get(t.toLowerCase());
    }, h.prototype.getBoolean = function(t, r) {
        var i = this.innerProps.get(t.toLowerCase());
        return void 0 !== i ? convertUtil_1.ConvertUtil.toBoolean(i.toString()) : r;
    }, h.prototype.getNumber = function(t, r, i, n) {
        var e, o = this.innerProps.get(t.toLowerCase());
        return o ? (e = stringUtil_1.StringUtil.equalsIgnoreCase(o.toString(), "true") ? 1 : stringUtil_1.StringUtil.equalsIgnoreCase(o.toString(), "false") ? 0 : Number.parseInt(o), 
        isNaN(e) || void 0 !== i && e < i || void 0 !== n && n < e ? r : e) : r;
    }, h.prototype.getTrimString = function(t, r) {
        var i = this.innerProps.get(t.toLowerCase());
        return i ? i.trim() : r;
    }, h.prototype.getStringArray = function(t, r) {
        var i = this.getTrimString(t, "");
        if (i) {
            var n = i.split(",");
            if (0 < n.length) return n;
        }
        return r;
    }, h.prototype.size = function() {
        return this.innerProps.size;
    }, h;
}();

exports.Properties = Properties;
