// dmdb/src/convert/db2j.js
"use strict";

var __awaiter = this && this.__awaiter || function(e, c, s, _) {
    return new (s = s || Promise)(function(t, r) {
        function n(e) {
            try {
                u(_.next(e));
            } catch (e) {
                r(e);
            }
        }
        function a(e) {
            try {
                u(_.throw(e));
            } catch (e) {
                r(e);
            }
        }
        function u(e) {
            var r;
            e.done ? t(e.value) : ((r = e.value) instanceof s ? r : new s(function(e) {
                e(r);
            })).then(n, a);
        }
        u((_ = _.apply(e, c || [])).next());
    });
}, __generator = this && this.__generator || function(t, n) {
    var a, u, c, s = {
        label: 0,
        sent: function() {
            if (1 & c[0]) throw c[1];
            return c[1];
        },
        trys: [],
        ops: []
    }, e = {
        next: r(0),
        throw: r(1),
        return: r(2)
    };
    return "function" == typeof Symbol && (e[Symbol.iterator] = function() {
        return this;
    }), e;
    function r(r) {
        return function(e) {
            return function(r) {
                if (a) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (a = 1, u && (c = 2 & r[0] ? u.return : r[0] ? u.throw || ((c = u.return) && c.call(u), 
                    0) : u.next) && !(c = c.call(u, r[1])).done) return c;
                    switch (u = 0, c && (r = [ 2 & r[0], c.value ]), r[0]) {
                      case 0:
                      case 1:
                        c = r;
                        break;

                      case 4:
                        return s.label++, {
                            value: r[1],
                            done: !1
                        };

                      case 5:
                        s.label++, u = r[1], r = [ 0 ];
                        continue;

                      case 7:
                        r = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(c = 0 < (c = s.trys).length && c[c.length - 1]) && (6 === r[0] || 2 === r[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === r[0] && (!c || r[1] > c[0] && r[1] < c[3])) {
                            s.label = r[1];
                            break;
                        }
                        if (6 === r[0] && s.label < c[1]) {
                            s.label = c[1], c = r;
                            break;
                        }
                        if (c && s.label < c[2]) {
                            s.label = c[2], s.ops.push(r);
                            break;
                        }
                        c[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    r = n.call(t, s);
                } catch (e) {
                    r = [ 6, e ], u = 0;
                } finally {
                    a = c = 0;
                }
                if (5 & r[0]) throw r[1];
                return {
                    value: r[0] ? r[1] : void 0,
                    done: !0
                };
            }([ r, e ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var blob_1 = require("../driver/blob"), dbtype_1 = require("../driver/dbtype"), clob_1 = require("../driver/clob"), error_1 = require("../driver/error"), dateUtil_1 = require("../utils/dateUtil"), bufferUtil_1 = require("../utils/bufferUtil"), stringUtil_1 = require("../utils/stringUtil"), intervalDT_1 = require("../driver/intervalDT"), intervalYM_1 = require("../driver/intervalYM"), numeric_1 = require("../driver/numeric"), lob_1 = require("../driver/lob"), dm_1 = require("../dm"), timestamp_1 = require("../driver/timestamp"), convertUtil_1 = require("../utils/convertUtil"), rowid_1 = require("../driver/rowid"), DB2J = function() {
    function y() {}
    return y.toNaturalJS = function(b, o, p, f) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t, n, a, u, c, s, _, i, d;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    if (null == b) return [ 2, null ];
                    switch (o.type) {
                      case dbtype_1.DBType.BOOLEAN:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.BIT:
                        return [ 3, 2 ];

                      case dbtype_1.DBType.TINYINT:
                      case dbtype_1.DBType.SMALLINT:
                      case dbtype_1.DBType.INT:
                      case dbtype_1.DBType.BIGINT:
                      case dbtype_1.DBType.REAL:
                      case dbtype_1.DBType.DOUBLE:
                      case dbtype_1.DBType.DECIMAL:
                        return [ 3, 3 ];

                      case dbtype_1.DBType.DATE:
                      case dbtype_1.DBType.TIME:
                      case dbtype_1.DBType.DATETIME:
                      case dbtype_1.DBType.DATETIME2:
                        return [ 3, 4 ];

                      case dbtype_1.DBType.TIME_TZ:
                      case dbtype_1.DBType.DATETIME_TZ:
                      case dbtype_1.DBType.DATETIME2_TZ:
                        return [ 3, 5 ];

                      case dbtype_1.DBType.INTERVAL_DT:
                        return [ 3, 6 ];

                      case dbtype_1.DBType.INTERVAL_YM:
                        return [ 3, 7 ];

                      case dbtype_1.DBType.BINARY:
                      case dbtype_1.DBType.VARBINARY:
                        return [ 3, 8 ];

                      case dbtype_1.DBType.BLOB:
                        return [ 3, 9 ];

                      case dbtype_1.DBType.CLOB:
                        return [ 3, 11 ];

                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                        return [ 3, 19 ];

                      case dbtype_1.DBType.ROWID:
                        return [ 3, 20 ];
                    }
                    return [ 3, 21 ];

                  case 1:
                    return [ 2, 0 != bufferUtil_1.BufferUtil.a(b, 0) ];

                  case 2:
                    return stringUtil_1.StringUtil.equalsIgnoreCase(o.typeName, "boolean") ? [ 2, 0 != bufferUtil_1.BufferUtil.a(b, 0) ] : [ 2, bufferUtil_1.BufferUtil.a(b, 0) ];

                  case 3:
                    return void 0 !== f && f === dm_1.dmdb.STRING || void 0 === f && dm_1.dmdb.fetchAsString.includes(dm_1.dmdb.NUMBER) ? [ 2, y.toString(b, o, p) ] : [ 2, y.toNumber(b, o, p) ];

                  case 4:
                    return void 0 !== f && f === dm_1.dmdb.STRING || void 0 === f && dm_1.dmdb.fetchAsString.includes(dm_1.dmdb.DATE) ? [ 2, y.toString(b, o, p) ] : (r = dateUtil_1.DateUtil.decode(b, o, p.conn_prop_localTimezone, p.dbTimezone), 
                    [ 2, y.toDate(b, o, p, r) ]);

                  case 5:
                    return void 0 !== f && f === dm_1.dmdb.STRING || void 0 === f && dm_1.dmdb.fetchAsString.includes(dm_1.dmdb.DATE) ? [ 2, y.toString(b, o, p) ] : ((r = dateUtil_1.DateUtil.decode(b, o, p.conn_prop_localTimezone, p.dbTimezone))[timestamp_1.Timestamp.OFFSET_TIMEZONE] !== timestamp_1.Timestamp.INVALID_VALUE && dateUtil_1.DateUtil.transformTZ(r, r[timestamp_1.Timestamp.OFFSET_TIMEZONE], p.conn_prop_localTimezone), 
                    [ 2, y.toDate(b, o, p, r) ]);

                  case 6:
                    return [ 2, new intervalDT_1.IntervalDT(b).toString() ];

                  case 7:
                    return [ 2, new intervalYM_1.IntervalYM(b).toString() ];

                  case 8:
                    return void 0 !== f && f === dm_1.dmdb.STRING || void 0 === f && dm_1.dmdb.fetchAsString.includes(dm_1.dmdb.BUFFER) ? [ 2, y.toString(b, o, p) ] : [ 2, b ];

                  case 9:
                    return void 0 !== f && f === dm_1.dmdb.BUFFER || void 0 === f && dm_1.dmdb.fetchAsBuffer.includes(dm_1.dmdb.BLOB) ? [ 2, y.toBuffer(b, o, p) ] : (t = lob_1.Lob.bind, 
                    [ 4, y.toBlob(b, o, p) ]);

                  case 10:
                    return [ 2, new (t.apply(lob_1.Lob, [ void 0, e.sent() ]))() ];

                  case 11:
                    return [ 4, y.toClob(b, o, p) ];

                  case 12:
                    return (n = e.sent(), !dm_1.dmdb.parseJson || 8192 !== o.scale && 16384 !== o.scale) ? [ 3, 15 ] : (c = (u = n).getSubString, 
                    s = [ 1 ], [ 4, n.getLength() ]);

                  case 13:
                    return [ 4, c.apply(u, s.concat([ e.sent() ])) ];

                  case 14:
                    a = e.sent()[0];
                    try {
                        return [ 2, JSON.parse(a) ];
                    } catch (e) {
                        return [ 2, a ];
                    }
                    e.label = 15;

                  case 15:
                    return void 0 !== f && f === dm_1.dmdb.STRING || void 0 === f && dm_1.dmdb.fetchAsString.includes(dm_1.dmdb.CLOB) ? (i = (_ = n).getSubString, 
                    d = [ 1 ], [ 4, n.getLength() ]) : [ 3, 18 ];

                  case 16:
                    return [ 4, i.apply(_, d.concat([ e.sent() ])) ];

                  case 17:
                    return [ 2, e.sent()[0] ];

                  case 18:
                    return [ 2, new lob_1.Lob(n) ];

                  case 19:
                    return [ 2, bufferUtil_1.BufferUtil.u(b, 0, b.length, p.getServerEncoding()) ];

                  case 20:
                    return [ 2, new rowid_1.RowId(b).toString() ];

                  case 21:
                    return [ 2, null ];
                }
            });
        });
    }, y.toNaturalJS4Bind = function(c, s, _, i) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t, n, a, u;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    if (null == c) return [ 2, null ];
                    if (void 0 === i) return [ 3, 10 ];
                    switch (i) {
                      case dm_1.dmdb.BLOB:
                        return [ 3, 1 ];

                      case dm_1.dmdb.BUFFER:
                        return [ 3, 3 ];

                      case dm_1.dmdb.CLOB:
                        return [ 3, 4 ];

                      case dm_1.dmdb.DATE:
                        return [ 3, 6 ];

                      case dm_1.dmdb.NUMBER:
                        return [ 3, 7 ];

                      case dm_1.dmdb.STRING:
                        return [ 3, 8 ];

                      case dm_1.dmdb.CURSOR:
                      case dm_1.dmdb.DEFAULT:
                        return [ 3, 9 ];
                    }
                    return [ 3, 9 ];

                  case 1:
                    return r = lob_1.Lob.bind, [ 4, y.toBlob(c, s, _) ];

                  case 2:
                    return [ 2, new (r.apply(lob_1.Lob, [ void 0, e.sent() ]))() ];

                  case 3:
                    return [ 2, y.toBuffer(c, s, _) ];

                  case 4:
                    return t = lob_1.Lob.bind, [ 4, y.toClob(c, s, _) ];

                  case 5:
                    return [ 2, new (t.apply(lob_1.Lob, [ void 0, e.sent() ]))() ];

                  case 6:
                    return [ 2, y.toDate(c, s, _) ];

                  case 7:
                    return [ 2, y.toNumber(c, s, _) ];

                  case 8:
                    return [ 2, y.toString(c, s, _) ];

                  case 9:
                    return [ 3, 10 ];

                  case 10:
                    switch (s.type) {
                      case dbtype_1.DBType.BOOLEAN:
                        return [ 3, 11 ];

                      case dbtype_1.DBType.BIT:
                        return [ 3, 12 ];

                      case dbtype_1.DBType.TINYINT:
                        return [ 3, 13 ];

                      case dbtype_1.DBType.SMALLINT:
                        return [ 3, 14 ];

                      case dbtype_1.DBType.INT:
                        return [ 3, 15 ];

                      case dbtype_1.DBType.BIGINT:
                        return [ 3, 16 ];

                      case dbtype_1.DBType.REAL:
                        return [ 3, 17 ];

                      case dbtype_1.DBType.DOUBLE:
                        return [ 3, 18 ];

                      case dbtype_1.DBType.DATE:
                      case dbtype_1.DBType.TIME:
                      case dbtype_1.DBType.DATETIME:
                      case dbtype_1.DBType.DATETIME2:
                        return [ 3, 19 ];

                      case dbtype_1.DBType.TIME_TZ:
                      case dbtype_1.DBType.DATETIME_TZ:
                      case dbtype_1.DBType.DATETIME2_TZ:
                        return [ 3, 20 ];

                      case dbtype_1.DBType.INTERVAL_DT:
                        return [ 3, 21 ];

                      case dbtype_1.DBType.INTERVAL_YM:
                        return [ 3, 22 ];

                      case dbtype_1.DBType.DECIMAL:
                        return [ 3, 23 ];

                      case dbtype_1.DBType.BINARY:
                      case dbtype_1.DBType.VARBINARY:
                        return [ 3, 24 ];

                      case dbtype_1.DBType.BLOB:
                        return [ 3, 25 ];

                      case dbtype_1.DBType.CLOB:
                        return [ 3, 27 ];

                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                        return [ 3, 29 ];

                      case dbtype_1.DBType.ROWID:
                        return [ 3, 30 ];
                    }
                    return [ 3, 31 ];

                  case 11:
                    return [ 2, 0 != bufferUtil_1.BufferUtil.a(c, 0) ];

                  case 12:
                    return stringUtil_1.StringUtil.equalsIgnoreCase(s.typeName, "boolean") ? [ 2, 0 != bufferUtil_1.BufferUtil.a(c, 0) ] : [ 2, bufferUtil_1.BufferUtil.a(c, 0) ];

                  case 13:
                    return [ 2, bufferUtil_1.BufferUtil.a(c, 0) ];

                  case 14:
                    return [ 2, bufferUtil_1.BufferUtil.c(c, 0) ];

                  case 15:
                    return [ 2, bufferUtil_1.BufferUtil.s(c, 0) ];

                  case 16:
                    return [ 2, Number(bufferUtil_1.BufferUtil.i(c, 0)) ];

                  case 17:
                    return [ 2, bufferUtil_1.BufferUtil.d(c, 0) ];

                  case 18:
                    return [ 2, bufferUtil_1.BufferUtil.b(c, 0) ];

                  case 19:
                    return n = dateUtil_1.DateUtil.decode(c, s, _.conn_prop_localTimezone, _.dbTimezone), 
                    [ 2, y.toDate(c, s, _, n) ];

                  case 20:
                    return (n = dateUtil_1.DateUtil.decode(c, s, _.conn_prop_localTimezone, _.dbTimezone))[timestamp_1.Timestamp.OFFSET_TIMEZONE] !== timestamp_1.Timestamp.INVALID_VALUE && dateUtil_1.DateUtil.transformTZ(n, n[timestamp_1.Timestamp.OFFSET_TIMEZONE], _.conn_prop_localTimezone), 
                    [ 2, y.toDate(c, s, _, n) ];

                  case 21:
                    return [ 2, new intervalDT_1.IntervalDT(c).toString() ];

                  case 22:
                    return [ 2, new intervalYM_1.IntervalYM(c).toString() ];

                  case 23:
                    return [ 2, new numeric_1.Numeric(s.prec, s.scale, c).toNumber() ];

                  case 24:
                    return [ 2, c ];

                  case 25:
                    return a = lob_1.Lob.bind, [ 4, y.toBlob(c, s, _) ];

                  case 26:
                    return [ 2, new (a.apply(lob_1.Lob, [ void 0, e.sent() ]))() ];

                  case 27:
                    return u = lob_1.Lob.bind, [ 4, y.toClob(c, s, _) ];

                  case 28:
                    return [ 2, new (u.apply(lob_1.Lob, [ void 0, e.sent() ]))() ];

                  case 29:
                    return [ 2, bufferUtil_1.BufferUtil.u(c, 0, c.length, _.getServerEncoding()) ];

                  case 30:
                    return [ 2, new rowid_1.RowId(c).toString() ];

                  case 31:
                    return [ 2, null ];
                }
            });
        });
    }, y.toBoolean = function(n, a, u) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (a.type) {
                      case dbtype_1.DBType.BIT:
                      case dbtype_1.DBType.BOOLEAN:
                      case dbtype_1.DBType.TINYINT:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.SMALLINT:
                        return [ 3, 2 ];

                      case dbtype_1.DBType.INT:
                        return [ 3, 3 ];

                      case dbtype_1.DBType.BIGINT:
                        return [ 3, 4 ];

                      case dbtype_1.DBType.REAL:
                        return [ 3, 5 ];

                      case dbtype_1.DBType.DOUBLE:
                        return [ 3, 6 ];

                      case dbtype_1.DBType.DECIMAL:
                        return [ 3, 7 ];

                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                      case dbtype_1.DBType.CLOB:
                        return [ 3, 8 ];
                    }
                    return [ 3, 10 ];

                  case 1:
                    return [ 2, 0 !== bufferUtil_1.BufferUtil.a(n, 0) ];

                  case 2:
                    return [ 2, 0 !== bufferUtil_1.BufferUtil.c(n, 0) ];

                  case 3:
                    return [ 2, 0 !== bufferUtil_1.BufferUtil.s(n, 0) ];

                  case 4:
                    return [ 2, bufferUtil_1.BufferUtil.i(n, 0) !== BigInt(0) ];

                  case 5:
                    return [ 2, 0 !== bufferUtil_1.BufferUtil.d(n, 0) ];

                  case 6:
                    return [ 2, 0 !== bufferUtil_1.BufferUtil.b(n, 0) ];

                  case 7:
                    return [ 2, 0 === new numeric_1.Numeric(a.prec, a.scale, n).toNumber() ];

                  case 8:
                    return t = (r = convertUtil_1.ConvertUtil).toBoolean, [ 4, y.charToString(n, a, u) ];

                  case 9:
                    return [ 2, t.apply(r, [ e.sent() ]) ];

                  case 10:
                    throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
                }
            });
        });
    }, y.toNumber = function(t, n, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (n.type) {
                      case dbtype_1.DBType.BIT:
                      case dbtype_1.DBType.BOOLEAN:
                      case dbtype_1.DBType.TINYINT:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.SMALLINT:
                        return [ 3, 2 ];

                      case dbtype_1.DBType.INT:
                        return [ 3, 3 ];

                      case dbtype_1.DBType.BIGINT:
                        return [ 3, 4 ];

                      case dbtype_1.DBType.REAL:
                        return [ 3, 5 ];

                      case dbtype_1.DBType.DOUBLE:
                        return [ 3, 6 ];

                      case dbtype_1.DBType.DECIMAL:
                        return [ 3, 7 ];

                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                      case dbtype_1.DBType.CLOB:
                        return [ 3, 8 ];
                    }
                    return [ 3, 10 ];

                  case 1:
                    return [ 2, bufferUtil_1.BufferUtil.a(t, 0) ];

                  case 2:
                    return [ 2, bufferUtil_1.BufferUtil.c(t, 0) ];

                  case 3:
                    return [ 2, bufferUtil_1.BufferUtil.s(t, 0) ];

                  case 4:
                    return [ 2, Number(bufferUtil_1.BufferUtil.i(t, 0)) ];

                  case 5:
                    return [ 2, bufferUtil_1.BufferUtil.d(t, 0) ];

                  case 6:
                    return [ 2, bufferUtil_1.BufferUtil.b(t, 0) ];

                  case 7:
                    return [ 2, new numeric_1.Numeric(n.prec, n.scale, t).toNumber() ];

                  case 8:
                    return r = Number, [ 4, y.charToString(t, n, a) ];

                  case 9:
                    return [ 2, r.apply(void 0, [ e.sent() ]) ];

                  case 10:
                    throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
                }
            });
        });
    }, y.toBuffer = function(d, b, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t, n, a, u, c, s, _, i;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (b.type) {
                      case dbtype_1.DBType.CLOB:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.BLOB:
                        return [ 3, 5 ];
                    }
                    return [ 3, 8 ];

                  case 1:
                    return [ 4, clob_1.Clob.newClobFromDB(d, o, b, !0) ];

                  case 2:
                    return r = e.sent(), a = (n = r).getSubString, u = [ 1 ], [ 4, r.getLength() ];

                  case 3:
                    return [ 4, a.apply(n, u.concat([ e.sent() ])) ];

                  case 4:
                    return t = e.sent()[0], [ 2, bufferUtil_1.BufferUtil.o(t, o.getServerEncoding()) ];

                  case 5:
                    return [ 4, blob_1.Blob.newBlobFromDB(d, o, b, !0) ];

                  case 6:
                    return c = e.sent(), _ = (s = c).getBytes, i = [ 1 ], [ 4, c.getLength() ];

                  case 7:
                    return [ 2, _.apply(s, i.concat([ e.sent() ])) ];

                  case 8:
                    return [ 2, d ];
                }
            });
        });
    }, y.toBlob = function(r, t, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(e) {
                switch (t.type) {
                  case dbtype_1.DBType.BLOB:
                    return [ 2, blob_1.Blob.newBlobFromDB(r, n, t, n.lobFetchAll()) ];

                  default:
                    return [ 2, blob_1.Blob.newBlobOfLocal(r, n) ];
                }
                return [ 2 ];
            });
        });
    }, y.toClob = function(n, a, u) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (a.type) {
                      case dbtype_1.DBType.CLOB:
                        return [ 3, 1 ];
                    }
                    return [ 3, 2 ];

                  case 1:
                    return [ 2, clob_1.Clob.newClobFromDB(n, u, a, u.lobFetchAll()) ];

                  case 2:
                    return t = (r = clob_1.Clob).newClobOfLocal, [ 4, y.toString(n, a, u) ];

                  case 3:
                    return [ 2, t.apply(r, [ e.sent(), u ]) ];
                }
            });
        });
    }, y.toString = function(d, b, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t, n, a, u, c, s, _, i;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (b.type) {
                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.BIT:
                      case dbtype_1.DBType.BOOLEAN:
                      case dbtype_1.DBType.TINYINT:
                        return [ 3, 2 ];

                      case dbtype_1.DBType.SMALLINT:
                        return [ 3, 3 ];

                      case dbtype_1.DBType.INT:
                        return [ 3, 4 ];

                      case dbtype_1.DBType.BIGINT:
                        return [ 3, 5 ];

                      case dbtype_1.DBType.REAL:
                        return [ 3, 6 ];

                      case dbtype_1.DBType.DOUBLE:
                        return [ 3, 7 ];

                      case dbtype_1.DBType.DECIMAL:
                        return [ 3, 8 ];

                      case dbtype_1.DBType.BINARY:
                      case dbtype_1.DBType.VARBINARY:
                        return [ 3, 9 ];

                      case dbtype_1.DBType.BLOB:
                        return [ 3, 10 ];

                      case dbtype_1.DBType.CLOB:
                        return [ 3, 14 ];

                      case dbtype_1.DBType.DATE:
                      case dbtype_1.DBType.TIME:
                      case dbtype_1.DBType.DATETIME:
                      case dbtype_1.DBType.DATETIME2:
                        return [ 3, 18 ];

                      case dbtype_1.DBType.TIME_TZ:
                      case dbtype_1.DBType.DATETIME_TZ:
                      case dbtype_1.DBType.DATETIME2_TZ:
                        return [ 3, 19 ];

                      case dbtype_1.DBType.INTERVAL_DT:
                        return [ 3, 20 ];

                      case dbtype_1.DBType.INTERVAL_YM:
                        return [ 3, 21 ];

                      case dbtype_1.DBType.ROWID:
                        return [ 3, 22 ];
                    }
                    return [ 3, 23 ];

                  case 1:
                    return [ 2, y.charToString(d, b, o) ];

                  case 2:
                    return [ 2, bufferUtil_1.BufferUtil.a(d, 0).toString() ];

                  case 3:
                    return [ 2, bufferUtil_1.BufferUtil.c(d, 0).toString() ];

                  case 4:
                    return [ 2, bufferUtil_1.BufferUtil.s(d, 0).toString() ];

                  case 5:
                    return [ 2, bufferUtil_1.BufferUtil.i(d, 0).toString() ];

                  case 6:
                    return [ 2, y.replaceNumberPoint(bufferUtil_1.BufferUtil.d(d, 0).toString(), o.formatNLS) ];

                  case 7:
                    return [ 2, y.replaceNumberPoint(bufferUtil_1.BufferUtil.b(d, 0).toString(), o.formatNLS) ];

                  case 8:
                    return [ 2, new numeric_1.Numeric(b.prec, b.scale, d).toString() ];

                  case 9:
                    return [ 2, d.toString("hex") ];

                  case 10:
                    return [ 4, blob_1.Blob.newBlobFromDB(d, o, b, !0) ];

                  case 11:
                    return r = e.sent(), n = (t = r).getBytes, a = [ 1 ], [ 4, r.getLength() ];

                  case 12:
                    return [ 4, n.apply(t, a.concat([ e.sent() ])) ];

                  case 13:
                    return [ 2, e.sent().toString("hex") ];

                  case 14:
                    return [ 4, clob_1.Clob.newClobFromDB(d, o, b, !0) ];

                  case 15:
                    return u = e.sent(), s = (c = u).getSubString, _ = [ 1 ], [ 4, u.getLength() ];

                  case 16:
                    return [ 4, s.apply(c, _.concat([ e.sent() ])) ];

                  case 17:
                    return [ 2, e.sent()[0] ];

                  case 18:
                    return i = dateUtil_1.DateUtil.decode(d, b, o.conn_prop_localTimezone, o.dbTimezone), 
                    [ 2, y.replaceNumberPoint(dateUtil_1.DateUtil.toStringFromDT(i, y.getFormatString(b, o), b.type, b.scale, o.oracleDateLanguage), o.formatNLS) ];

                  case 19:
                    return (i = dateUtil_1.DateUtil.decode(d, b, o.conn_prop_localTimezone, o.dbTimezone))[timestamp_1.Timestamp.OFFSET_TIMEZONE] !== timestamp_1.Timestamp.INVALID_VALUE && dateUtil_1.DateUtil.transformTZ(i, i[timestamp_1.Timestamp.OFFSET_TIMEZONE], o.conn_prop_localTimezone), 
                    [ 2, y.replaceNumberPoint(dateUtil_1.DateUtil.toStringFromDT(i, y.getFormatString(b, o), b.type, b.scale, o.oracleDateLanguage), o.formatNLS) ];

                  case 20:
                    return [ 2, new intervalDT_1.IntervalDT(d).toString() ];

                  case 21:
                    return [ 2, new intervalYM_1.IntervalYM(d).toString() ];

                  case 22:
                    return [ 2, new rowid_1.RowId(d).toString() ];

                  case 23:
                    throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
                }
            });
        });
    }, y.toDate = function(n, a, u, c) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    switch (a.type) {
                      case dbtype_1.DBType.DATE:
                      case dbtype_1.DBType.TIME:
                      case dbtype_1.DBType.DATETIME:
                      case dbtype_1.DBType.DATETIME2:
                        return [ 3, 1 ];

                      case dbtype_1.DBType.TIME_TZ:
                      case dbtype_1.DBType.DATETIME_TZ:
                      case dbtype_1.DBType.DATETIME2_TZ:
                        return [ 3, 2 ];

                      case dbtype_1.DBType.CHAR:
                      case dbtype_1.DBType.VARCHAR:
                      case dbtype_1.DBType.VARCHAR2:
                      case dbtype_1.DBType.CLOB:
                        return [ 3, 3 ];
                    }
                    return [ 3, 5 ];

                  case 1:
                    return c = c || dateUtil_1.DateUtil.decode(n, a, u.conn_prop_localTimezone, u.dbTimezone), 
                    [ 2, dateUtil_1.DateUtil.toDateFromDT(c, -new Date().getTimezoneOffset()) ];

                  case 2:
                    return c = c || dateUtil_1.DateUtil.decode(n, a, u.conn_prop_localTimezone, u.dbTimezone), 
                    [ 2, dateUtil_1.DateUtil.toDateFromDT(c, u.conn_prop_localTimezone) ];

                  case 3:
                    return t = (r = dateUtil_1.DateUtil).toDateFromString, [ 4, y.charToString(n, a, u) ];

                  case 4:
                    return [ 2, t.apply(r, [ e.sent(), u.conn_prop_localTimezone ]) ];

                  case 5:
                    throw error_1.DBError.ECJS_DATA_CONVERTION_ERROR();
                }
            });
        });
    }, y.charToString = function(u, c, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r, t, n, a;
            return __generator(this, function(e) {
                switch (e.label) {
                  case 0:
                    return c.type !== dbtype_1.DBType.VARCHAR2 ? [ 3, 1 ] : (u = y.processVarchar2(u, c.prec), 
                    [ 3, 5 ]);

                  case 1:
                    return c.type !== dbtype_1.DBType.CLOB ? [ 3, 5 ] : [ 4, clob_1.Clob.newClobFromDB(u, s, c, !0) ];

                  case 2:
                    return r = e.sent(), n = (t = r).getSubString, a = [ 1 ], [ 4, r.getLength() ];

                  case 3:
                    return [ 4, n.apply(t, a.concat([ e.sent() ])) ];

                  case 4:
                    return [ 2, e.sent()[0] ];

                  case 5:
                    return [ 2, bufferUtil_1.BufferUtil.u(u, 0, u.length, s.getServerEncoding()) ];
                }
            });
        });
    }, y.processVarchar2 = function(e, r) {
        var t = Buffer.alloc(r);
        t.copy(e, 0, 0, e.length);
        for (var n = e.length; n < t.length; n++) t[n] = 32;
        return t;
    }, y.replaceNumberPoint = function(e, r) {
        return r ? e.replace(".", r) : e;
    }, y.getFormatString = function(e, r) {
        switch (e.type) {
          case dbtype_1.DBType.DATE:
            return r.formatDate || "";

          case dbtype_1.DBType.TIME:
            return r.formatTime || "";

          case dbtype_1.DBType.DATETIME:
          case dbtype_1.DBType.DATETIME2:
            return r.formatTimestamp || "";

          case dbtype_1.DBType.TIME_TZ:
            return r.formatTimeTZ || "";

          case dbtype_1.DBType.DATETIME_TZ:
          case dbtype_1.DBType.DATETIME2_TZ:
            return r.formatTimestampTZ || "";

          default:
            return "";
        }
    }, y;
}();

exports.DB2J = DB2J;
