// dmdb/src/convert/offRowBufferBinder.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, n) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, n) {
            t.__proto__ = n;
        } || function(t, n) {
            for (var i in n) n.hasOwnProperty(i) && (t[i] = n[i]);
        })(t, n);
    };
    return function(t, n) {
        function i() {
            this.constructor = t;
        }
        r(t, n), t.prototype = null === n ? Object.create(n) : (i.prototype = n.prototype, 
        new i());
    };
}(), __awaiter = this && this.__awaiter || function(t, u, f, c) {
    return new (f = f || Promise)(function(i, n) {
        function r(t) {
            try {
                o(c.next(t));
            } catch (t) {
                n(t);
            }
        }
        function e(t) {
            try {
                o(c.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function o(t) {
            var n;
            t.done ? i(t.value) : ((n = t.value) instanceof f ? n : new f(function(t) {
                t(n);
            })).then(r, e);
        }
        o((c = c.apply(t, u || [])).next());
    });
}, __generator = this && this.__generator || function(i, r) {
    var e, o, u, f = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (e) throw new TypeError("Generator is already executing.");
                for (;f; ) try {
                    if (e = 1, o && (u = 2 & n[0] ? o.return : n[0] ? o.throw || ((u = o.return) && u.call(o), 
                    0) : o.next) && !(u = u.call(o, n[1])).done) return u;
                    switch (o = 0, u && (n = [ 2 & n[0], u.value ]), n[0]) {
                      case 0:
                      case 1:
                        u = n;
                        break;

                      case 4:
                        return f.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        f.label++, o = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = f.ops.pop(), f.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = f.trys).length && u[u.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            f = 0;
                            continue;
                        }
                        if (3 === n[0] && (!u || n[1] > u[0] && n[1] < u[3])) {
                            f.label = n[1];
                            break;
                        }
                        if (6 === n[0] && f.label < u[1]) {
                            f.label = u[1], u = n;
                            break;
                        }
                        if (u && f.label < u[2]) {
                            f.label = u[2], f.ops.push(n);
                            break;
                        }
                        u[2] && f.ops.pop(), f.trys.pop();
                        continue;
                    }
                    n = r.call(i, f);
                } catch (t) {
                    n = [ 6, t ], o = 0;
                } finally {
                    e = u = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var offRowBinder_1 = require("./offRowBinder"), OffRowBufferBinder = function(i) {
    function t(t, n) {
        return i.call(this, t, n) || this;
    }
    return __extends(t, i), t.prototype.read = function(n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return 0 < this.getBufferLeaveLen() ? this.readFromBuffer(n) : this.readOver || (n.putBuffer(this.obj, 0, this.obj.length), 
                this.readOver = !0), [ 2 ];
            });
        });
    }, t.prototype.readAll = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.obj ];
            });
        });
    }, t.prototype.reset = function() {
        this.readOver = !1, this.buffer.clear(), this.buffer.putBuffer(this.obj, 0, this.obj.length), 
        this.readOver = !0, this.offRow = this.buffer.length > this.offRowLen;
    }, t;
}(offRowBinder_1.OffRowBinder);

exports.OffRowBufferBinder = OffRowBufferBinder;
