// dmdb/src/convert/offRowBinder.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, o, s, c) {
    return new (s = s || Promise)(function(r, n) {
        function i(t) {
            try {
                u(c.next(t));
            } catch (t) {
                n(t);
            }
        }
        function e(t) {
            try {
                u(c.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function u(t) {
            var n;
            t.done ? r(t.value) : ((n = t.value) instanceof s ? n : new s(function(t) {
                t(n);
            })).then(i, e);
        }
        u((c = c.apply(t, o || [])).next());
    });
}, __generator = this && this.__generator || function(r, i) {
    var e, u, o, s = {
        label: 0,
        sent: function() {
            if (1 & o[0]) throw o[1];
            return o[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (e) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (e = 1, u && (o = 2 & n[0] ? u.return : n[0] ? u.throw || ((o = u.return) && o.call(u), 
                    0) : u.next) && !(o = o.call(u, n[1])).done) return o;
                    switch (u = 0, o && (n = [ 2 & n[0], o.value ]), n[0]) {
                      case 0:
                      case 1:
                        o = n;
                        break;

                      case 4:
                        return s.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        s.label++, u = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(o = 0 < (o = s.trys).length && o[o.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === n[0] && (!o || n[1] > o[0] && n[1] < o[3])) {
                            s.label = n[1];
                            break;
                        }
                        if (6 === n[0] && s.label < o[1]) {
                            s.label = o[1], o = n;
                            break;
                        }
                        if (o && s.label < o[2]) {
                            s.label = o[2], s.ops.push(n);
                            break;
                        }
                        o[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    n = i.call(r, s);
                } catch (t) {
                    n = [ 6, t ], u = 0;
                } finally {
                    e = o = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var const_1 = require("../desc/const"), bufferArray_1 = require("../utils/bufferArray"), OffRowBinder = function() {
    function t(t, n, r) {
        this.readOver = !1, this.position = 0, this.offRow = !1, this.obj = t, this.encoding = n.getServerEncoding(), 
        this.readLen = n.lobReadWriteLen, this.offRowLen = n.lobOffRowLen, this.targetLength = r || const_1.Const.IGNORE_TARGET_LENGTH, 
        this.buffer = new bufferArray_1.BufferArray();
    }
    return t.prototype.init = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.read(this.buffer) ];

                  case 1:
                    return t.sent(), this.offRow = this.buffer.length > this.offRowLen, [ 2, this ];
                }
            });
        });
    }, t.prototype.getBufferLeaveLen = function() {
        return this.buffer.length;
    }, t.prototype.readFromBuffer = function(t) {
        t.append(this.buffer);
    }, t.prototype.isReadOver = function() {
        return this.readOver && 0 === this.getBufferLeaveLen();
    }, t;
}();

exports.OffRowBinder = OffRowBinder;
